// @divi/types/src/module/element/decoration/filters/index.ts
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';


export namespace Filters {
  export interface AttributeValue {
    hueRotate?: string;
    saturate?: string;
    brightness?: string;
    contrast?: string;
    invert?: string;
    sepia?: string;
    opacity?: string;
    blur?: string;
    blendMode?: 'normal' | 'multiply' | 'screen' | 'overlay' | 'darken' | 'lighten' | 'color-dodge' | 'color-burn' | 'hard-light' | 'soft-light' | 'difference' | 'exclusion' | 'hue' | 'saturation' | 'color' | 'luminosity';
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Style {

    // TODO replace `FiltersStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.Filters>;
      important?: boolean | FormatStyleImportant<Css.Properties.Filters>;
    }

  }
}
