// @divi/types/src/module/library/toggle/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface ToggleCssAttr extends Css.AttributeValue {
  openToggle?: string;
  toggleTitle?: string;
  toggleIcon?: string;
  toggleContent?: string;
}

export type ToggleCssGroupAttr = FormatBreakpointStateAttr<ToggleCssAttr>;

export interface ToggleAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: ToggleCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      open?: FormatBreakpointStateAttr<string>;
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<
      'bodyFont' |
      'inlineFont'
    >;
  };

  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  }
  closedTitle?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  }

  openToggle?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };
  openToggleIcon?: {
    decoration?: Element.Decoration.PickedAttributes<'icon'>;
  };

  closedToggle?: {
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  }
  closedToggleIcon?: {
    decoration?: Element.Decoration.PickedAttributes<'icon'>;
  };
}
