// @divi/types/src/shortcode-module/field/index.ts
/* eslint-disable @typescript-eslint/naming-convention */
import { type ShortcodeModuleAdvancedFieldDefinition } from '../advanced-field';
import {
  type ShortcodeModuleBoxShadowFieldDefault,
  type ShortcodeModuleBoxShadowFieldPresets,
} from '../advanced-field/box-shadow';
import {
  type ShortcodeModuleTextShadowFieldDefault,
  type ShortcodeModuleTextShadowFieldPresets,
} from '../advanced-field/text-shadow';


export namespace Field {
  export interface CssItem {
    label: string;
    no_space_before_selector?: boolean;
    option_category?: string;
    selector?: string;
    toggle_slug: string;
  }

  export type CssItems = Record<string, CssItem>;

  export type AdvancedItems = ShortcodeModuleAdvancedFieldDefinition;

  export interface Item {
    additional_att?: string;
    additional_button?: string;
    additional_button_type?: string;
    additional_button_attrs?: Record<string, string>;
    advanced_fields?: boolean;
    affects?: string[];
    affects_mobile?: boolean;
    after?: Record<string, string | Record<string, string>>[];
    allow_empty?: boolean;
    allowed_units?: string[];
    allowed_values?: string[] | Record<string, string[]>;
    allow_dynamic?: string[];
    attributes?: Record<string, string> | string;
    attr_suffix?: string;
    background_fields?: Record<string, Item>;
    background_tab?: string;
    base_name?: string;
    bb_support?: boolean;
    button_options?: Record<string, string>;
    class?: string | string[];
    classes?: string;
    className?: string;
    composite_structure?: Record<string, CompositeStructure>;
    composite_type?: string;
    computed_affects?: string[];
    computed_callback?: string[];
    computed_depends_on?: string[];
    computed_minimum?: string[];

    // This looks like a typo but this exist on et_pb_wc_price module.
    computed_minumum?: string[];
    computed_variables?: Record<string, string>;
    checkbox?: boolean;
    choose_text?: string;
    context?: string;
    copy_with?: string[];
    custom_color?: boolean;
    data_type?: string;
    default?:
    boolean |
    string |
    number |
    Record<number, string |
      Record<string, string>> |
    ShortcodeModuleBoxShadowFieldDefault;
    default_from?: string;
    default_tablet?: string | Record<number, string | Record<string, string>>;
    default_phone?: string | Record<number, string | Record<string, string>>;
    default_on_child?: boolean;

    // Looks like typo, but this exist on et_pb_fullwidth_menu module.
    default_on_fron?: string;
    default_on_front?: string;
    default_sticky?: string;
    default_value_depends?: string | string[];
    default_values_mapping?: Record<string, string>;
    depends_default?: null | string; // TODO (D5) Need to be double checked.
    depends_on?: string | string[];
    depends_on_responsive?: string[];
    depends_show_if?: string | string[];
    depends_show_if_not?: string | string[];
    default_unit?: string;
    description?: string;
    displayRecent?: boolean;
    display_if?: boolean;
    dynamic_content?: string;
    fixed_range?: boolean;
    field_template?: string;
    fixed_unit?: string;
    flip?: string;
    group_label?: string;
    hover?: string;
    hide_metadata?: boolean;
    hide_sync?: boolean;
    i10n?: Record<string, string>;
    is_fb_content?: boolean;
    is_global_default?: boolean;
    label?: string;
    labels?: Record<string, string>;
    last_edited?: string;
    linked_responsive?: string[];
    main_tab_setting?: string;
    max_accounts?: number;
    meta_categories?: Record<string, string>;
    message?: string;
    multi_selection?: boolean;
    mobile_options?: boolean;
    mode?: string;
    name?: string;
    no_space_before_selector?: boolean;
    not_required?: boolean;
    number_validation?: boolean;
    option_category?: string;
    options_category?: string;
    option_class?: string;
    options?: Options;
    options_icon?: string;
    order?: string[];
    overwrite_onchange?: string[];
    post_type?: string;
    presets?:
    ShortcodeModuleBoxShadowFieldPresets |
    ShortcodeModuleTextShadowFieldPresets;
    priority?: number | string;
    radio?: boolean;
    readonly_if?: Record<string, string[]>;
    readonly_if_not?: Record<string, string[]>;
    renderer_options?: Record<string, boolean | string>;
    range_settings?: Record<string, number> | Record<string, string>;
    reset_animation?: boolean;
    resolver?: string;
    responsive?: boolean;
    responsive_affects?: string[];
    right_actions?: string;
    right_actions_readonly?: string;
    searchable?: boolean;
    selector?: string;
    shortcut_index?: string;
    show_if?: Record<string, string | string[]>;
    show_if_not?: Record<string, string | string[]>;
    specialty_only?: boolean | string;
    sticky?: boolean;
    sub_toggle?: null | string | boolean; // TODO (D5) Need to be double checked.
    sync_affects?: string[];
    tab_filler?: boolean;
    tab_slug?: string;
    taxonomy_name?: string;
    toggleable?: boolean;
    toggle_slug?: string;
    tooltip?: string;
    type?: string;
    unitless?: boolean;
    upload_button_text?: string;
    update_text?: string;
    use_container_wrapper?: boolean;
    validate_input?: boolean;
    validate_unit?: boolean;
    value?: boolean;
    value_type?: string;
    value_min?: number;
    value_max?: number;
    value_overwrite?: Record<string, string>;
  }

  export type Items = Record<string, Item>;

  export type CompositeStructure = {
    controls?: Record<string, Item>;
    icon?: string;
    tooltip?: string;
    label?: string;
  }

  export type Options = Record<string,
    string | Record<string, string | Record<string, string>>
  > |
    string[];

  export type ValuePreset = Record<number, string | Record<string, string>>;

  export type AdvancedDefaults = Record<string,
    string |
    number |
    boolean |
    ShortcodeModuleBoxShadowFieldDefault |
    ShortcodeModuleTextShadowFieldDefault
  >;

}
