// @divi/types/src/module/settings/group/library/position/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Position {
  export type Sides = 'top' | 'right' | 'bottom' | 'left';

  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Position.Attributes;
    fields?: {
      mode?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Position.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      originRelative?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Position.Attributes;
        component?: FieldLibrary.ComponentMap['divi/corner-picker'];
      };
      originAbsolute?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Position.Attributes;
        component?: FieldLibrary.ComponentMap['divi/cornerside-picker'];
      };
      originFixed?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Position.Attributes;
        component?: FieldLibrary.ComponentMap['divi/cornerside-picker'];
      };
      offsetVertical?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Position.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      offsetHorizontal?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Position.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
    }
  }
}
