// @divi/types/src/module/edit/index.ts
import {
  type ElementType,
  type MutableRefObject,
  type ReactElement,
  type ReactNode,
} from 'react';

import { type ModuleElementsType } from '../class-module-elements';
import { type Classnames } from '../classnames';
import { type Element } from '../element';
import { type ScriptData } from '../script-data';
import { type ModuleFlatObject } from '../state';
import { type Styles } from '../styles';


export namespace Edit {
  export namespace Component {

    // TODO Replace `ModuleBaseProps` with this.
    export interface BaseProps<
      TAttrs,
      TParentAttrs = unknown,
    > {
      id: string;
      name: string;
      className?: string;
      children?: ReactNode;
      childrenIds?: string[];
      hasModule?: boolean;
      isLast?: boolean;
      isFirst?: boolean;
      hasModuleClassName?: boolean;
      hasHoveredClassName?: boolean;
      parentAttrs?: TParentAttrs;
      parentId?: string;
      parentName?: string;
      stylesComponent?: (props: Styles.Component.Props<TAttrs>) => ReactElement;
      scriptDataComponent?: (props: ScriptData.Component.Props<TAttrs>) => ReactElement;

      // feat(D5, Dev) :: To be removed
      moduleClassName?: string;
      moduleOrderClassName?: string;

      // feat(D5, Dev) :: Once all module has this, make this mandatory.
      classnamesFunction?: (params: Classnames.Params<TAttrs>) => void;
      wrapperClassnamesFunction?: (params: Classnames.Params<TAttrs>) => void;

      htmlAttributesFunction?: (params: Element.Advanced.IdClasses.Attributes) => Record<'id' | 'classNames', string>;
    }

    // TODO Replace `ModuleContainerProps` with this.
    export type ContainerProps<TAttrs> = Omit<BaseProps<TAttrs>, 'moduleClassName' | 'moduleOrderClassName'> & {
      attrs: TAttrs;
      parentAttrs?: TAttrs;
      defaultPrintedStyleAttrs?: TAttrs;
      domRef?: MutableRefObject<HTMLDivElement>;
      elements?: ModuleElementsType<TAttrs>;
      htmlAttrs?: Record<string, string | number>;
      innerHTML?: string;
      originalName?: string;
      siblingAttrs?: {
        previous?: TAttrs;
        next?: TAttrs;
      };
      pageSettings?: {
        sectionBackgroundColor?: string,
      }
      tag?: ElementType;
      hasModuleWrapper?: boolean;
      wrapperTag?: ElementType;
      wrapperHtmlAttrs?: Record<string, string | number>;
      wrapperClassname?: string;
      cssPosition?: 'before' | 'after';
    }

    // TODO Replace `ModuleEditProps` with this.
    export interface Props<
      TAttrs,
      TParentAttrs = unknown,
      TChildAttrs = unknown
    > {
      attrs: TAttrs;
      defaultPrintedStyleAttrs?: TAttrs;
      childModules?: ModuleFlatObject<TChildAttrs>[];
      isFirst?: boolean;

      // Params can be anything at this moment, hence unknown.
      // TODO feat(Refactor) Provide stricter type / interface
      parentCallback?: (params?: unknown) => void;
      parentName?: string;
      parentAttrs?: TParentAttrs;
      siblingAttrs?: {
        previous?: TAttrs;
        next?: TAttrs;
      }
      pageSettings?: {
        sectionBackgroundColor?: string,
      },
      elements?: ModuleElementsType<TAttrs>;
    }
  }
}
