// @divi/types/src/module/library/image/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type Decoration } from '../../element/decoration';
import { type InternalAttrs } from '../internal';


export interface ImageSpacingAttr extends Element.Decoration.Spacing.AttributeValue {
  showBottomSpace?: OnOff;
}

export type ImageSpacingGroupAttr = FormatBreakpointStateAttr<ImageSpacingAttr>;

export interface ImageAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      align?: FormatBreakpointStateAttr<string>;
      sizing?: Element.Decoration.Sizing.Attributes;
      spacing?: ImageSpacingGroupAttr;
      forceFullwidth?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Image
  image?: Element.Types.ImageLink.Attributes & {
    decoration?: Decoration.PickedAttributes<
    'border' |
    'boxShadow'
    >;
  };
}
