// @divi/types/src/module/library/team-member/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface TeamMemberAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  image?: {
    innerContent?: FormatBreakpointStateAttr<{
      url?: string;
      animation?: string;
    }>;
    decoration?: Element.Decoration.PickedAttributes<
      'border' |
      'boxShadow' |
      'filters'
    >;
  };

  social?: {
    innerContent?: FormatBreakpointStateAttr<{
      facebookUrl?: string;
      twitterUrl?: string;
      googleUrl?: string;
      linkedinUrl?: string;
    }>;
    decoration?: Element.Decoration.PickedAttributes<'icon'>;
  }

  // Name/Title
  name?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Content
  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<
      'bodyFont' |
      'inlineFont'
    >;
  };

  // Position
  position?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };
}
