// @divi/types/src/global-data/presets/index.ts
import { type ModuleAttrs } from '../../module';


export namespace Presets {

  // In the future, there will be presets on different level of settings. Hence the `module` property.
  export type Items = {
    module: Record<Module.Name, {
      items: Module.Items;

      /**
       * ID of the default preset currently set.
       * Modules without a `modulePreset` attribute will apply this preset.
       */
      default?: string;
    }>;
  }

  /**
   * In D4, there is only module-level preset.
   * In the future, there will be presets on different level of settings.
   */
  export namespace Module {

    // Refers to the preset's name. Easier to be read than `string`.
    export type Name = string;

    // Refers to preset item id. Easier to be read than `string`.
    export type Id = string;

    export type Item = {
      type: 'module';
      id: Id;
      name: string;
      created: number;
      updated: number;
      version: string;
      attrs?: ModuleAttrs;
      renderAttrs?: ModuleAttrs;
      styleAttrs?: ModuleAttrs;
      moduleName: string;
    }

    // eslint-disable-next-line no-shadow
    export type Items = Record<Id, Item>;
  }
}
