// @divi/types/src/field-library/spam-protection-account/index.ts
import { type FormEvent } from 'react';
import { type ImmutableObject } from 'seamless-immutable';

import {
  type AttrState,
  type Breakpoint,
} from '../../format-attr';
import { type SpamProtection } from '../../module/element/advanced/spam-protection';
import {
  type SpamProtection as SpamProtectionNamespace,
} from '../../spam-protection';


export namespace SpamProtectionAccount {

  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: string;
  }

  export interface OnRemoveCallbackParams {
    provider: string;
    account: string;
  }

  export interface OnAddCallbackParams extends OnRemoveCallbackParams {
    data: Record<string, string>;
  }

  export type OnChangeCallback = (params: OnChangeCallbackParams) => void;
  export type OnRemoveCallback = (params: OnRemoveCallbackParams) => void;
  export type OnAddCallback = (params: OnAddCallbackParams) => void;

  export interface ContainerProps {
    appBreakpoint?: Breakpoint;
    appState?: AttrState;
    attr?: SpamProtection.Attributes;
    defaultAttr?: Partial<SpamProtection.Attributes>;
    name?: string;
  }

  export interface Props {
    defaultValue?: string;
    name: string;
    onAdd: OnAddCallback;
    onChange: OnChangeCallback;
    onRemove: OnRemoveCallback;
    provider: ImmutableObject<SpamProtectionNamespace.Service.Provider>;
    value: string;
  }
}
