// @divi/types/src/module/library/video-slider/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface SliderFieldOption {
  label: string;
}

export interface SliderControlsAttr {
  color?: 'dark' | 'light';
  useArrows?: OnOff;
  useThumbnails?: 'on' | 'off';
}

export interface VideoSliderCssAttr extends Css.AttributeValue {
  playButton?: string;
  thumbnailItem?: string;
  sliderArrows?: string;
}
export type VideoSliderCssGroupAttr = FormatBreakpointStateAttr<VideoSliderCssAttr>;

export interface VideoSliderAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: VideoSliderCssGroupAttr;

  // Module.
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'spacing' |
      'sizing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Video.
  video?: {
    decoration?: Element.Decoration.PickedAttributes<
      'border' |
      'boxShadow'
    >;
  };

  // Video Play Icon.
  playIcon?: {
    decoration?: Element.Decoration.PickedAttributes<'icon'>;
  };

  // Video Overlay.
  overlay?: {
    advanced?: FormatBreakpointStateAttr<{
      showImageOverlay?: OnOff;
    }>
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  };

  // Video Overlay Thumbnail Image.
  thumbnail?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
    }>
  };

  // Slider Controls.
  sliderControls?: {
    advanced?: FormatBreakpointStateAttr<{
      color?: 'dark' | 'light';
      useArrows?: OnOff;
      useThumbnails?: 'on' | 'off';
    }>
  };

}
