// @divi/types/src/module/settings/group/library/sizing/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Sizing {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Sizing.Attributes;
    fields?: {
      width?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sizing.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      maxWidth?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sizing.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      alignment?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sizing.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      minHeight?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sizing.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      height?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sizing.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      maxHeight?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sizing.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
    }
  }
}
