// @divi/types/src/field-library/transform-origin/index.ts
import { type NumericInput } from '../common-numeric-input';


export namespace TransformOrigin {

  /**
   * @typedef {('x'|'y')} Axis
   */
  export type Axis = 'x' | 'y';

  /**
   * @typedef {object} Value
   * @property {string} x X value.
   * @property {string} y Y value.
   */
  export interface Value {
    x: string;
    y: string;
  }

  /**
   * @typedef {object} ValueMappings
   * @property {object[]} x X value.
   * @property {object[]} y Y value.
   */
  export interface ValueMappings {
    x: NumericInput.ValueMapping[];
    y: NumericInput.ValueMapping[];
  }

  /**
   * @typedef {object} ValuePreset
   * @property {object} x X value.
   * @property {object} y Y value.
   */
  export interface ValuePreset {
    x: NumericInput.ValueMapping;
    y: NumericInput.ValueMapping;
  }

  /**
   * @typedef {object} Step
   * @property {number} x X value.
   * @property {number} y Y value.
   */
  export interface Step {
    x: number;
    y: number;
  }

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {object} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    inputValue?: Value;
  }

  /**
   * @typedef {object} Props
   * @property {string[]} [allowedUnits] Allowed units.
   * @property {string} [className] Class name.
   * @property {string} [defaultUnit] Default unit.
   * @property {object} [defaultValue] Default value.
   * @property {number} [max] Max value.
   * @property {number} [min] Min value.
   * @property {string} [name] Name.
   * @property {boolean} ["readonly"] Read only.
   * @property {object} value Value.
   * @property {Function} [formatter] Formatter.
   * @property {Function} onChange On change callback.
   */
  export interface Props {
    allowedUnits?: string[];
    className?: string;
    defaultUnit?: string;
    defaultValue?: Value;
    max?: number;
    min?: number;
    name?: string;
    readonly?: boolean;
    value: Value;
    formatter?: NumericInput.Formatter;
    onChange: (params: OnChangeCallbackParams) => void;
  }
}
