// @divi/types/src/module/library/accordion/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface AccordionCssAttr extends Css.AttributeValue {
  toggle?: string;
  openToggle?: string;
  toggleTitle?: string;
  toggleIcon?: string;
  toggleContent?: string;
}

export type AccordionCssGroupAttr = FormatBreakpointStateAttr<AccordionCssAttr>;

export interface AccordionAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: AccordionCssGroupAttr;

  // Module.
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'spacing' |
      'sizing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Title.
  title?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Content.
  content?: {
    decoration?: Element.Decoration.PickedAttributes<'bodyFont'>;
  };

  // Open Toggle.
  openToggle?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };

  // Closed Toggle.
  closedToggle?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };

  // Closed Toggle Icon.
  closedToggleIcon?: {
    decoration?: Element.Decoration.PickedAttributes<'icon' | 'font'>;
  };
}
