// @divi/types/src/global-data/store/index.ts
import { type CSSProperties } from 'react';
import {
  type ImmutableObject,
} from 'seamless-immutable';

import { type Data as DataNamespace } from '../data';
import { type Presets as PresetsNamespace } from '../presets';


export namespace Store {

   /**
   * The Global Data mutable store state.
   *
   * This is the state that is used in the reducer, and is not immutable.
   *
   * @since ??
   */
  export interface State {
    globalColors: DataNamespace.Colors;

    /**
     * List of presets.
     */
    presets: PresetsNamespace.Items;

    /**
     * Flag to indicate whether preset modal is active.
     */
    isActivePreset: boolean;

    fetchLog: Record<string, string[]>,
    fetchError: Record<string, string[]>,
  }

  /**
   * The Global Data immutable store state.
   *
   * This is the state that is used in the components, and is immutable.
   *
   * @since ??
   */
  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {

    /**
     * A function that dispatches an action within the store.
     *
     * @since ??
     *
     * @example
     *
     * ```ts
     * // Dispatching an action within the store
     * store.dispatch();
     * ```
     */
    dispatch: () => void;

    /**
     * A function that retrieves the current state of the store.
     *
     * @since ??
     *
     * @example
     *
     * ```ts
     * // Retrieving the current state of the store
     * const currentState = store.getState();
     * ```
     */
    getState: () => ImmutableState;
  }

  export namespace Actions {

    /**
     * Represents an action to add a global color.
     *
     * This interface defines the properties of an object that represents an action to add a global color palette.
     * It is typically used as the type of the action in a reducer function.
     *
     * @since ??
     */
    export namespace AddGlobalColor {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {

        /**
         * The type of the action. This should always be set to 'ADD_GLOBAL_COLOR'.
         *
         * @since ??
         */
        type: 'ADD_GLOBAL_COLOR',

        /**
         * The new color value. It should be a string representing a color in a valid CSS format.
         *
         * @since ??
         *
         * @example '#ff0000'
         * @example 'rgb(255, 0, 0)'
         * @example 'rgba(255, 0, 0, 1)'
         * @example 'hsl(0, 100%, 50%)'
         * @example 'hsla(0, 100%, 50%, 1)'
         * @example 'red'
         */
        color: CSSProperties['color'],
      }
    }


    /**
     * Represents an action to delete a global color palette.
     *
     * This interface defines the properties of an object that represents an action to delete a global color palette.
     * It is typically used as the type of the action in a reducer function.
     *
     * @since ??
     */
    export namespace DeleteGlobalColor {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {

        /**
         * The type of the action. This should always be set to 'UPDATE_GLOBAL_COLOR'.
         *
         * @since ??
         */
        type: 'DELETE_GLOBAL_COLOR',

        /**
         * The ID of the color to delete. It should be a unique identifier for a color in the palette.
         *
         * @since ??
         */
        id: string,
      }
    }

    /**
     * Represents an action to update a global color palette.
     *
     * This interface defines the properties of an object that represents an action to update a global color palette.
     * It is typically used as the type of the action in a reducer function.
     *
     * @since ??
     */

    export namespace UpdateGlobalColor {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {

        /**
         * The type of the action. This should always be set to 'UPDATE_GLOBAL_COLOR'.
         *
         * @since ??
         */
        type: 'UPDATE_GLOBAL_COLOR',

        /**
         * The new color value. It should be a string representing a color in a valid CSS format.
         *
         * @since ??
         *
         * @example '#ff0000'
         * @example 'rgb(255, 0, 0)'
         * @example 'rgba(255, 0, 0, 1)'
         * @example 'hsl(0, 100%, 50%)'
         * @example 'hsla(0, 100%, 50%, 1)'
         * @example 'red'
         */
        color: CSSProperties['color'],

        /**
         * The ID of the color to update. It should be a unique identifier for a color in the palette.
         *
         * @since ??
         */
        id: string,
      }
    }

    /**
     * Represents an action to update the global colors' palette.
     *
     * This interface defines the properties of an object that
     * represents an action to update the global colors' palette.
     * It is typically used as the type of the action in a reducer function.
     *
     * @since ??
     */
    export namespace UpdateGlobalColors {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {

        /**
         * The type of the action. This should always be set to 'UPDATE_GLOBAL_COLORS'.
         *
         * @since ??
         */
        type: 'UPDATE_GLOBAL_COLORS',

        /**
         * The global colors' palette.
         *
         * @since ??
         */
        palette: DataNamespace.Colors;
      }
    }

    export type FunctionMap = {
      addGlobalColor: (color: AddGlobalColor.Params['color']) => AddGlobalColor.ReturnValue;
      deleteGlobalColor: (id: DeleteGlobalColor.Params['id']) => DeleteGlobalColor.ReturnValue;
      updateGlobalColor: (id: UpdateGlobalColor.Params['id'], color: UpdateGlobalColor.Params['color']) => UpdateGlobalColor.ReturnValue;
      updateGlobalColors: (palette: UpdateGlobalColors.Params['palette']) => UpdateGlobalColors.ReturnValue;
    };

    export type ReturnValue =
      AddGlobalColor.ReturnValue |
      DeleteGlobalColor.ReturnValue |
      UpdateGlobalColor.ReturnValue;
  }

  export namespace Selectors {
    export namespace GetModulePreset {
      export type Params = {
        moduleName: keyof PresetsNamespace.Items['module'];
        id: keyof PresetsNamespace.Items['module'][keyof PresetsNamespace.Items['module']]['items'];

        /**
         * Whether to use the default preset as fallback if the preset is not found.
         */
        useDefaultFallback?: boolean;
      }

      export type ReturnValue = ImmutableState['presets']['module'][Params['moduleName']]['items'][Params['id']];
    }

    export namespace GetModulePresets {
      export type Params = {
        presets: Record<string, GetModulePreset.Params>
      }

      export type ReturnValue = Record<string, GetModulePreset.ReturnValue>;
    }

    export namespace GetModuleDefaultPreset {
      export type Params = {
        moduleName: keyof PresetsNamespace.Items['module'];
      }

      export type ReturnValue = ImmutableObject<PresetsNamespace.Module.Item>;
    }

    export namespace GetModuleDefaultPresetID {
      export type Params = {
        moduleName: keyof PresetsNamespace.Items['module'];
      }

      export type ReturnValue = ImmutableState['presets']['module'][Params['moduleName']]['default'];
    }

    export type FunctionMap = {
      getGlobalColors: () => DataNamespace.Colors;
      getGlobalColorsByStatus: (status: string) => DataNamespace.Colors;
      getGlobalColorsToExport: (ids: string[]) => DataNamespace.ExportColors;
      getGlobalColorValue: (colorId: string) => DataNamespace.Color['color'];
      getModulePreset: (params: GetModulePreset.Params) => GetModulePreset.ReturnValue;
      getModulePresets: (Params: GetModulePresets.Params) => GetModulePresets.ReturnValue;
      getModuleDefaultPreset: (params: GetModuleDefaultPreset.Params) => GetModuleDefaultPreset.ReturnValue;
      getModuleDefaultPresetID: (params: GetModuleDefaultPresetID.Params) => GetModuleDefaultPresetID.ReturnValue;
      isActivePreset: () => boolean;
      isActiveGlobalColor: (colorId: string) => boolean;
    }
  }
}
