// @divi/types/src/module/settings/group/library/filters/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Filters {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Filters.Attributes;
    fields?: {
      hueRotate?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      saturate?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      brightness?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      contrast?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      invert?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      sepia?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      opacity?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      blur?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      blendMode?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Filters.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
    }
  }
}
