// @divi/types/src/module/metadata/element/index.ts
import { type Element as ElementType } from '../../element';
import { type Decoration } from '../../element/decoration';
import { type Type as ModuleElementType } from '../../element/types';
import { type Settings } from '../settings';


export namespace Element {

  /**
   * Module element's properties.
   */
  export interface Properties<TAttributes = ElementType.Attributes> {
    type?: 'object';
    allowHtml?: boolean;
    allowEmptyValue?: boolean;
    attributes?: Record<string, string>;
    childrenSanitizer?: string;
    default?: TAttributes;
    defaultPrintedStyle?: TAttributes;
    elementProps?: Record<string, unknown>;
    elementType?: ModuleElementType;
    inlineEditor?: 'plainText' | 'richText' | false;
    label?: string;
    selector?: string;
    tagName?: keyof HTMLElementTagNameMap;
    styleProps?: unknown;
    styleComponentsProps?: unknown;
    scriptDataProps?: unknown;
    settings?: {

      // Inner content always has direct BreakpointStateAttr value.
      innerContent?: Partial<Settings.AttributeGroup.Value> | Settings.AttributeGroup.DefaultValue;

      // The rest of the properties are object of BreakpointStateAttr value.
      decoration?: Partial<Record<
        keyof Decoration.Attributes, Partial<Settings.AttributeGroup.Value> | Settings.AttributeGroup.DefaultValue
      >>;
      advanced?: {
        [key: string]: Partial<Settings.AttributeGroup.Value> | Settings.AttributeGroup.DefaultValue;
      };
      meta?: {
        [key: string]: Partial<Settings.AttributeGroup.Value> | Settings.AttributeGroup.DefaultValue;
      };
    };
  }
}
