// @divi/types/src/module/element/types/image-link/index.ts
import { type OnOff } from '../../../../attr';
import { type FormatBreakpointStateAttr } from '../../../../format-attr';
import { type Icon } from '../../../../icon';


export namespace ImageLink {
  export namespace InnerContent {
    export interface AttributeValue {
      src?: string;
      alt?: string;
      titleText?: string;
      linkUrl?: string;
      linkTarget?: string;
    }

    export type Attributes = FormatBreakpointStateAttr<AttributeValue>;
  }

  export namespace Advanced {

    // TODO refactor this
    // Extending Record<string, unknown> is needed to ensure compatibility with `ModuleElementAttrAdvanced`
    // which is used at `ModuleElementAttr`. This is however needs to be refactored for better
    export interface Attributes extends Record<string, unknown> {
      lightbox?: FormatBreakpointStateAttr<OnOff>;
      overlay?: FormatBreakpointStateAttr<{
        use?: OnOff;
        backgroundColor?: string;
      }>;
      overlayIcon?: FormatBreakpointStateAttr<{
        hoverIcon?: Icon.Font.AttributeValue;
        iconColor?: string;
      }>;
    }
  }

  export interface Attributes {
    innerContent?: InnerContent.Attributes;
    advanced?: Advanced.Attributes;
  }
}
