// @divi/types/src/field-library/range/index.ts
import { type MouseEvent } from 'react';

import { type NumericInput } from '../common-numeric-input';


export namespace Range {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {MouseEvent} [event] Event.
   * @property {string} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: MouseEvent;
    inputValue?: string;
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} [className] Class name.
   * @property {Function} onChange On change callback.
   * @property {string} value Value.
   * @property {string} [defaultValue] Default value.
   * @property {number} [min] Min value.
   * @property {number} [max] Max value.
   * @property {number} [minLimit] Min limit.
   * @property {number} [maxLimit] Max limit.
   * @property {number} [step] Step.
   * @property {boolean} ["readonly"] Read only.
   * @property {boolean} [allowEmpty] Allow empty.
   * @property {string} [getDefaultFrom] Get default from.
   * @property {object} [defaultValuesMapping] Default values mapping.
   * @property {string[]} [allowedValues] Allowed values.
   * @property {string[]} [allowedUnits] Allowed units.
   * @property {string} [defaultUnit] Default unit.
   * @property {string} [placeholder] Placeholder.
   * @property {object[]} [valueMapping] Value mapping.
   */
  export interface Props {
    name: string;
    className?: string;
    onChange: (params: OnChangeCallbackParams) => void;
    value: string;
    defaultValue?: string;
    min?: number;
    max?: number;
    minLimit?: number;
    maxLimit?: number;
    step?: number;
    readonly?: boolean;
    allowEmpty?: boolean;
    getDefaultFrom?: string;
    defaultValuesMapping?: Record<string, string>;
    allowedValues?: string[];
    allowedUnits?: string[];
    defaultUnit?: string;
    placeholder?: string;
    valueMapping?: NumericInput.ValueMapping[];
  }
}
