// @divi/types/src/modal-library/d4-design/index.ts
// Types of modal library prior to D5 Enhanced design implementation.
// These types are needed mostly for transforming D4 design modal state into enhanced design format.
export namespace D4Design {
  export type Modal<TAttributes = Record<string, any>> = {
    name: string;
    owner: string;
    dimension: {
      height: number;
      width: number;
    };
    offset: {
      top: number;
      left: number;
    };
    centered?: boolean;

    // Modal specific configuration.
    attributes?: TAttributes;

    tab: string;
    type: 'single' | 'multi' | 'floating' | 'confirmation';
    group: Record<string, string>;
    groupTab: Record<string, string>;
    expand: boolean;
    snap: false | 'right' | 'bottom' | 'left';
    panelSize?: number;
    bodySiblingHeight?: number;
    isResizing?: boolean;
  }

  export type ModalsProps = {
    single: Modal;
    multi: Record<string, Modal>;
    floating: Modal;
    confirmation: Record<string, Modal>;
  }

  export type ModalTypes = keyof ModalsProps;
}
