// @divi/types/src/email-marketing/service/index.ts
export namespace Service {
  export interface NewAccountField {
    label: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    not_required?: boolean;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    apply_password_mask?: boolean;
    required?: 'http' | 'https';
  }
  export interface CustomField {
    field_id?:    string|number;
    field_tag?:   string;
    hidden?:      boolean;
    name?:        string;
    native_type?: string;
    type?:        string;
  }
  export interface CustomFields {
    [key: string|number]: CustomField;
  }
  export interface Account {
    provider: string;
    slug: string;
    credentials?: Record<string, string>;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    is_authorized?: boolean;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    custom_fields?: CustomFields;
    lists?: Record<string, List>;
    custom_data?: Record<string, unknown>;
  }
  export interface List {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    list_id?: string;
    name?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    subscribers_count?: number;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    custom_fields?: CustomFields;
  }
  export interface Provider {
    slug: string;
    label: string;
    newAccountFields?: Record<string, NewAccountField>;
    accounts?: Record<string, Account>;
    customFields?: {
      enable?: boolean;
      isPredefined?: boolean;
    }
    nameFields: {
      name: boolean;
      useSingleNameField: boolean;
      showFirstNameField: boolean;
      showLastNameField: boolean;
    };
  }
}
