// @divi/types/src/field-library/common-preview/index.ts
import {
  type CSSProperties,
  type MouseEvent,
  type ReactNode,
} from 'react';


export namespace Preview {

  /**
   * @typedef {object} Props
   * @property {ReactNode} [children] Children.
   * @property {string} [className] Class name.
   * @property {string} [caption] Caption.
   * @property {string} [addTitle] Add title.
   * @property {CSSProperties} [previewStyle] Preview style.
   * @property {boolean} [showAddButton] Show add button.
   * @property {boolean} [showEditButton] Show edit button.
   * @property {boolean} [showSwapButton] Show swap button.
   * @property {boolean} [showRemoveButton] Show remove button.
   * @property {boolean} [showResetButton] Show reset button.
   * @property {boolean} [showReverseButton] Show reverse button.
   * @property {boolean} [isEmptyValue] Is empty value.
   * @property {Function} [onClickCanvas] On click canvas callback.
   * @property {Function} [onClickAddButton] On click add button callback.
   * @property {Function} [onClickEditButton] On click edit button callback.
   * @property {Function} [onClickSwapButton] On click swap button callback.
   * @property {Function} [onClickRemoveButton] On click remove button callback.
   * @property {Function} [onClickResetButton] On click reset button callback.
   * @property {Function} [onClickReverseButton] On click reverse button callback.
   */
  export interface Props {
    children?: ReactNode;
    className?: string;
    caption?: string;
    addTitle?: string;
    previewStyle?: CSSProperties;
    showAddButton?: boolean;
    showEditButton?: boolean;
    showSwapButton?: boolean;
    showRemoveButton?: boolean;
    showResetButton?: boolean;
    showReverseButton?: boolean;
    isEmptyValue?: boolean;
    onClickCanvas?: (event: MouseEvent) => void;
    onClickAddButton?: (event: MouseEvent) => void;
    onClickEditButton?: (event: MouseEvent) => void;
    onClickSwapButton?: (event: MouseEvent) => void;
    onClickRemoveButton?: (event: MouseEvent) => void;
    onClickResetButton?: (event: MouseEvent) => void;
    onClickReverseButton?: (event: MouseEvent) => void;
  }
}
