// @divi/types/src/module/library/fullwidth-portfolio/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface FullwidthPortfolioThumbnail {
    src: string;
    width: number;
    height: number;
    altText?: string;
}

export interface FullwidthPortfolioPost {
    id: number;
    title: string;
    date?: string;
    permalink: string;
    orientation: string;
    thumbnail: FullwidthPortfolioThumbnail;
    classNames: string[];
}

export interface FullwidthPortfolioCssAttr extends Css.AttributeValue {
    portfolioTitle?: string;
    portfolioItem?: string;
    portfolioOverlay?: string;
    portfolioItemTitle?: string;
    portfolioMeta?: string;
    portfolioArrows?: string;
}
export type FullwidthPortfolioCssGroupAttr = FormatBreakpointStateAttr<FullwidthPortfolioCssAttr>;

export interface FullwidthPortfolioAttrs extends InternalAttrs {

    // CSS options is used across multiple elements inside the module thus it deserves its own top property.
    css?: FullwidthPortfolioCssGroupAttr;

    module?: {
        meta?: Element.Meta.Attributes;
        advanced?: {
            autoRotate?: FormatBreakpointStateAttr<OnOff>;
            autoRotateSpeed?: FormatBreakpointStateAttr<string>;
            htmlAttributes?: Element.Advanced.IdClasses.Attributes;
            link?: Element.Advanced.Link.Attributes;
            text?: Element.Advanced.Text.Attributes;
        };
        decoration?: Element.Decoration.PickedAttributes<
            'animation' |
            'background' |
            'border' |
            'boxShadow' |
            'disabledOn' |
            'filters' |
            'overflow' |
            'position' |
            'scroll' |
            'sizing' |
            'spacing' |
            'sticky' |
            'transform' |
            'transition' |
            'zIndex'
        >;
    };

    // Title.
    title?: {
        innerContent?: FormatBreakpointStateAttr<string>;
        decoration?: Element.Decoration.PickedAttributes<'font'>;
    };

    // Meta.
    meta?: {
        decoration?: Element.Decoration.PickedAttributes<'font'>;
    };

    // Portfolio.
    portfolio?: {
        innerContent?: FormatBreakpointStateAttr<{
            postsNumber?: string;
            includedCategories?: string[];
        }>;
        advanced?: {
            showTitle?: FormatBreakpointStateAttr<OnOff>;
            showDate?: FormatBreakpointStateAttr<OnOff>;
            layout?: FormatBreakpointStateAttr<OnOff>;
        }
        decoration?: Element.Decoration.PickedAttributes<'font'>;
    }

    // Portfolio Image.
    image?: {
        decoration?: Element.Decoration.PickedAttributes<
            'border' |
            'boxShadow' |
            'filters'
        >;
    };

    // Portfolio Item Image Overlay.
    overlay?: {
        decoration?: Element.Decoration.PickedAttributes<
            'background' |
            'icon'
        >;
    }
}
