// @divi/types/src/shortcode-module/advanced-field/border/index.ts
/* eslint-disable @typescript-eslint/naming-convention */
import { type Field } from '../../field';


export interface ShortcodeModuleAdvancedFieldDefinitionBorder {
  additional_elements?: Record<number, Record<string, string[]>>
  color_type?: string;
  css?: {
    defaults?: ShortcodeModuleAdvancedFieldDefinitionBorder['defaults'];
    main?: string | {
      border_radii?: string;
      border_radii_hover?: string;
      border_styles?: string;
      border_styles_hover?: string;
      defaults?: {
        border_radii?: string;
        border_styles?: Record<string, string>;
      }
    };
    important?: string | string[] | boolean;
  };
  defaults?: {
    border_radii?: string;
    border_styles?: {
      color?: string;
      style?: string;
      width?: string;
    }

    // This looks like a typo but this exist on et_pb_wc_checkout_payment_info module.
    border_style?: {
      width?: string;
      style?: string;
    }
    composite?: Record<string, Record<string, string>>;
  };
  depends_on?: Field.Item['depends_on'];
  depends_show_if_not?: Field.Item['depends_show_if_not'];
  depends_show_if?: Field.Item['depends_show_if'];
  fields_after?: Record<string, Field.Item>;
  important?: string;
  label_prefix?: string;
  name?: string;
  option_category?: Field.Item['option_category'];
  sub_toggle?: Field.Item['sub_toggle'];
  suffix?: string;
  tab_slug?: Field.Item['tab_slug'];
  toggle_slug?: Field.Item['toggle_slug'];
  use_radius?: boolean;
  use_focus_borders?: boolean;
}
