// @divi/types/src/shortcode-module/advanced-field/form-field/index.ts
/* eslint-disable @typescript-eslint/naming-convention */
import { type ShortcodeModuleAdvancedFieldDefinitionBorder } from '../border';
import { type ShortcodeModuleAdvancedFieldDefinitionBoxShadow } from '../box-shadow';
import { type ShortcodeModuleAdvancedFieldDefinitionFont } from '../font';
import { type ShortcodeModuleAdvancedFieldDefinitionSpacing } from '../spacing';


export interface ShortcodeModuleAdvancedFieldDefinitionFormField {
  alternating_background_color?: boolean | Record<string, string>;
  background_color?: boolean | Record<string, string>;
  label?: string;
  main?: string;
  css?: {
    alternating_background_color?: string;
    alternating_background_color_hover?: string;
    background_color_hover?: string;
    background_color?: string;
    focus_background_color?: string;
    focus_background_color_hover?: string;
    focus_hover?: string;
    focus?: string;
    focus_text_color?: string;
    focus_text_color_hover?: string;
    form_text_color?: string;
    form_text_color_hover?: string;
    hover?: string;
    important?: string[];
    main?: string;
    margin?: string;
    margin_padding?: {
      css?: {
        main?: string;
        important?: string[];
      }
    };
    padding?: string;
    placeholder_focus?: string;
    placeholder?: string;
    text_shadow?: string;
  }
  box_shadow?: boolean | ShortcodeModuleAdvancedFieldDefinitionBoxShadow;
  border_styles?: boolean | Record<string, ShortcodeModuleAdvancedFieldDefinitionBorder>;
  focus_background_color?: boolean;
  focus_text_color?: boolean;
  font_field?: boolean | ShortcodeModuleAdvancedFieldDefinitionFont;
  margin_padding?: boolean | ShortcodeModuleAdvancedFieldDefinitionSpacing;
  placeholder?: boolean;
  toggle_slug?: string;
  tab_slug?: string;
  text_color?: boolean | Record<string, string>;
  width?: [];
  toggle_priority?: number;

  // Looks like a bug; used in et_pb_wc_additional_info module.
  table_row?: {
    label?: string;
    border_styles?: {
      table_row?: ShortcodeModuleAdvancedFieldDefinitionBorder;
    };
    css?: {
      main?: string;
    };
    font_field?: boolean;
    margin_padding?: {
      css?: {
        main?: string;
      };
      use_margin?: boolean;
    };
    text_color?: boolean;
    focus_background_color?: false;
    focus_text_color?: false;
  }
}
