// @divi/types/src/module/library/sidebar/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface SidebarHTML {
  html?: string;
}

export interface SidebarAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' | 'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
       'transition' |
       'zIndex'
      >;
  };

  // Title
  title?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  sidebar?: {
    innerContent?: FormatBreakpointStateAttr<{
      area?: string;
    }>;
    advanced?: {
      layout?: FormatBreakpointStateAttr<{
        alignment?: string;
        showBorder?: OnOff;
      }>;
    }
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };
}
