// @divi/types/src/module/element/decoration/font/index.ts
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';
import { type Border } from '../border';
import { TextShadow as TextShadowNamespace } from '../text-shadow';


export namespace Font {
  export import TextShadow = TextShadowNamespace;

  export type StyleValue = 'italic' | 'uppercase' | 'capitalize' | 'underline' | 'strikethrough';

  export interface AttributeValue {
    headingLevel?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
    family?: string;
    weight?: '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900';
    style?: StyleValue[];
    lineColor?: string;
    lineStyle?: '' | 'solid' | 'double' | 'dotted' | 'dashed' | 'wavy';
    textAlign?: 'left' | 'center' | 'right' | 'justify';
    color?: string;
    size?: string;
    letterSpacing?: string;
    lineHeight?: string;
  }

  export type ListStyleUnordered = 'disc' | 'circle' | 'square' | 'none';

  export type ListStyleOrdered = 'decimal' |
    'armenian' |
    'cjk-ideographic' |
    'decimal-leading-zero' |
    'georgian' |
    'hebrew' |
    'hiragana' |
    'hiragana-iroha' |
    'katakana' |
    'katakana-iroha' |
    'lower-alpha' |
    'lower-greek' |
    'lower-latin' |
    'lower-roman' |
    'upper-alpha' |
    'upper-greek' |
    'upper-latin' |
    'upper-roman' |
    'none';

    export interface ListAttributeValue {
    type?: ListStyleUnordered | ListStyleOrdered;
    position?: 'outside' | 'inside';
    itemIndent?: string;
  }

  export interface BorderAttributeValue {
    styles?: {
      left?: {
        width?: string;
        color?: string;
      }
    }
  }


  export interface Attributes {
    font?: FormatBreakpointStateAttr<AttributeValue>;
    textShadow?: TextShadowNamespace.Attributes;
    list?: FormatBreakpointStateAttr<ListAttributeValue>;
    border?: Border.Attributes;
  }

  export namespace Style {

    // TODO Replace `FontStyleProps` with this
    // <FontStyle /> has additional props; Extending `StyleProps` on styles/font/types will cause circular
    export interface Props extends Components.Style.Props<Attributes> {
      headingLevel?: string | boolean;
      propertySelectors?: {
        font?: FormatPropertySelectors<Css.Properties.Font>;
        textShadow?: TextShadowNamespace.Style.Props['propertySelectors'];
      };
      important?: boolean | {
        font?: FormatStyleImportant<Css.Properties.Font>;
        textShadow?: TextShadowNamespace.Style.Props['important'];
      }
    }
  }
}
