// @divi/types/src/shortcode-module/index.ts
// TODO feat(D5) There might be more property out there.
/* eslint-disable @typescript-eslint/naming-convention */
import { Field as FieldNamespace } from './field';
import { GroupField as GroupFieldNamespace } from './group-field';
import { Help as HelpNamespace } from './help';
import { Module as ModuleNamespace } from './module';
import { Panel as PanelNamespace } from './panel';
import { Store as StoreNamespace } from './store';


export * from './advanced-field';
export * from './d4';
export * from './field';

export namespace ShortcodeModule {
  /* eslint-disable @typescript-eslint/no-unused-vars */
  // TODO figure out a way to declare nested namespace without disabling no-unused-vars rule
  // imported namespace as marked as unused by eslint using this config; it shouldn't happen in the first place.
  export import Field = FieldNamespace;
  export import GroupField = GroupFieldNamespace;
  export import Help = HelpNamespace;
  export import Panel = PanelNamespace;
  export import Store = StoreNamespace;
  export import Module = ModuleNamespace;
  /* eslint-enable @typescript-eslint/no-unused-vars */

  export interface Attributes extends Record<string, unknown> {
    builderVersion?: string;
    modulePreset?: string;
    nonconvertible?: string;
    innerHTML?: string;
    registeredShortcode?: string;
    shortcodeName?: string;
    originalName?: string;
  }

  export interface RenderedContent {
    html?: string;
    error?: string;
  }
}
