// @divi/types/src/right-click-options/owner/index.ts
import { type ReactNode } from 'react';
import { type ImmutableObject } from 'seamless-immutable';

import { type Clipboard } from '../../clipboard';
import { type ModuleType } from '../../module/category';


export namespace Owner {
  export interface ModalGroup {
    optionType: 'modalGroup';

    // @todo Need to circle back on what this id is meant and whether it is really needed or not.
    id?: string;

    // @todo Need to circle back whether moduleId is really needed or not.
    moduleId?: string;
    modalName: string;
    modalTab: string;
    modalGroup?: string;
    modalGroupLabel: string;
    clipboardItem?: ImmutableObject<Clipboard.Item.ModuleFieldAttribute>
    | ImmutableObject<Clipboard.Item.ModuleGroupTabAttribute>
    | ImmutableObject<Clipboard.Item.ModuleGroupAttribute>;
    groupName?: string;
    fieldsNode?: ReactNode;
    groupId?: string;
    attrsMaps?: Clipboard.Attrs.Maps;
  }

  export interface ModalTab {
    optionType: 'modalTab';
    id?: string;
    modalName: string;
    modalTab: string;
    fieldsNode?: ReactNode;
  }

  export interface ModalHeader {
    optionType: 'modalHeader';
    id: string;
    moduleName: string;
    fieldsNode: ReactNode;
  }

  export interface ModalField {
    optionType: 'modalField';
    id: string;
    modalName: string;
    fieldLabel: string;
    attrName: string;
    subName?: string;
    fieldName?: string;
    fieldOptions?: Record<string, any>;
    fieldsNode?: ReactNode;
    clipboardItem?: ImmutableObject<Clipboard.Item.ModuleFieldAttribute>;
    groupName?: string;
    attrsMaps?: Clipboard.Attrs.Maps;
  }

  export interface ModalGroupTab {
    optionType: 'modalGroupTab';
    modalName: string;
    moduleId: string;
    groupId?: string;
    groupTabName?: string;
    groupTabLabel: string;
    fieldsNode?: ReactNode;
    id?: string;
    attrsMaps?: Clipboard.Attrs.Maps;
    clipboardItem?: ImmutableObject<Clipboard.Item.ModuleGroupTabAttribute>
    | ImmutableObject<Clipboard.Item.ModuleGroupAttribute>
    | ImmutableObject<Clipboard.Item.ModuleFieldAttribute>;
  }

  export interface Module {
    optionType: 'module';
    id: string;
    moduleType?: ModuleType;
    modalName?: string;
    clipboardItem?: ImmutableObject<Clipboard.Item.Module> | ImmutableObject<Clipboard.Item.Modules>;
    clipboardItemStyle?: ImmutableObject<Clipboard.Item.ModuleAttributes>
    | ImmutableObject<Clipboard.Item.ModuleGroupAttribute>
    | ImmutableObject<Clipboard.Item.ModuleFieldAttribute>;
    selectedModuleIds?: string[];
    fieldsNode?: ReactNode;
    attrsMaps?: Clipboard.Attrs.Maps;
  }

  export interface DraggableChildModuleField {
    optionType: 'modalDraggableChildModuleList';
    id: string;
    moduleName: string;
    clipboardItem?: ImmutableObject<Clipboard.Item.Module> | ImmutableObject<Clipboard.Item.Modules>;
    clipboardItemStyle?: ImmutableObject<Clipboard.Item.ModuleAttributes>
    | ImmutableObject<Clipboard.Item.ModuleGroupAttribute>
    | ImmutableObject<Clipboard.Item.ModuleFieldAttribute>;
    fieldsNode?: ReactNode;
    attrsMaps?: Clipboard.Attrs.Maps;
  }

  export interface ColorPicker {
    optionType: 'colorPicker';
    colorPickerAttrs: {
      color: string | null;
      colorId: number | string | null;
      colorPalette: string | null;
      isCurrentColorSwatch: boolean,
    }
    detachCurrentColorFromGlobal: (colorId: string, color: string) => void,
    convertToGlobalColor: (color: string) => void,
    onEditGlobalColor: (colorId: string, color: string) => void,
    onDeleteGlobalColor: (colorId: string) => void,
    replaceGlobalColor: (colorId: string, color: string) => void,
  }

  export interface GradientStop {
    optionType: 'gradientStop';
    fieldsNode?: ReactNode;
    removeStop: () => void;
  }

  export interface SidebarRowNav {
    optionType: 'sidebarRowNav';
    modalName: string;
    id: string;
    trigger: 'navItem' | 'navWrapper';
  }

  export type Item = null
    | ModalGroup
    | ModalTab
    | ModalHeader
    | ModalField
    | ModalGroupTab
    | Module
    | ColorPicker
    | GradientStop
    | DraggableChildModuleField
    | SidebarRowNav;
}
