// @divi/types/src/module/element/decoration/background/index.ts
import { type OnOff } from '../../../../attr';
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';


export namespace Background {

  export interface TabOption {
    icon: string;
    label: string;
  }

  export interface TabOptions {
    [key: string]: TabOption;
  }

  export type GradientStop = {
    position: string;
    color: string;
  };

  export interface Color {
    color?: string;
    enableColor?: OnOff; // Some modules have support for enabling and disabling background color: https://elegantthemes.slack.com/archives/C01CW343ZJ9/p1651717784463199
  }

  export interface Gradient {
    enabled?: OnOff;
    start?: string;
    end?: string;
    type?: 'linear' | 'circular' | 'elliptical' | 'conic';
    direction?: string;
    directionRadial?: 'center' | 'top left' | 'top' | 'top right' | 'right' | 'bottom right' | 'bottom' | 'bottom left' | 'left';
    startPosition?: string;
    endPosition?: string;
    overlaysImage?: OnOff;
    repeat?: OnOff;
    length?: string;
    stops?: GradientStop[];
  }

  export interface Image {
    enabled?: OnOff; // Some modules have support for enabling and disabling background image: https://elegantthemes.slack.com/archives/C01CW343ZJ9/p1651717784463199
    url?: string;
    parallax?: {
      enabled?: OnOff;
      method?: OnOff;
    };
    title?: string;
    size?: 'cover' | 'contain' | 'initial' | 'stretch' | 'custom';
    width?: string;
    height?: string;
    position?: 'left top' | 'center top' | 'right top' | 'left center' | 'center' | 'right center' | 'left bottom' | 'center bottom' | 'right bottom';
    horizontalOffset?: string;
    verticalOffset?: string;
    repeat?: 'no-repeat' | 'repeat' | 'repeat-x' | 'repeat-y' | 'space' | 'round';
    blend?: 'normal' | 'multiply' | 'screen' | 'overlay' | 'darken' | 'lighten' | 'color-dodge' | 'color-burn' | 'hard-light' | 'soft-light' | 'difference' | 'exclusion' | 'hue' | 'saturation' | 'color' | 'luminosity';
  }

  export interface Video {
    enabledMp4?: OnOff;
    enabledWebm?: OnOff;
    mp4?: string;
    webm?: string;
    width?: string;
    height?: string;
    allowPlayerPause?: OnOff;
    pauseOutsideViewport?: OnOff;
  }

  export interface Pattern {
    enabled?: OnOff;
    style?: string;
    color?: string;
    transform?: string[];
    size?: 'initial' | 'cover' | 'contain' | 'stretch' | 'custom';
    width?: string;
    height?: string;
    repeatOrigin?: 'left top' | 'center top' | 'right top' | 'left center' | 'center' | 'right center' | 'left bottom' | 'center bottom' | 'right bottom';
    horizontalOffset?: string;
    verticalOffset?: string;
    repeat?: 'repeat' | 'repeat-x' | 'repeat-y' | 'space' | 'round';
    blend?: 'normal' | 'multiply' | 'screen' | 'overlay' | 'darken' | 'lighten' | 'color-dodge' | 'color-burn' | 'hard-light' | 'soft-light' | 'difference' | 'exclusion' | 'hue' | 'saturation' | 'color' | 'luminosity';
  }

  export interface Mask {
    enabled?: OnOff;
    style?: string;
    color?: string;
    transform?: string[];
    aspectRatio?: 'landscape' | 'square' | 'portrait';
    size?:  'stretch' | 'cover' | 'contain' | 'custom';
    width?: string;
    height?: string;
    position?: 'left top' | 'center top' | 'right top' | 'left center' | 'center' | 'right center' | 'left bottom' | 'center bottom' | 'right bottom';
    horizontalOffset?: string;
    verticalOffset?: string;
    blend?: 'normal' | 'multiply' | 'screen' | 'overlay' | 'darken' | 'lighten' | 'color-dodge' | 'color-burn' | 'hard-light' | 'soft-light' | 'difference' | 'exclusion' | 'hue' | 'saturation' | 'color' | 'luminosity';
  }

  export type AttributeValueAny = Gradient | Image | Video | Pattern | Mask;

  export interface AttributeValue extends Color {
    gradient?: Gradient;
    image?: Image;
    video?: Video;
    pattern?: Pattern;
    mask?: Mask;
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Declaration {

    // TODO replace `DeclarationParams` with this.
    export interface Params {
      important?: boolean;
      attrValue: AttributeValue;
    }
  }

  export namespace Style {

    // TODO replace `BackgroundStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.Background>;
      featureSelectors?: {
        mask?: FormatBreakpointStateAttr<string>;
        pattern?: FormatBreakpointStateAttr<string>;
      };
      important?: boolean | FormatStyleImportant<Css.Properties.Background>;
    }
  }

  export namespace Component {

    // TODO replace `BackgroundComponentProps` with this.
    export interface Props {
      attr: Attributes;
      id: string;
      settings?: {
        mask?: boolean;
        parallax?: boolean;
        pattern?: boolean;
        video?: boolean;
      } | boolean;
    }

    // TODO replace `BackgroundComponentContainerProps` with this.
    export interface ContainerProps {
      backgroundAttr: Attributes;
      moduleId: string;
    }

    /* eslint-disable-next-line no-shadow */
    export namespace Video {

      // TODO replace `BackgroundComponentVideoProps` with this.
      /* eslint-disable-next-line no-shadow */
      export interface Props extends ContainerProps {
        className?: string;
        mp4?: string;
        webm?: string;
        width?: string;
        height?: string;
        allowPlayerPause?: OnOff;
        pauseOutsideViewport?: OnOff;
      }
    }

    export namespace Parallax {

      // TODO replace `BackgroundComponentParallaxContainerProps` with this.
      /* eslint-disable-next-line no-shadow */
      export interface ContainerProps {
        backgroundAttr: Attributes;
        moduleId: string;
      }

      // TODO replace `BackgroundComponentParallaxProps` with this.
      /* eslint-disable-next-line no-shadow */
      export interface Props {
        backgroundAttr: Attributes;
        moduleId: string;
      }

      export namespace Item {

        // TODO replace `BackgroundComponentParallaxItemProps` with this.
        /* eslint-disable-next-line no-shadow */
        export interface Props {
          breakpoint?: string;
          cssParallax?: boolean;
          existClassNames?: {
            [key: string]: boolean;
          };
          moduleId: string;
          state: string;
          url: string;
          setTrueParallax: string;
          unsetTrueParallax: string;
        }
      }
    }

    export namespace MaskPattern {

      // TODO replace `BackgroundComponentMaskPatternContainerProps` with this.
      /* eslint-disable-next-line no-shadow */
      export interface ContainerProps {
        backgroundAttr: Attributes;
        moduleId: string;
      }

      // TODO replace `BackgroundComponentMaskPatternProps` with this.
      /* eslint-disable-next-line no-shadow */
      export interface Props extends ContainerProps {
        className?: string;
        style?: string;
        enable?: OnOff;
      }
    }
  }
}
