// @divi/types/src/field-library/common-numeric-input/index.ts
import { type FocusEvent } from 'react';


export namespace NumericInput {

  /**
   * @typedef {object} ValueNumberUnit
   * @property {number} valueNumber Value number.
   * @property {string} valueUnit Value unit.
   */
  export interface ValueNumberUnit {
    valueNumber: number;
    valueUnit: string;
  }

  /**
   * @typedef {object} ValueMapping
   * @property {string} original Original.
   * @property {string} alias Alias.
   */
  export interface ValueMapping {
    original: string;
    alias: string;
  }

  /**
   * @type {(value: string) => string} Props
   */
  export type Formatter = (value: string) => string;

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string[]} [allowedUnits] Allowed units.
   * @property {string} [className] Class name.
   * @property {string} [defaultUnit] Default unit.
   * @property {string} [defaultValue] Default value.
   * @property {number} [max] Max value.
   * @property {number} [min] Min value.
   * @property {number} [maxResetStepValue] Max reset step value.
   * @property {number} [minResetStepValue] Min reset step value.
   * @property {boolean} ["readonly"] Read only.
   * @property {number} [step] Step.
   * @property {string} value Value.
   * @property {object[]} [valueMapping] Value mapping.
   * @property {string[]} [allowedValues] Allowed values.
   * @property {Function} [formatter] Formatter.
   * @property {Function} [onChange] On change.
   * @property {Function} [onFocus] On focus.
   * @property {Function} [onBlur] On blur.
   */
  export interface Props {
    name: string;
    allowedUnits?: string[];
    className?: string;
    defaultUnit?: string;
    defaultValue?: string;
    max?:number | null;
    min?:number | null;
    maxResetStepValue?: number;
    minResetStepValue?: number;
    readonly?: boolean;
    step?:number;
    value: string;
    valueMapping?: ValueMapping[];
    allowedValues?: string[];
    formatter?: Formatter;
    onChange?: (valueNew: string, updateSource: string) => void;
    onFocus?: (event: FocusEvent<HTMLInputElement>) => void;
    onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
  }
}
