// @divi/types/src/field-library/select-widget-area/index.ts
import { type FormEvent } from 'react';

import { type KeyboardShortcuts } from '../../keyboard-shortcuts';
import { type Select } from '../select';


export namespace SelectWidgetArea {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {object} [event] Form Event.
   * @property {string} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: string;
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} [defaultValue] Default value.
   * @property {string} value Value.
   * @property {object} options Options.
   * @property {Function} onChange On change callback.
   * @property {Function} [registerShortcuts] Register shortcuts.
   * @property {Function} [unregisterShortcuts] Unregister shortcuts.
   */
  export interface Props {
    name: string;
    defaultValue?: string;
    value: string
    options: Select.Options;
    onChange: (params: OnChangeCallbackParams) => void;
    registerShortcuts?: (shortcuts: KeyboardShortcuts.Shortcut.Item[]) => void,
    unregisterShortcuts?: (names: string[]) => void;
  }
}
