// @divi/types/src/field-library/gradient/index.ts
import { type MouseEvent } from 'react';

import {
  type AttrState,
  type Breakpoint,
} from '../../format-attr';
import { type Background } from '../../module/element/decoration/background';
import { type RightClickOptions } from '../../right-click-options';


export namespace Gradient {

  /**
   * @typedef {object} StopsOnChangeCallbackParams
   * @property {Event} [event] Event.
   * @property {object} [inputValue] Input value.
   */
  export interface StopsOnChangeCallbackParams {
    event?: CustomEvent | MouseEvent;
    inputValue?: Background.AttributeValue['gradient']['stops'];
  }

  /**
   * @typedef {object} Props
   * @property {object} backgroundAttrs Background attributes.
   * @property {object} defaultValue Default value.
   * @property {string} length Length.
   * @property {object} value Value.
   * @property {Function} [onChange] On change callback.
   * @property {boolean} ["readonly"] Read only.
   * @property {Function} [setRightClickOptions] Set right click options callback.
   */
  export type Props = {
    backgroundAttrs: Background.AttributeValue;
    defaultValue: Background.AttributeValue['gradient']['stops'] | null;
    length: string;
    value: Background.AttributeValue['gradient']['stops'] | null;
    onChange?: (props: StopsOnChangeCallbackParams) => void;
    readonly?: boolean;
    setRightClickOptions?: (params: {
      options: RightClickOptions.Option.Items;
      cursorPosition: RightClickOptions.Cursor.Position;
      optionType: 'gradientStop';
      owner: RightClickOptions.Owner.Item;
      windowLocation: RightClickOptions.Window.Location;
    }) => void,
  }

  /**
   * @typedef {object} ContainerProps
   * @property {object} [attr] Background attributes.
   * @property {string} [appBreakpoint] App breakpoint.
   * @property {string} [appState] App state.
   */
  export type ContainerProps = Partial<Props> & {
    attr?: Background.Attributes;
    appBreakpoint?: Breakpoint,
    appState?: AttrState,
  }
}
