// @divi/types/src/app-ui/sidebar/index.ts
import { Modal as ModalNamespace } from './modal';


export namespace Sidebar {
  /* eslint-disable @typescript-eslint/no-unused-vars */
  // TODO figure out a way to declare nested namespace without disabling no-unused-vars rule
  // imported namespace as marked as unused by eslint using this config; it shouldn't happen in the first place.
  export import Modal = ModalNamespace;
  /* eslint-enable @typescript-eslint/no-unused-vars */

  export type Name = 'sidebarLeft' | 'sidebarRight';

  export type Names = Name[];

  // Sidebar position.
  export type Position = 'left' | 'right';

  export interface Offset {
    top: number;
    left: number;
  }

  export interface Dimension {
    height?: number;
  }

  /**
   * Row data; this will be displayed as panel in sidebar row.
   */
  export interface RowItem {
    id: string;
    modalIds: string[];
    activeModal: string;
    multiModal?: boolean;
    dimension?: Dimension;
    maximum?: Dimension;
    minimum?: Dimension;
    resizing?: boolean;
  }

  /**
   * Collection of row items.
   */
  export type RowItems = Record<string, RowItem>;

  /**
   * Sidebar component props.
   */
  export interface Props {
    rowIds: RowItem['id'][];
  }

  export type DockingArea = false | 'sidebar';
}

export namespace SidebarRow {
  export type DockingArea = false | 'sidebar-row' | 'sidebar-row-last' | 'sidebar-modal';
}
