// @divi/types/src/module/library/signup/index.ts
import { type OnOff } from '../../../attr';
import {
  type FormatBreakpointStateAttr,
} from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import {
  type EmailServiceGroupAttr,
} from '../../options';
import {
  type InternalAttrs,
} from '../internal';


export interface SignupAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.AttributeValue;

  module?: {
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
      layout?: FormatBreakpointStateAttr<string>;
      emailService?: EmailServiceGroupAttr;
      spamProtection?: Element.Advanced.SpamProtection.Attributes;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'animation' |
      'background' |
      'conditions' |
      'disabledOn' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex' |
      'border' |
      'boxShadow' |
      'filters'
    >;
    meta?: Element.Meta.Attributes;
  };

  // Signup Form Title.
  title?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
    innerContent?: FormatBreakpointStateAttr<string>;
  };

  // Signup Form Description.
  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Pick<Element.Decoration.Attributes,
      'bodyFont'
    >;
  };

  description?: {
    innerContent?: FormatBreakpointStateAttr<string>;
  };

  resultMessage?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  formField?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  // Button: Form submit button.
  button?: Element.Types.Button.Attributes;

  // Field: Normal, Focused, Placeholder.
  field?: {
    decoration?: Pick<Element.Decoration.Attributes, 'background' | 'border' | 'boxShadow' | 'font' | 'spacing'>;
    advanced?: {
      focusUseBorder?: FormatBreakpointStateAttr<OnOff>;
      focus?: Pick<Element.Decoration.Attributes, 'background' | 'border' | 'font'>;
      placeholder?: Pick<Element.Decoration.Attributes, 'font'>;
      emailFullwidth?: FormatBreakpointStateAttr<OnOff>;
      firstNameField?: FormatBreakpointStateAttr<OnOff>;
      firstNameFullwidth?: FormatBreakpointStateAttr<OnOff>;
      ipAddress?: FormatBreakpointStateAttr<OnOff>;
      lastNameField?: FormatBreakpointStateAttr<OnOff>;
      lastNameFullwidth?: FormatBreakpointStateAttr<OnOff>;
      nameField?: FormatBreakpointStateAttr<OnOff>;
      nameFieldOnly?: FormatBreakpointStateAttr<OnOff>;
      nameFullwidth?: FormatBreakpointStateAttr<OnOff>;
    }
  }

  // Signup Form Footer Content.
  footerContent?: {
    innerContent?: FormatBreakpointStateAttr<string>;
  };

  // SignUp Form Success Action.
  success?: {
    advanced?: {
      action?: FormatBreakpointStateAttr<string>;
      message?: FormatBreakpointStateAttr<string>;
      redirectQuery?: FormatBreakpointStateAttr<string[]>;
      redirectUrl?: FormatBreakpointStateAttr<string>;
    };
  };

  // Custom Fields.
  customFields?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
      notice?: FormatBreakpointStateAttr<string>;
    };
  }
}
