// @divi/types/src/module/library/post-nav/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface PostNavData {
  title?: string;
  date?: string;
  permalink?: string;
}


export interface PostNavAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      inSameTerm?: FormatBreakpointStateAttr<OnOff>;
      taxonomyName?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'transform' |
      'transition' |
      'zIndex' |
      'sizing' |
      'sticky'
    >;
  };

  // Links
  links?: {
    advanced?: {
      prevText?: FormatBreakpointStateAttr<string>;
      nextText?: FormatBreakpointStateAttr<string>;
      showPrev?: FormatBreakpointStateAttr<OnOff>;
      showNext?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'font' |
      'background' |
      'border' |
      'boxShadow' |
      'spacing'
    >;
  };
}
