// @divi/types/src/module/element/decoration/disabled-on/index.ts
import { type AppPreferences } from '../../../../app-preferences';
import { type OnOff } from '../../../../attr';
import { type Css } from '../../../../css';
import {
  type AttrState,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type Components } from '../../../components';


export namespace DisabledOn {
  export type AttributeValue = OnOff;

  export interface Attributes {
    desktopAbove?: {
      value?: AttributeValue;
    };
    tabletOnly?: {
      value?: AttributeValue;
    };
    phone?: {
      value?: AttributeValue;
    };
  }

  export namespace Style {

    // TODO replace `DisabledModuleStyleProps` with this.
    export interface Props extends Omit<Components.Style.Props<Attributes>, 'selectors'> {
      selectors?: {
        desktopAbove?: Partial<Record<AttrState, string>>;
        tabletOnly?: Partial<Record<AttrState, string>>;
        phone?: Partial<Record<AttrState, string>>;
      };
      propertySelectors?: FormatPropertySelectors<Css.Properties.DisabledOn>;
      disabledModuleVisibility: AppPreferences.DisabledModule.Visibility;
    }

  }
}
