// @divi/types/src/module/metadata/index.ts
/* eslint-disable jsdoc/require-description-complete-sentence -- Needed to write proper code documentation. */
import { type Css } from '../css';
import { type Default } from '../default';
import { type InternalAttrs } from '../library';
import { Element as ElementNamespace } from './element';
import { ElementType as ElementTypeNamespace } from './element-type';
import { Settings as SettingsNamespace } from './settings';


type RemoveInternalAttrsAndCss<TAttrs> = Omit<Omit<TAttrs, 'css'>, keyof InternalAttrs>


export namespace Metadata {
  /* eslint-disable @typescript-eslint/no-unused-vars */
  // TODO figure out a way to declare nested namespace without disabling no-unused-vars rule
  // imported namespace as marked as unused by eslint using this config; it shouldn't happen in the first place.
  export import Settings = SettingsNamespace;
  export import Element = ElementNamespace;
  export import ElementType = ElementTypeNamespace;
  /* eslint-enable @typescript-eslint/no-unused-vars */

  /**
   * Video properties of module.json's videos property.
   */
  export interface HelpVideo {
    id: string;
    name: string;
  }

  /**
   * Module metadata values.
   */
  export interface Values<TAttributes = Default.Attributes> {
    name: string;
    title: string;
    titles: string;
    moduleClassName?: string;
    moduleOrderClassName?: string;
    moduleIcon?: string;
    category: string;
    folder?: string;
    videos?: Record<number, HelpVideo>;
    attributes: {

      /**
       * Module Metadata Attributes and Module Attributes have different structures.
       *
       * Module Attributes (passed generic `TAttributes`):
       *
       * ```
       * title: {
       *   [value]
       * }
       * ```
       *
       * Module Metadata Attributes:
       *
       * ```
       * title: {
       *   default: {
       *     [value]
       *   }.
       * }
       * ```
       *
       * So we use the following type to convert the Module Attributes (`TAttributes`)
       * to Module Metadata Attributes structure.
       */
      [K in keyof RemoveInternalAttrsAndCss<TAttributes>]: Element.Properties<TAttributes[K]>;
    };
    d4Shortcode?: string;
    customCssFields?: Record<string, Css.Field>;
    settings?: {
      content?: 'auto';
      design?: 'auto';
      advanced?: 'auto';

      // Group settings configuration.
      groups?: Record<string, Omit<SettingsNamespace.AttributeGroup.Item.Group, 'groupType'>>;
    };

    /**
     * TODO(D5, Types): Replace `string` with module library name type (eg. `divi/counter` | `divi/accordion-item`)
     *
     * How `childModuleName` differs from `ModuleRegisterDefinition['childrenName]`: The type for
     * ``ModuleRegisterDefinition['childrenName]` is `string[]` which implies that a module can have
     * multiple possible type of child module. This was added assuming in D6 a module can have multiple
     * type of modules inside of it (container module have two different kind of modules, etc). In
     * `Metadata.Values['childModuleName']`, the type is set into `string` because this is used
     * by `getAutoGenerateModuleSettingsPanel()` to automatically added `<DraggableChildModuleListContainer />`
     * into auto generate content panel. In this case, at least right now `<DraggableChildModuleListContainer />`
     * can only have one type of child module thus the value of `childModuleName` is set into `string`.
     */
    childModuleName?: string;
    childModuleTitle?: string;
    childModuleTitleAttribute?: string;

    /**
     * TODO(D5, Types): Temporary using `any` instead of `string[]` because `ModuleDefinitionSettings`
     * is based on Gutenberg's `BlockConfiguration` which specifies `script` as `string` and changing
     * it would have an unnecessary huge ripple effect.
     *
     * Gutenberg newer versions support array of string for `script` however our GB types are not updated
     * which needs to be done on its own task. Existing types folder in `/packages` should also be removed
     * because it is causing outdated errors regarding `script` being string[].
     *
     * Once above is addressed, then this type issue can be changed.
     */
    script?: any;
  }
}


// TODO To be removed.
export type ModuleMetadata<TAttributes = Default.Attributes> = Metadata.Values<TAttributes>;
