// @divi/types/src/module/library/filterable-portfolio/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


interface FilterablePortfolioCategory {
  id: number;
  slug: string;
  label: string;
  permalink: string;
}


export interface FilterablePortfolioPost {
  id?: number;
  title?: string;
  permalink?: string;
  thumbnail?: string;
  categories?: FilterablePortfolioCategory[];
  classNames?: string[];
}
export interface ProjectTermType {
  id?: string;
  label?: string;
  slug?: string;
}
export interface FilterablePortfolioAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Title
  title?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Meta
  meta?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Filter Criteria
  filter?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Pagination
  pagination?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  portfolio?: {
    advanced?: {
      postsNumber?: FormatBreakpointStateAttr<string>;
      includedCategories?: FormatBreakpointStateAttr<string[]>;
      showTitle?: FormatBreakpointStateAttr<OnOff>;
      showCategories?: FormatBreakpointStateAttr<OnOff>;
      showPagination?: FormatBreakpointStateAttr<OnOff>;
      layout?: FormatBreakpointStateAttr<OnOff>;
    }
  };

  portfolioItem?: {
    decoration?: Element.Decoration.PickedAttributes<'border'>;
  };

  image?: {
    decoration?: Element.Decoration.PickedAttributes<
    'border' |
    'boxShadow' |
    'filters'
    >;
  };

  overlay?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'icon'
    >;
  };
}
