// @divi/types/src/module/library/number-counter/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface NumberCounterCssAttr extends Css.AttributeValue {
  percent?: string;
  numberCounterTitle?: string;
}

export type NumberCounterCssGroupAttr = FormatBreakpointStateAttr<NumberCounterCssAttr>;
export interface NumberCounterAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: NumberCounterCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Title
  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Content
  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
  };

  // Numbers.
  number?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    advanced?: {
      enablePercentSign: FormatBreakpointStateAttr<OnOff>;
    }
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };
}
