// @divi/types/src/module/library/social-media-follow-item/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface SocialMediaItemProps {
  label?: string;
  background?: string;
}

export interface SocialNetworkProps {
  [key: string]: SocialMediaItemProps
}

export type IconSizeAttrs = {
  useSize?: string;
  size?: string;
};
export interface SocialMediaFollowItemAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  // Button
  button?: Element.Types.Button.Attributes;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'border' |
      'boxShadow' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  socialNetwork?: {
    innerContent?: FormatBreakpointStateAttr<{
      title?: string;
      link?: string;
      skypeUrl?: string;
      skypeAction?: string;
    }>;
    advanced?: {
      followButton?: FormatBreakpointStateAttr<OnOff>;
    };
  };
  icon?: {
    advanced?: {
      color?: FormatBreakpointStateAttr<string>;
      size?: FormatBreakpointStateAttr<IconSizeAttrs>;
    };
  };
}
