// @divi/types/src/module/library/video-slider-item/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface VideoSlideThumbnailURL {
  thumbnail?: string;
  cover?: string;
}

export interface VideoSliderItemAttrs extends InternalAttrs {

  // Module.
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Video.
  video?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
      webm?: string;
    }>;
    decoration?: Element.Decoration.PickedAttributes<
      'border' |
      'boxShadow'
    >;
  };

  // Video Play Icon.
  playIcon?: {
    decoration?: Element.Decoration.PickedAttributes<'icon'>;
  };

  // Video Overlay.
  overlay?: {
    advanced?: FormatBreakpointStateAttr<{
      showImageOverlay?: OnOff;
    }>
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  };

  // Video Overlay Thumbnail Image.
  thumbnail?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
    }>
  };

  // Video Overlay Cover Image.
  cover?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
    }>
  };

  // Slider Controls.
  sliderControls?: {
    advanced?: FormatBreakpointStateAttr<{
      color?: 'dark' | 'light';
    }>
  };
}
