// @divi/types/src/module/library/tabs/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface TabsCssAttr extends Css.AttributeValue {
  tabsControls?: string;
  tab?: string;
  activeTab?: string;
  tabsContent?: string;
}

export type TabsCssGroupAttr = FormatBreakpointStateAttr<TabsCssAttr>;

export interface TabsAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: TabsCssGroupAttr;

  // Module.
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'border' |
      'boxShadow' |
      'conditions' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Content.
  content?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'bodyFont'
    >;
  };

  // Tab.
  tab?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // InActive Tab.
  inactiveTab?: {
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  };

  // Active Tab.
  activeTab?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };
}
