// @divi/types/src/field-library/select-divider/index.ts
import { type FormEvent } from 'react';

import {
  type AttrState,
  type Breakpoint,
  type FormatBreakpointStateAttr,
} from '../../format-attr';
import { type Dividers } from '../../module/element/advanced/dividers';


export namespace SelectDivider {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {object} [event] Form Event.
   * @property {string} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: string;
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {Function} onChange On change callback.
   * @property {string} value Value.
   * @property {string} [className] Class name.
   * @property {string} [defaultValue] Default value.
   * @property {string[]} [flip] Flip.
   * @property {('top'|'bottom')} [placement] Placement.
   * @property {object} [dividersAttr] Dividers attr.
   */
  export interface Props {
    name: string;
    onChange: (params: OnChangeCallbackParams) => void;
    value: string;
    className?: string;
    defaultValue?: string;
    flip?: string[];
    placement?: 'top' | 'bottom';
    dividersAttr?: Partial<Dividers.AttributeValue>;
  }

  /**
   * @typedef {object} ContainerProps
   * @property {string} [attr] Dividers component attr.
   * @property {string} [appBreakpoint] App breakpoint.
   * @property {string} [appState] App state.
   */
  export type ContainerProps = Partial<Props> & {
    attr?: FormatBreakpointStateAttr<Dividers.AttributeValue>;
    appBreakpoint?: Breakpoint,
    appState?: AttrState,
  }
}
