// @divi/types/src/field-library/text/index.ts
import {
  type FormEvent,
  type KeyboardEvent,
  type ReactNode,
} from 'react';


export namespace Text {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {object} [event] Form Event.
   * @property {string} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: string;
  }

  /**
   * @typedef {Function} OnChangeCallback
   * @param {object} params Callback params.
   */
  export type OnChangeCallback = (params: OnChangeCallbackParams) => void;

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} [className] Class name.
   * @property {Function} onChange On change callback.
   * @property {Function} [onKeyDown] On key down callback.
   * @property {string} value Value.
   * @property {string} [defaultValue] Default value.
   * @property {boolean} [autofocus] Autofocus.
   * @property {boolean} ["readonly"] Read only.
   * @property {boolean} [showPlaceholderOnEmpty] Show placeholder on empty.
   * @property {Function} [renderBefore] Render before.
   * @property {Function} [renderAfter] Render after.
   */
  export interface Props {
    name: string;
    className?: string;
    onChange: OnChangeCallback;
    onKeyDown?: (event: KeyboardEvent<HTMLInputElement>) => void;
    value: string;
    defaultValue?: string;
    autofocus?: boolean;
    readonly?: boolean;
    showPlaceholderOnEmpty?: boolean;
    renderBefore?: (state: string) => ReactNode;
    renderAfter?: (state: string) => ReactNode;
  }
}
