// @divi/types/src/module/settings/group/library/spam-protection/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace SpamProtection {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Advanced.SpamProtection.Attributes;
    fields?: {
      enabled?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      provider?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/spam-protection-provider'];
      };
      account?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/spam-protection-account'];
      };
      minScore?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      useBasicCaptcha?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
    };
  }
}
