// @divi/types/src/module/library/map/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface GrayscaleFilterAttrs {
  enabled?: OnOff;
  amount?: string;
}

export interface MapAttrs extends InternalAttrs {

  map?: {
    innerContent?: Element.InnerContent.Map.Attributes,
    advanced?: {
      grayscaleFilter?: FormatBreakpointStateAttr<GrayscaleFilterAttrs>;
      googleAPIKey?: FormatBreakpointStateAttr<string>;
      mouseWheel?: FormatBreakpointStateAttr<OnOff>;
      mobileDragging?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<'filters'>;
  }

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };
}
