// @divi/types/src/module/settings/group/library/scroll/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Scroll {
  export type Tab = 'verticalMotion' | 'horizontalMotion' | 'fade' | 'scaling' | 'rotating' | 'blur';

  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Scroll.Attributes;
    defaultTab?: Tab;
    fields?: {
      tabNav?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      enable?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Scroll.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      set?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Scroll.Attributes;
        component?: FieldLibrary.ComponentMap['divi/scroll-effect'];
      };
      motionTriggerStart?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Scroll.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      }
    }
  }
}
