// @divi/types/src/global-layouts/store/index.ts
import { type ImmutableArray, type ImmutableObject } from 'seamless-immutable';

import { type ModuleFlatObjects } from '../../module/state';
import { type Layout } from '../layout';


export namespace Store {
  export interface State {
      layouts: Layout.Items,
      fetchLog: Record<string, string[]>,
      fetchError: Record<string, string[]>,
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace AddFetchError {
      export type ReturnValue = {
        type: 'ADD_FETCH_ERROR';

        // Global Layout CPT ID.
        id: string;

        // Global Module UUID.
        owner: string;
      }
    }

    export namespace AddFetchLog {
      export type ReturnValue = {
        type: 'ADD_FETCH_LOG';

        // Global Layout CPT ID.
        id: string;

        // Global Module UUID.
        owner: string;
      }
    }

    export namespace DeleteFetchError {
      export type ReturnValue = {
        type: 'DELETE_FETCH_ERROR';

        // Global Layout CPT ID.
        id: string;
      }
    }

    export namespace FetchData {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'FETCH_DATA';

        // Global Layout CPT ID.
        id: string;

        // Global Module UUID.
        owner: string;
      }
    }

    export namespace DeleteFetchLog {
      export type ReturnValue = {
        type: 'DELETE_FETCH_LOG';

        // Global Layout CPT ID.
        id: string;
      }
    }

    export namespace SetGlobalLayoutMap {
      export type ReturnValue = {
        type: 'SET_GLOBAL_LAYOUT_MAP';
        id: string;
        map: Layout.ImmutableMap;
      }
    }

    export namespace SetLayoutContent {
      export type ReturnValue = {
        type: 'SET_LAYOUT_CONTENT';
        id: string;
        content: ModuleFlatObjects;
        owner: string;
        syncToServer?: boolean;
      }
    }

    export type FunctionMap = {
      addFetchError: (
        id: string,
        owner: string,
      ) => AddFetchError.ReturnValue,
      addFetchLog: (
        id: string,
        owner: string,
      ) => AddFetchLog.ReturnValue,
      deleteFetchError: (
        id: string,
      ) => DeleteFetchError.ReturnValue,
      deleteFetchLog: (
        id: string,
      ) => DeleteFetchLog.ReturnValue,
      fetchData: (
        params: FetchData.Params,
      ) => FetchData.ReturnValue,
      setGlobalLayoutMap: (
        id: string,
        map: Layout.ImmutableMap,
      ) => SetGlobalLayoutMap.ReturnValue,
      setLayoutContent: (
        id: string,
        content: ModuleFlatObjects,
        owner: string,
        syncToServer: boolean,
      ) => SetLayoutContent.ReturnValue,
    };

    export type ReturnValue =
      AddFetchError.ReturnValue |
      AddFetchLog.ReturnValue |
      DeleteFetchError.ReturnValue |
      DeleteFetchLog.ReturnValue |
      FetchData.ReturnValue |
      SetGlobalLayoutMap.ReturnValue |
      SetLayoutContent.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getFetchError: (id: string) => ImmutableArray<string>;
      getFetchLog: (id: string) => ImmutableArray<string>;
      getGlobalLayoutMap: (id: string) => Layout.ImmutableMap;
      getLayout: (id: string) => ImmutableObject<Layout.Item>;
      getLayoutContent: (id: string) => ModuleFlatObjects;
      getLayoutRootUUID: (id: string) => ImmutableArray<string>;
    }
  }
}
