// @divi/types/src/shortcode-module/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Settings } from '../../settings';
import { type Module } from '../module';


export namespace Store {
  export interface State {
    shortcodeModules: Settings.Store.ImmutableState['shortcodeModuleDefinitions'],
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace SetSomething {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_SOMETHING',
        shortcodeName: string;
        setting: unknown;
      }
    }

    export type FunctionMap = {
      setSomething: (params: SetSomething.Params) => SetSomething.ReturnValue;
    };

    export type ReturnValue = SetSomething.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getShortcodeModule: (shortcodeName: string) => ImmutableObject<Module.Definition>;
    }
  }
}
