// @divi/types/src/field-library/select-pattern/index.ts
import { type MouseEvent } from 'react';

import {
  type AttrState,
  type Breakpoint,
} from '../../format-attr';
import { type Background } from '../../module/element/decoration/background';
import { type Select } from '../select';


export namespace SelectPattern {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {object} [event] Event.
   * @property {string} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: CustomEvent | MouseEvent;
    inputValue?: string;
  }

  /**
   * @typedef {object} Props
   * @property {object} [backgroundAttr] Background attribute.
   * @property {string} [name] Name.
   * @property {Function} [onChange] On change callback.
   * @property {Function} [onReset] On reset callback.
   * @property {string} [value] Value.
   * @property {string} [addTitle] Add title.
   * @property {string} [className] Class name.
   * @property {string} [defaultValue] Default value.
   * @property {boolean} [hasManager] Has manager.
   * @property {boolean} [hasPreview] Has preview.
   * @property {object} [options] Options.
   */
  export interface Props {
    backgroundAttr: Partial<Background.AttributeValue>;
    name: string;
    onChange: (params: OnChangeCallbackParams) => void;
    onReset: (event: MouseEvent) => void;
    value: string;

    // Optional.
    addTitle?: string;
    className?: string;
    defaultValue?: string;
    hasManager?: boolean;
    hasPreview?: boolean;
    options?: Select.Options,
  }

  /**
   * @typedef {object} PatternContainerProps
   * @property {object} [attr] Background attribute.
   * @property {string} [appBreakpoint] App breakpoint.
   * @property {string} [appState] App state.
   */
  export type ContainerProps = Partial<Props> & {
    attr?: Background.Attributes;
    appBreakpoint?: Breakpoint,
    appState?: AttrState,
  }
}
