// @divi/types/src/right-click-options/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Cursor } from '../cursor';
import { type Option } from '../option';
import { type Owner } from '../owner';
import { type Window } from '../window';


export namespace Store {
  export interface State {
    active: boolean;
    cursorPosition: Cursor.Position;
    optionType: Owner.Item['optionType'];
    options: Option.Items;
    parent: number[];
    selected: number;
    length: number;
    owner: Owner.Item;
    windowLocation: Window.Location;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace SelectChild {

      /**
       * @typedef {object} ReturnValue
       * @property {string} type Action type.
       */
      export type ReturnValue = {
        type: 'SELECT_CHILD';
      }
    }

    export namespace SelectNext {

      /**
       * @typedef {object} ReturnValue
       * @property {string} type Action type.
       */
      export type ReturnValue = {
        type: 'SELECT_NEXT';
      }
    }

    export namespace SelectParent {

      /**
       * @typedef {object} ReturnValue
       * @property {string} type Action type.
       */
      export type ReturnValue = {
        type: 'SELECT_PARENT';
      }
    }

    export namespace SelectPrev {

      /**
       * @typedef {object} ReturnValue
       * @property {string} type Action type.
       */
      export type ReturnValue = {
        type: 'SELECT_PREV';
      }
    }

    export namespace Set {

      /**
       * @typedef {object} Params
       * @property {object} cursorPosition Cursor position.
       * @property {string} optionType Option type.
       * @property {object} owner Owner.
       * @property {string} windowLocation Window location.
       */
      export type Params = Omit<ReturnValue, 'type' | 'options'>;

      /**
       * @typedef {object} ReturnValue
       * @property {string} type Action type.
       * @property {object} options Options.
       * @property {object} cursorPosition Cursor position.
       * @property {string} optionType Option type.
       * @property {object} owner Owner.
       * @property {string} windowLocation Window location.
       */
      export type ReturnValue = {
        type: 'SET';
        options: Option.Items;
        cursorPosition: Cursor.Position;
        optionType: Owner.Item['optionType'];
        owner: Owner.Item;
        windowLocation: Window.Location;
      }
    }

    export namespace SetOwnerData {

      /**
       * @typedef {object} Params
       * @property {string} name Owner data name.
       * @property {boolean | string | number} value Owner data value.
       */
      export type Params = Omit<ReturnValue, 'type'>;

      /**
       * @typedef {object} ReturnValue
       * @property {string} type Action type.
       * @property {string} name Owner data name.
       * @property {boolean | string | number} value Owner data value.
       */
      export type ReturnValue = {
        type: 'SET_OWNER_DATA';
        name: string;
        value: boolean | string | number;
      }
    }

    export namespace SetSelected {

      /**
       * @typedef {object} Params
       * @property {number} selected Selected.
       * @property {number[]} parent Parent.
       */
      export type Params = Omit<ReturnValue, 'type'>;

      /**
       * @typedef {object} ReturnValue
       * @property {string} type Action type.
       * @property {number} selected Selected.
       * @property {number[]} parent Parent.
       */
      export type ReturnValue = {
        type: 'SET_SELECTED';
        selected: number;
        parent: number[];
      }
    }

    export namespace Unset {

      /**
       * @typedef {object} ReturnValue
       * @property {string} type Action type.
       */
      export type ReturnValue = {
        type: 'UNSET';
      }
    }

    export type FunctionMap = {
      selectChild: () => SelectChild.ReturnValue;
      selectNext: () => SelectNext.ReturnValue;
      selectParent: () => SelectParent.ReturnValue;
      selectPrev: () => SelectPrev.ReturnValue;
      set: (params: Set.Params) => Set.ReturnValue;
      setOwnerData: (params: SetOwnerData.Params) => SetOwnerData.ReturnValue;
      setSelected: (params: SetSelected.Params) => SetSelected.ReturnValue;
      unset: () => Unset.ReturnValue;
    };

    export type ReturnValue =
      SelectChild.ReturnValue |
      SelectNext.ReturnValue |
      SelectParent.ReturnValue |
      SelectPrev.ReturnValue |
      Set.ReturnValue |
      SetOwnerData.ReturnValue |
      SetSelected.ReturnValue |
      Unset.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getState: () => ImmutableState;
      isActive: () => boolean;
    }
  }
}
