// @divi/types/src/module/library/slide/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface SlideCssAttr extends Css.AttributeValue {
  slideTitle?: string;
  slideContainer?: string;
  slideDescription?: string;
  slideButton?: string;
  slideImage?: string;
}

export type SlideCssGroupAttr = FormatBreakpointStateAttr<SlideCssAttr>;

export interface SlideAttrs extends InternalAttrs {

  // Button
  button?: {
    advanced?: {
      showOnMobile?: FormatBreakpointStateAttr<OnOff>; // No options available in setting panel in Slide module. But this value will be inherited from Slider module.
    };
    innerContent?: Element.Types.Button.InnerContent.Attributes;
    decoration?: Element.Types.Button.Decoration.Attributes;
  };


  css?: SlideCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'filters' |
      'transform' |
      'overflow' |
      'zIndex' |
      'transition'
    >;
  };

  image?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
      alt?: string;
    }>;
    advanced?: {
      alignment?: FormatBreakpointStateAttr<string>;
    }
    decoration?: Element.Decoration.PickedAttributes<
      'border' |
      'boxShadow' |
      'filters'
    >;
  }

  title?: {
    innerContent?: FormatBreakpointStateAttr<string>; // heading
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  }

  content?: {
    advanced?: {
      showOnMobile?: FormatBreakpointStateAttr<OnOff>; // No options available in setting panel in Slide module. But this value will be inherited from Slider module.
    };
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<
      'bodyFont' |
      'inlineFont' |
      'spacing' |
      'sizing'
    >;
  }

  video?: {
    innerContent?: FormatBreakpointStateAttr<string>;
  }

  contentOverlay?: {
    advanced?: {
      use?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<'background' | 'border'>;
  }

  slideOverlay?: {
    advanced?: {
      use?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  }

  arrows?: {
    advanced?: {
      color?: FormatBreakpointStateAttr<string>;
    };
  };

  dotNav?: {
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  }
}
