// @divi/types/src/module/library/search/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import {
  type FieldElementAttr,
} from '../../options';
import { type InternalAttrs } from '../internal';

// Search Custom CSS Attr.
export interface SearchCssAttr extends Css.AttributeValue {
  inputField?: string;
  button?: string;
}

export type SearchCssGroupAttr = FormatBreakpointStateAttr<SearchCssAttr>;

export interface SearchAttrs extends InternalAttrs {

  // Advanced tab.
  css?: SearchCssGroupAttr;

  // Module.
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  searchPlaceholder?: {
    innerContent?: FormatBreakpointStateAttr<string>;
  };

  search?: {
    advanced?: {
      showButton?: FormatBreakpointStateAttr<OnOff>;
      excludePages?: FormatBreakpointStateAttr<OnOff>;
      excludePosts?: FormatBreakpointStateAttr<OnOff>;
      excludedCategories?: FormatBreakpointStateAttr<string[]>;
    };
  };

  // Field: Normal, Focused, Placeholder.
  field?: FieldElementAttr;

  // Button.
  button?: {
    decoration?: Element.Decoration.PickedAttributes<
    'background' |
    'font'
    >;
  };
}
