// @divi/types/src/module/element/decoration/border/index.ts
import { type OnOff } from '../../../../attr';
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';


export namespace Border {
  export type Side = 'all' | 'top' | 'right' | 'bottom' | 'left';

  export interface SideValues {
    width?: string;
    color?: string;
    style?: 'solid' | 'dashed' | 'dotted' | 'double' | 'groove' | 'ridge' | 'inset' | 'outset' | 'none';
  }

  export interface AttributeValue {
    radius?: {
      topLeft?: string;
      topRight?: string;
      bottomLeft?: string;
      bottomRight?: string;
      sync?: OnOff;
    };
    styles?: Partial<Record<Side, SideValues>>;
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Style {

    // TODO replace `BorderStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.Border>;
      important?: boolean | FormatStyleImportant<Css.Properties.Border>;
    }
  }
}
