// @divi/types/src/module/settings/group/library/border/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Border {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Border.Attributes;
    fields?: {
      radius?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Border.Attributes;
        component?: FieldLibrary.ComponentMap['divi/border-radius'];
      };
      stylesTabNav?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      styles?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Border.Attributes;
        component?: FieldLibrary.ComponentMap['divi/border-styles-preview'];
      };
      stylesTabbedWidth?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Border.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      stylesTabbedColor?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Border.Attributes;
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      stylesTabbedStyle?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Border.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      }
    }
  }
}
