// @divi/types/src/shortcode-module/module/index.ts
import { type Field } from '../field';
import { type GroupField } from '../group-field';
import { type Help } from '../help';
import { type Panel } from '../panel';


export namespace Module {
  export type Toggles = Record<string, string | GroupField.Setting>;

  export interface Fields {
    module?: Field.Items;
    advanced?: boolean | Field.AdvancedItems;
    css?: [] | Field.CssItems;
  }

  export interface Definition {
    name: string;
    plural: string;
    slug: string;
    selector: string;
    helpVideos?: Help.Video[];
    folder?: null | string;
    icon?: {
      name?: null | string;
      path?: null | string;
    };
    fields?: Fields;
    is: {
      child: boolean;
      parent: boolean;
      official: boolean;
      vbSupport: boolean;
      fullwidth: boolean;
      uniqueId: boolean;
      rawContent: boolean;
    };
    child?: {
      slug?: null | string;
      itemText?: null | string;
      titleText?: null | string;
      titleVar?: null | string;
      titleFallbackVar?: null | string;
    };
    settings?: {
      group?: Partial<Record<Panel.Name, {
        toggles?: Toggles;
      }>>;
    };
    render?: {
      wrapper?: [] | Record<string, boolean>;
    }
  }
}
