// @divi/types/src/module/state/index.ts
import {
  type ModuleAttrs,
  type ModuleAttrsByName,
} from '../library';


export interface ModuleFlatObjectNamed<TName extends string> {
  children: string[];
  id: string;
  name: TName;
  parent?: string;
  isGlobal?: boolean;
  props: {
    attrs?: ModuleAttrsByName<TName>;
    innerHTML?: string;
  },
}

/**
 * Module's flat object interface which its state is managed at divi/edit-post store.
 *
 * @since ??
 */
export interface ModuleFlatObject<TAttrs = ModuleAttrs> {
  children: string[];
  id: string;
  name: string;
  parent?: string;
  props: {
    attrs?: TAttrs;
    innerHTML?: string;
  },
}

export type ModuleFlatObjectItems = (
  ModuleFlatObjectNamed<'divi/accordion'> |
  ModuleFlatObjectNamed<'divi/accordion-item'> |
  ModuleFlatObjectNamed<'divi/audio'> |
  ModuleFlatObjectNamed<'divi/blog'> |
  ModuleFlatObjectNamed<'divi/blurb'> |
  ModuleFlatObjectNamed<'divi/button'> |
  ModuleFlatObjectNamed<'divi/circle-counter'> |
  ModuleFlatObjectNamed<'divi/code'> |
  ModuleFlatObjectNamed<'divi/column'> |
  ModuleFlatObjectNamed<'divi/column-inner'> |
  ModuleFlatObjectNamed<'divi/comments'> |
  ModuleFlatObjectNamed<'divi/contact-field'> |
  ModuleFlatObjectNamed<'divi/contact-form'> |
  ModuleFlatObjectNamed<'divi/countdown-timer'> |
  ModuleFlatObjectNamed<'divi/counter'> |
  ModuleFlatObjectNamed<'divi/counters'> |
  ModuleFlatObjectNamed<'divi/cta'> |
  ModuleFlatObjectNamed<'divi/divider'> |
  ModuleFlatObjectNamed<'divi/filterable-portfolio'> |
  ModuleFlatObjectNamed<'divi/fullwidth-code'> |
  ModuleFlatObjectNamed<'divi/fullwidth-header'> |
  ModuleFlatObjectNamed<'divi/fullwidth-image'> |
  ModuleFlatObjectNamed<'divi/fullwidth-map'> |
  ModuleFlatObjectNamed<'divi/fullwidth-menu'> |
  ModuleFlatObjectNamed<'divi/fullwidth-portfolio'> |
  ModuleFlatObjectNamed<'divi/fullwidth-post-title'> |
  ModuleFlatObjectNamed<'divi/fullwidth-slider'> |
  ModuleFlatObjectNamed<'divi/gallery'> |
  ModuleFlatObjectNamed<'divi/heading'> |
  ModuleFlatObjectNamed<'divi/icon'> |
  ModuleFlatObjectNamed<'divi/image'> |
  ModuleFlatObjectNamed<'divi/login'> |
  ModuleFlatObjectNamed<'divi/map'> |
  ModuleFlatObjectNamed<'divi/map-pin'> |
  ModuleFlatObjectNamed<'divi/menu'> |
  ModuleFlatObjectNamed<'divi/number-counter'> |
  ModuleFlatObjectNamed<'divi/portfolio'> |
  ModuleFlatObjectNamed<'divi/post-nav'> |
  ModuleFlatObjectNamed<'divi/post-slider'> |
  ModuleFlatObjectNamed<'divi/post-title'> |
  ModuleFlatObjectNamed<'divi/pricing-table'> |
  ModuleFlatObjectNamed<'divi/pricing-tables'> |
  ModuleFlatObjectNamed<'divi/root'> |
  ModuleFlatObjectNamed<'divi/row'> |
  ModuleFlatObjectNamed<'divi/row-inner'> |
  ModuleFlatObjectNamed<'divi/search'> |
  ModuleFlatObjectNamed<'divi/section'> |
  ModuleFlatObjectNamed<'divi/shortcode-module'> |
  ModuleFlatObjectNamed<'divi/sidebar'> |
  ModuleFlatObjectNamed<'divi/signup'> |
  ModuleFlatObjectNamed<'divi/signup-custom-field'> |
  ModuleFlatObjectNamed<'divi/slide'> |
  ModuleFlatObjectNamed<'divi/slider'> |
  ModuleFlatObjectNamed<'divi/social-media-follow'> |
  ModuleFlatObjectNamed<'divi/social-media-follow-network'> |
  ModuleFlatObjectNamed<'divi/tab'> |
  ModuleFlatObjectNamed<'divi/tabs'> |
  ModuleFlatObjectNamed<'divi/team-member'> |
  ModuleFlatObjectNamed<'divi/testimonial'> |
  ModuleFlatObjectNamed<'divi/text'> |
  ModuleFlatObjectNamed<'divi/toggle'> |
  ModuleFlatObjectNamed<'divi/video'> |
  ModuleFlatObjectNamed<'divi/video-slider'> |
  ModuleFlatObjectNamed<'divi/video-slider-item'> |
  ModuleFlatObjectNamed<'unknown'>
)

/**
 * Module flat objects; collection of module flat object.
 *
 * @since ??
 */
export type ModuleFlatObjects<TOtherModuleFlatObjectItems = null> = Record<
  string,
  ModuleFlatObjectItems | TOtherModuleFlatObjectItems
>;
