// @divi/types/src/serialized-post/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type PostStatus } from '@divi/types-external';

import { type ModuleFlatObjects } from '../../module/state';
import { type ThemeBuilderLocation } from '../../theme-builder';


export namespace Store {
  export type ItemPath = string | string[];

  export interface SerializedPostSchema {
    title: string;
    content: string;
  }

  export type State = Partial<Record<ThemeBuilderLocation, SerializedPostSchema>>;

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace SyncToServer {
      export interface Params {
        postStatus?: PostStatus;
      }

      export type ReturnValue = {
        type: 'SYNC_TO_SERVER';
        payload: Partial<Record<ThemeBuilderLocation, ImmutableObject<ModuleFlatObjects>>>;
        postStatus?: PostStatus;
      }

      export type EffectFunction = (action: ReturnValue, store: Instance) => void;
    }

    export type FunctionMap = {
      syncToServer: (params: SyncToServer.Params) => SyncToServer.ReturnValue;
    };

    export type ReturnValue = SyncToServer.ReturnValue;
  };

  export namespace Selectors {

    /**
     * Interface for getState function overloading.
     *
     * @since ??
     */
    export interface GetState<T = ImmutableState> {
      // eslint-disable-next-line max-len
      <T1 extends keyof T, T2 extends keyof T[T1], T3 extends keyof T[T1][T2], T4 extends keyof T[T1][T2][T3]>(path: [T1, T2, T3, T4], defaultValue?: T[T1][T2][T3][T4]): T[T1][T2][T3][T4];
      // eslint-disable-next-line max-len
      <T1 extends keyof T, T2 extends keyof T[T1], T3 extends keyof T[T1][T2]>(path: [T1, T2, T3], defaultValue?: T[T1][T2][T3]): T[T1][T2][T3];
      <T1 extends keyof T, T2 extends keyof T[T1]>(path: [T1, T2], defaultValue?: T[T1][T2]): T[T1][T2];
      <T1 extends keyof T>(path: [T1], defaultValue?: T[T1]): T[T1];
      <T1 extends keyof T>(path: T1, defaultValue?: T[T1]): T[T1];
    }

    export type FunctionMap = {
      getState: GetState;
      getUsedGlobalColorIds: () => string[];
    }
  }

  export namespace Effects {
    export type FunctionMap = {
      SYNC_TO_SERVER: Actions.SyncToServer.EffectFunction[];
    }
  }
}
