// @divi/types/src/module/metadata/settings/index.ts
import { type FieldLibrary } from '../../../field-library';
import { type Settings as ModuleSettings } from '../../settings';
import { type FieldOrGroup } from '../../settings/field-or-group';


export namespace Settings {
  export namespace AttributeGroup {

    export namespace Item {

      /**
       * Group type "group" is used for one element attribute (eg. `title.decoration.background`)
       * that has its own group component.
       */
      export interface Group {
        groupType: 'group';
        panel: ModuleSettings.Panel.Name;
        priority: number;
        groupName: string;

        groupLabel?: string;
        overwrite?: boolean;
        multiElements?: boolean;

        component: ModuleSettings.Group.Library.ComponentMapItem;
      }

      /**
       * Group type `group-items` is used when the field of this group are added into pre-registered group.
       *
       * `group-items` was added because we notice when we're using `shared-group`
       * there are too much repetition / boilerplate-y code. This is an attempt to simplify it.
       */
      export interface GroupItems {
        groupType: 'group-items';
        items: Record<string, FieldOrGroup.Item & {

          /**
           * Slug of registered module settings group.
           *
           * TODO maybe rename this. A better name for this would be `groupName` but it has been
           * used on other occassion (even on `Field.ContainerProps` that builds `FieldOrGroup.Item`)
           * so it is better to use different name that is more specific to this use case.
           */
          groupSlug: string;
        }>;
      }

      /**
       * Group type `group-item` is used when the field of this group is added into pre-registered group.
       * `group-items` are used when there are multiple groups / fields registered into the same group
       * while `group-item` is the simplification of it where there's only one field / group registered
       * into a group.
       */
      export interface GroupItem {
        groupType: 'group-item';
        item: FieldOrGroup.Item & {

          /**
           * Slug of registered module settings group.
           *
           * TODO maybe rename this. A better name for this would be `groupName` but it has been
           * used on other occassion (even on `Field.ContainerProps` that builds `FieldOrGroup.Item`)
           * so it is better to use different name that is more specific to this use case.
           */
          groupSlug: string;
        };
      }

      /**
       * Group type "inside-group" is used for one element attribute (eg. `title.decoration.zIndex`)
       * that has its group component inside another group component (eg. `<ZIndexGroup />`
       * inside `<PositionSettingsGroup />`).
       */
      export interface InsideGroup {
        groupType: 'inside-group';
        panel: ModuleSettings.Panel.Name;
        groupName: string;

        component?: {
          props?: FieldLibrary.ComponentMap;
        }
      }

      /**
       * Group type "shared-group" is used for group component that is possibly shared by multiple
       * element attribute. For example in CTA module: `title.innerContent`, `content.innerContent`
       * and `button.innerContent` all use the same group component `<CompositeGroup />` on
       * `Content (panel) > Text (group)`.
       */
      export interface SharedGroup {
        groupType: 'shared-group';

        // Multiple group will have the same group property setup. Only one will be rendered,
        // and the rest acts as "render if not exist yet, otherwise skip"
        sharedGroup: Omit<Group, 'groupType'>;
        fields?: ModuleSettings.FieldOrGroup.Items;
      }

      /**
       * Group type "into-multiple-groups" is used for one element attribute (eg. `button.innerContent`)
       * that requires multiple group components (eg. `<CompositeGroup />` and `<ButtonLinkGroup />`).
       */
      export interface IntoMultipleGroups {
        groupType: 'into-multiple-groups';
        groups?: Record<string, Group | GroupItem | GroupItems | InsideGroup | SharedGroup>;
      }
    }

    /**
     * Refers to empty object.
     */
    export type DefaultValue = Record<string, never>;

    /**
     * Element Attribute's Group Settings.
     */
    export type Value =
      Item.Group |
      Item.GroupItems |
      Item.GroupItem |
      Item.InsideGroup |
      Item.SharedGroup |
      Item.IntoMultipleGroups;
  }
}
