// @divi/types/src/module/element/decoration/body-font/index.ts
import { type Css } from '../../../../css';
import { type FormatPropertySelectors } from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';
import { type Font } from '../font';
import { type TextShadow } from '../text-shadow';


export namespace BodyFont {
  export interface Attributes {
    body?: Font.Attributes;
    link?: Font.Attributes;
    ul?: Font.Attributes;
    ol?: Font.Attributes;
    quote?: Font.Attributes;
  }

  export namespace Style {

    // TODO replace `FontBodyStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: {
        body?: Font.Style.Props['propertySelectors'];
        link?: Font.Style.Props['propertySelectors'];
        ul?: ListFont.Style.Props['propertySelectors'];
        ol?: ListFont.Style.Props['propertySelectors'];
        quote?: BlockquoteFont.Style.Props['propertySelectors'];
      };
      important?: boolean | {
        body?: Font.Style.Props['important'];
        link?: Font.Style.Props['important'];
        ul?: ListFont.Style.Props['important'];
        ol?: ListFont.Style.Props['important'];
        quote?: BlockquoteFont.Style.Props['important'];
      }
    }
  }
}

export namespace ListFont {
  export namespace Style {

    // TODO replace `ListDeclarationParams` with this.
    export interface ListDeclarationParams {
      important: boolean;
      attrValue: Font.ListAttributeValue;
      returnType?: 'string' | 'key_value_pair';
    }

    // TODO replace `ListFontStyleProps` with this.
    export interface Props extends Components.Style.Props<Font.Attributes> {
      propertySelectors?: {
        font?: Font.Style.Props['propertySelectors']['font'];
        textShadow?: TextShadow.Style.Props['propertySelectors'];
        list?: FormatPropertySelectors<Css.Properties.ListFont>;
      };
      important?: boolean | {
        font?: FormatStyleImportant<Css.Properties.Font>;
        textShadow?: TextShadow.Style.Props['important'];
        list?: FormatStyleImportant<Css.Properties.ListFont>;
      };
    }
  }
}

export namespace BlockquoteFont {
  export namespace Style {

    // TODO replace `BlockquoteFontStyleProps` with this.
    export interface Props extends Components.Style.Props<Font.Attributes> {
      propertySelectors?: {
        font?: Font.Style.Props['propertySelectors']['font'];
        textShadow?: TextShadow.Style.Props['propertySelectors'];
        border?: FormatPropertySelectors<Css.Properties.BlockquoteFont>;
      };
      important?: boolean | {
        font?: FormatStyleImportant<Css.Properties.Font>;
        textShadow?: TextShadow.Style.Props['important'];
        border?: FormatStyleImportant<Css.Properties.BlockquoteFont>;
      };
    }
  }
}
