// @divi/types/src/field-library/cornerside-picker/index.ts
export namespace CornerSidePicker {

  /**
   * @typedef {object} OptionValue
   * @property {string} buttonValue Button value.
   * @property {string} buttonLabel Button label.
   */
  export interface OptionValue {
    buttonValue: string;
    buttonLabel: string;
  }

  /**
   * @typedef {object} Option
   */
  export type Mode = 'absolute' | 'fixed';

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {object} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    inputValue?: string;
  }

  /**
   * @typedef {object} Props
   * @property {string} [className] Class name.
   * @property {string} [defaultValue] Default value.
   * @property {string} mode Mode.
   * @property {string} name Name.
   * @property {Function} onChange On change callback.
   * @property {string} value Value.
   */
  export interface Props {
    className?: string;
    defaultValue?: string;
    mode: Mode;
    name: string;
    onChange: (params: OnChangeCallbackParams) => void;
    value: string;
  }
}
