// @divi/types/src/module/library/column/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface ColumnAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      type?: FormatBreakpointStateAttr<string>;
      savedSpecialtyColumnType?: FormatBreakpointStateAttr<string>;
      specialtyColumns?: FormatBreakpointStateAttr<string>;
      sectionColumnType?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Element.Decoration.PickedAttributes<
    'animation' |
    'background' |
    'border' |
    'boxShadow' |
    'button' |
    'disabledOn' |
    'filters' |
    'overflow' |
    'position' |
    'scroll' |
    'sizing' |
    'spacing' |
    'sticky' |
    'transform' |
    'transition' |
    'zIndex'
    >;
  };
}
