// @divi/types/src/module/configuration/index.ts
import { type ReactElement } from 'react';

import { type BlockConfiguration } from '@wordpress/blocks';

import { type ModuleConversionOutline } from '../../conversion';
import { type KeysOfUnion } from '../../utils';
import { type Edit } from '../edit';
import { type ModuleAttrs } from '../library';
import { type Settings } from '../settings';


export namespace Configuration {

  /**
   * Module that gets added as part of template when module is added.
   *
   * @since ??
   */
  export type Template<TChildModuleAttrs = ModuleAttrs> = (string | TChildModuleAttrs)[];

  /**
   * Templates of module attributes that gets inserted when module is added.
   *
   * @since ??
   */
  export type Templates<TChildModuleAttrs = ModuleAttrs> = Template<TChildModuleAttrs>[];

  /**
   * Attribute interface of module definition's attribute.
   *
   * NOTE: This is based off Gutenberg's block definition. We may or may not use this later since
   * Visual Builder module attribute is defined by its fields.
   *
   * @since ??
   */
  export interface Attribute {
    type?: 'array' | 'boolean' | 'object' | 'null' | 'number' | 'string';
  }

  /**
   * Module attributes definition.
   *
   * @since ??
   */
  export type Attributes<
    TAttrs = ModuleAttrs
  > = Record<KeysOfUnion<TAttrs>, Attribute>;

  export interface Value<
    TAttrs = ModuleAttrs,
    TChildModuleAttrs = ModuleAttrs
  > extends BlockConfiguration {
    titles: string;
    placeholderContent?: TAttrs;
    defaultAttrs?: TAttrs;
    defaultPrintedStyleAttrs?: TAttrs;
    defaultSettingsAttrs?: TAttrs;
    settings?: {
      content?: (props: Settings.Panel.Props) => ReactElement,
      design?: (props: Settings.Panel.Props) => ReactElement,
      advanced?: (props: Settings.Panel.Props) => ReactElement,
    };
    template?: Templates<TChildModuleAttrs>;
    parentsName?: string[];
    childrenName?: string[];
    supportChildObjects?: boolean;
    folder?: string;

    attributes: Partial<Attributes<TAttrs>>;

    // Can't use `icon` because it has been used by `BlockConfiguration`.
    // @todo reconsider to completely drop block definition for module declaration.
    moduleIcon?: string;

    edit?: never; // Drop the usage edit property. Use the renderers.edit property instead.

    save?: never; // Drop the usage save property. Use the renderers.save property instead.

    renderers: {
      edit: (props: Edit.Component.Props<TAttrs>) => ReactElement;
      wireframe?: (props:Edit.Component.ContainerProps<TAttrs>) => ReactElement;
    };

    d4Shortcode?: string;
    official: boolean;
    conversionOutline?: ModuleConversionOutline;

    credit?: {
      author?: string,
      authorURI?: string,
      moduleURI?: string,
    }

    moduleClassName?: string;
    moduleOrderClassName?: string;
  }
}
