// @divi/types/src/module/class-module-elements.ts
import {
  type ReactElement,
  type ReactNode,
} from 'react';

import {
  type AttrState,
  type Breakpoint,
} from '../format-attr';
import { type Components } from './components';
import { type Element } from './element';
import { type Metadata } from './metadata';

/**
 * ModuleElements class' type.
 */
export type ModuleElementsType<TModuleAttrs> = {
  attrs: Record<string, Element.Attributes>;
  id: string;
  moduleMetadata: Metadata.Values;
  breakpoint: Breakpoint;
  state: AttrState;
  orderClass: string;
  wrapperOrderClass: string;
  orderId: string;

  /**
   * Create new instance of ModuleElements class.
   */
  constructor: (params: {
    attrs: TModuleAttrs;
    id: string;
    moduleMetadata: Metadata.Values<TModuleAttrs>;
    breakpoint?: Breakpoint;
    state?: AttrState
  }) => void;

  /**
   * Render the module element with the specified attributes.
   */
  render(params: {
    allowEmptyValue?: boolean;
    allowHtml?: boolean;
    attrName: string;
    attrSubName?: string;
    elementProps?: Omit<Element.Decoration.Button.Component.Props, 'buttonAttr' | 'className' | 'innerContent'>;
    elementType?: 'heading'
      | 'headingLink'
      | 'button'
      | 'image'
      | 'imageLink'
      | 'element';
    htmlAttributes?: Components.InlineTextEditor.AdditionalProps;
    inlineEditor?: 'plainText' | 'richText' | false;
    tagName?: keyof Components.InlineTextEditor.AllowedTags  ;
    dynamicDataLoaderProps?: Components.DynamicDataLoader.Props;
    inlineTextEditorProps?: Partial<Components.InlineTextEditor.ContainerProps>;
    className?: string;
    children?: ReactNode;
    valueResolver?: (
      value: string,
      resolverArgs: {
        breakpoint:Breakpoint,
        state:AttrState,
        attrName:string,
      }
    ) => string;
    imageWrapperClassName?: string;
  }): ReactElement;

  /**
   * Set orderClass.
   */
  setOrderClass(orderClass: string): void;

  /**
   * Set the wrapper order class.
   */
  setWrapperOrderClass(wrapperOrderClass: string): void;

  /**
   * Render script data.
   */
  scriptData(params: {
    attrName?: string;
    scriptDataProps?: Partial<Element.ScriptData.HookParams>;
    attrsResolver?: (attrs: Element.Decoration.Attributes) => Element.Decoration.Attributes;
  }): ReactElement;

  /**
   * Applies style to a React element based on provided parameters.
   */
  style(params: {
    attrName?: string;
    styleProps?: Partial<Element.Style.Props>;
  }): ReactElement;

  /**
   * Create style components based on provided parameters.
   */
  styleComponents(params: {
    attrName?: string;
    styleComponentsProps?: Partial<Element.Components.Props>;
  }): ReactElement;
}
