// @divi/types/src/module/index.ts
import { type CSSProperties, type ElementType, type MutableRefObject, type ReactElement, type ReactNode } from 'react';
import { type Editor as TinyMCEEditorReact } from '@tinymce/tinymce-react';

import { type AppPreferences } from '../app-preferences';
import { type Fonts } from '../fonts';
import { type AttrState, type Breakpoint, type FormatBreakpointStateAttr, type FormatPropertySelectors } from '../format-attr';
import { type FormatStyleImportant } from '../style';
import { type StyleLibrary } from '../style-library';
import { type Dividers } from './element/advanced/dividers';
import { type IdClasses } from './element/advanced/id-classes';
import { type Text } from './element/advanced/text';
import { type Animation } from './element/decoration/animation';
import { type Background } from './element/decoration/background';
import { type BodyFont } from './element/decoration/body-font';
import { type Border } from './element/decoration/border';
import { type BoxShadow } from './element/decoration/box-shadow';
import { type Button } from './element/decoration/button';
import { type DisabledOn } from './element/decoration/disabled-on';
import { type Filters } from './element/decoration/filters';
import { type Font } from './element/decoration/font';
import { type HeadingFont } from './element/decoration/heading-font';
import { type Icon } from './element/decoration/icon';
import { type Overflow } from './element/decoration/overflow';
import { type Position } from './element/decoration/position';
import { type Scroll } from './element/decoration/scroll';
import { type Sizing } from './element/decoration/sizing';
import { type Spacing } from './element/decoration/spacing';
import { type Sticky } from './element/decoration/sticky';
import { type TextShadow } from './element/decoration/text-shadow';
import { type Transform } from './element/decoration/transform';
import { type Transition } from './element/decoration/transition';
import { type ZIndex } from './element/decoration/z-index';
import { Css as CssNamespace } from './css';
import { Default as DefaultNamespace } from './default';
import { DropZone as DropZoneNamespace } from './drop-zone';
import { Element as ElementNamespace } from './element';
import { type ModuleAttrs  } from './library';
import { Metadata as MetadataNamespace, type ModuleMetadata } from './metadata';
import { ScriptData as ScriptDataNamespace } from './script-data';
import { Settings as SettingsNamespace } from './settings';
import { Store as StoreNamespace } from './store';
import { Styles as StylesNamespace } from './styles';
import { Unsync as UnsyncNamespace } from './unsync';


export namespace Module {

  /* eslint-disable @typescript-eslint/no-unused-vars */
  // TODO figure out a way to declare nested namespace without disabling no-unused-vars rule
  // imported namespace as marked as unused by eslint using this config; it shouldn't happen in the first place.
  export import Css = CssNamespace;
  export import Default = DefaultNamespace;
  export import DropZone = DropZoneNamespace;
  export import Element = ElementNamespace;
  export import Metadata = MetadataNamespace;
  export import ScriptData = ScriptDataNamespace;
  export import Settings = SettingsNamespace;
  export import Store = StoreNamespace;
  export import Styles = StylesNamespace;
  export import Unsync = UnsyncNamespace;
  /* eslint-enable @typescript-eslint/no-unused-vars */


  // module/attributes
  export type ModuleElementAttrInnerContentValue = Element.InnerContent.Attributes;
  export type ModuleElementAttrInnerContent<TAttr = ModuleElementAttrInnerContentValue> = TAttr;
  export type ModuleElementAttrAdvanced<TAttr = Record<string, unknown>>
  = Element.Advanced.Attributes<TAttr>;
  export interface ModuleElementAttr {
    innerContent?: ModuleElementAttrInnerContentValue;
    decoration?: Element.Decoration.Attributes;
    advanced?: ModuleElementAttrAdvanced;
    meta?: Element.Meta.Attributes;
  }

  // module/layout/component/dynamic-data-loader
  export interface DynamicDataLoaderProps {

    /**
     * The type of loader to use. Default is `text`.
     */
    loaderType?: 'text' | 'image';

    /**
     * The custom class name to be added for the loader element.
     */
    loaderClassName?: string;

    /**
     * The custom inline style to be added for the loader element.
     */
    loaderInlineStyle?: CSSProperties;

    /**
     * The custom width of the loader element. Default is `300` for `text` type and `100` for `image` type.
     */
    loaderWidth?: number;

    /**
     * The custom height of the loader element. Default is `8` for `text` type and `100` for `image` type.
     */
    loaderHeight?: number;
  }

  // module/options/button
  export interface ButtonComponentProps {
    allowEmptyUrl?: boolean;
    buttonAttr?: Element.Decoration.Attributes['button'];
    className?: string;
    hasWrapper?: boolean;
    hasPreloader?: boolean;
    hasTextWrapper?: boolean;
    id?: string;
    innerContent?: Module.Element.Types.Button.InnerContent.Attributes;
    name?: string;
    renderIconAsDataAttribute?: boolean;
    type?: 'link' | 'button';
    dynamicDataLoaderProps?: DynamicDataLoaderProps;
  }

  // module/options/sticky/script-data
  export interface UseStickyScriptDataContainerParams {
    affectingAttrs: {
      position?: Position.Attributes,
      sizing?: Sizing.Attributes,
    };
    attr: Sticky.Attributes,
    id: string;
    selector: string;
  }

  // module/options/element/script-data
  // In the UseStickyScriptDataContainerParams interface selector is mandatory.
  // But in ElementScriptData -> Sticky the selector is optional, hence handled separately.
  export type ElementScriptExcludedScriptProps = 'attr' | 'id';
  export type UseElementScriptDataParams = {

      // General Props.
      selector?: string;
      id?: string;
      attrs: Element.Decoration.Attributes;

      // Script specific settings
      sticky?: Omit<UseStickyScriptDataContainerParams, ElementScriptExcludedScriptProps>;
  }

  // module/layout/component/inline-text-editor
  export interface InlineTextEditorAdditionalProps {
    onClick?: (event: MouseEvent) => void;
    href?: string;
    title?: string;
    target?: string;
    style?: Record<string, string>;
  }
  export type InlineTextEditorAllowedTags = Pick<HTMLElementTagNameMap,
    'h1' |
    'h2' |
    'h3' |
    'h4' |
    'h5' |
    'h6' |
    'a' |
    'p' |
    'div' |
    'span' |
    'strong' |
    'section'>;
  export type InlineTextEditorContainerProps = {
    attrName: string;
    moduleId: string;
    tagName:  keyof InlineTextEditorAllowedTags;
    text: string;

    // Optional props.
    additionalProps?: InlineTextEditorAdditionalProps;
    allowHTML?: boolean;
    attrSubName?: string;
    callbackEditingStatusChange?: (event: FocusEvent, isEditing: boolean) => void;
    defaultText?: string;
    className?: string;
    handleBlur?: (event: FocusEvent) => void;
    handleClick?: (event: MouseEvent) => void;
    handleDoubleClick?: (event: MouseEvent) => void;
    handleKeyDown?: (event: KeyboardEvent) => void;
    handleKeyUp?: (event: KeyboardEvent) => void;
    handleMouseDown?: (event: MouseEvent) => void;
    handleMouseUp?: (event: MouseEvent) => void;
    handlePaste?: (event: ClipboardEvent) => void;
    handleShortcode?: (content: string) => string;
  }

  // module/layout/component/inline-richtext-editor
  export type InlineRichTextEditorAllowedTags = Pick<HTMLElementTagNameMap, 'div'>;

  export type InlineRichTextEditorEmptyEvent = Record<string, unknown>;

  export type InlineRichTextEditorActivateEvent = {
    relatedTarget: TinyMCEEditor;
    preventDefault(): void;
  };

  export type InlineRichTextEditorBlurEvent = {
    focusedEditor: TinyMCEEditor;
    preventDefault(): void;
  };

  export type InlineRichTextEditorFocusEvent = {
    blurredEditor: TinyMCEEditor;
    preventDefault(): void;
  };

  export interface InlineRichTextEditorExecCommandEvent {
    command: string;
    ui?: boolean;
    value?: any;
  }

  export interface InlineRichTextEditorNodeChangeEvent {
    element: Element;
    parents: Node[];
    selectionChange?: boolean;
    initial?: boolean;
  }

  export interface InlineRichTextEditorContainerState {
    editorActivated:   boolean,
    isSelected:        boolean,
    lastEditorContent: boolean | string,
    shortcodesDecoded: boolean,
    showPopover:       boolean,
    shortcodeHTML:     [],
    position?:   {
      x?: number;
      y?: number;
    };
    tinymceEditorId?:  string;
  }

  export interface InlineRichTextEditorContainerProps {
    tagName:  keyof InlineRichTextEditorAllowedTags;
    activateEditor?: () => void;
    attrName?: string;
    attrSubName?: string;
    breakpoint?: Breakpoint;
    className?: string;
    content?: string;
    contentProcessor?: (content: string) => string;
    defaultContent?: string;
    googleFonts?: Fonts.Google.Fonts;
    handleActivate?: (event: InlineRichTextEditorActivateEvent, editor?: TinyMCEEditor) => void;
    handleBlur?: (event: InlineRichTextEditorBlurEvent, editor?: TinyMCEEditor) => void;
    handleContextMenu?: (event: MouseEvent, editor?: TinyMCEEditor) => void;
    handleDoubleClick?: (event: MouseEvent, editor?: TinyMCEEditor) => void;
    handleEditorChange?: (newValue: string, editor?: TinyMCEEditor) => void;
    handleExecCommand?: (event: InlineRichTextEditorExecCommandEvent, editor?: TinyMCEEditor) => void;
    handleFocus?: (event: InlineRichTextEditorFocusEvent, editor?: TinyMCEEditor) => void;
    handleFocusOut?: (event: FocusEvent, editor?: TinyMCEEditor) => void;
    handleKeyDown?: (event: KeyboardEvent, editor?: TinyMCEEditor) => void;
    handleKeyUp?: (event: KeyboardEvent, editor?: TinyMCEEditor) => void;
    handleLoadContent?: (event: InlineRichTextEditorEmptyEvent, editor?: TinyMCEEditor) => void;
    handleMouseDown?: (event: MouseEvent, editor?: TinyMCEEditor) => void;
    handleMouseUp?: (event: MouseEvent, editor?: TinyMCEEditor) => void;
    handleNodeChange?: (event: InlineRichTextEditorNodeChangeEvent, editor?: TinyMCEEditor) => void;
    handleShortcode?: (content: string) => string;
    inlineEditorClick?: Module.Store.State['inlineEditorClick'];
    inlineEditorToolbar?: Module.Store.State['inlineEditorToolbar'];
    moduleId: string;
    rawContentProcessor?: (content: string) => string;
    setInlineEditorToolbar?: ({
      payload,
      clickInside,
      clickOutside,
    }: Module.Store.Actions.SetInlineEditorToolbar.Params) => void;
    state?: AttrState;
    tinyMCEPlugins?: string[];
  }

  export interface InlineRichTextEditorProps extends InlineRichTextEditorContainerProps {
    editable?: boolean;
    editing?:boolean;
  }


  // module/layout/component/style
  export interface SelectorFunctionParams<TAttr> {
    selector: string;
    breakpoint: string;
    state: string;
    attr: TAttr;
  }
  export type SelectorFunction = <TParams extends SelectorFunctionParams<TParams['attr']>>(params: TParams) => string;
  export type StyleMode = 'builder' | 'frontend';
  export interface StyleProps<TAttr> {
    selector?: string;
    selectors?: FormatBreakpointStateAttr<string>;
    selectorFunction?: SelectorFunction;
    attr: TAttr;
    defaultPrintedStyleAttr?: TAttr;
    asStyle?: boolean;
    orderClass?: string;

    // @todo set as required once mode is passed everywhere.
    mode?: StyleMode;
    state?: AttrState;
  }

  // module/options/css/style
  export interface CssStyleProps<TAttr> extends StyleProps<TAttr> {
    cssFields?: Record<string, Css.Field>;
  }

  // module/options/text-shadow/style
  export interface TextShadowStyleProps extends StyleProps<TextShadow.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.TextShadowStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.TextShadowStyleCssProperties>;
  }

  // module/options/text/style
  export interface TextStyleProps extends StyleProps<Text.Attributes> {
    propertySelectors?: {
      text?: FormatPropertySelectors<StyleLibrary.TextStyleCssProperties>;
      textShadow?: TextShadowStyleProps['propertySelectors'];
    };
    important?: boolean | {
      text?: boolean | FormatStyleImportant<StyleLibrary.TextStyleCssProperties>;
      textShadow?: TextShadowStyleProps['important'];
    };
    orientation?: boolean;
  }

  // module/layout/component/style-common
  export interface DeclarationFunctionProps<TAttrValue = string> {
    attrValue?: TAttrValue;
    defaultAttrValue?: TAttrValue;
    important?: boolean | Record<string, boolean>;
    returnType?: 'string' | 'key_value_pair';
    breakpoint?: Breakpoint;
    state?: AttrState;
  }
  export interface CommonStyleProps<TAttr, TReturn = string | Record<string, string>> extends StyleProps<TAttr> {
    property?: string;
    important?: boolean;
    declarationFunction?: (props: DeclarationFunctionProps<unknown>) => TReturn;
  }

  // module/options/animation/style
  export interface AnimationStyleProps extends StyleProps<Animation.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.AnimationStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.AnimationStyleCssProperties>;
  }

  // module/options/background/style
  export interface BackgroundStyleProps extends StyleProps<Background.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.BackgroundStyleCssProperties>;
    featureSelectors?: {
      mask?: FormatBreakpointStateAttr<string>;
      pattern?: FormatBreakpointStateAttr<string>;
    };
    important?: boolean | FormatStyleImportant<StyleLibrary.BackgroundStyleCssProperties>;
  }

  // module/options/font/style
  // <FontStyle /> has additional props; Extending `StyleProps` on styles/font/types will cause circular
  export interface FontStyleProps extends StyleProps<Font.Attributes> {
    headingLevel?: string | boolean;
    propertySelectors?: {
      font?: FormatPropertySelectors<StyleLibrary.FontStyleCssProperties>;
      textShadow?: TextShadowStyleProps['propertySelectors'];
    };
    important?: boolean | {
      font?: FormatStyleImportant<StyleLibrary.FontStyleCssProperties>;
      textShadow?: TextShadowStyleProps['important'];
    };
  }

  // module/options/font-body-group/style
  export interface BlockquoteFontStyleProps extends StyleProps<Font.Attributes> {
    propertySelectors?: {
      font?: FontStyleProps['propertySelectors']['font'];
      textShadow?: TextShadowStyleProps['propertySelectors'];
      border?: FormatPropertySelectors<StyleLibrary.BlockquoteFontStyleCssProperties>;
    };
    important?: boolean | {
      font?: FormatStyleImportant<StyleLibrary.FontStyleCssProperties>;
      textShadow?: TextShadowStyleProps['important'];
      border?: FormatStyleImportant<StyleLibrary.BlockquoteFontStyleCssProperties>;
    };
  }
  export interface ListFontStyleProps extends StyleProps<Font.Attributes> {
    propertySelectors?: {
      font?: FontStyleProps['propertySelectors']['font'];
      textShadow?: TextShadowStyleProps['propertySelectors'];
      list?: FormatPropertySelectors<StyleLibrary.ListFontStyleCssProperties>;
    };
    important?: boolean | {
      font?: FormatStyleImportant<StyleLibrary.FontStyleCssProperties>;
      textShadow?: TextShadowStyleProps['important'];
      list?: FormatStyleImportant<StyleLibrary.ListFontStyleCssProperties>;
    };
  }
  export interface FontBodyStyleProps extends StyleProps<BodyFont.Attributes> {
    propertySelectors?: {
      body?: FontStyleProps['propertySelectors'];
      link?: FontStyleProps['propertySelectors'];
      ul?: ListFontStyleProps['propertySelectors'];
      ol?: ListFontStyleProps['propertySelectors'];
      quote?: BlockquoteFontStyleProps['propertySelectors'];
    };
    important?: boolean | {
      body?: FontStyleProps['important'];
      link?: FontStyleProps['important'];
      ul?: ListFontStyleProps['important'];
      ol?: ListFontStyleProps['important'];
      quote?: BlockquoteFontStyleProps['important'];
    };
  }

  // module/options/border/style
  export interface BorderStyleProps extends StyleProps<Border.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.BorderStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.BorderStyleCssProperties>;
  }

  // module/options/box-shadow/style
  export interface BoxShadowStyleProps extends StyleProps<BoxShadow.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.BoxShadowCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.BoxShadowCssProperties>;
    useOverlay?: boolean;
  }

  // module/options/filter/style
  export interface FiltersStyleProps extends StyleProps<Filters.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.FiltersStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.FiltersStyleCssProperties>;
  }

  // module/options/transition/style
  export type TransitionElementAttrs = {
    animation?: Animation.Attributes;
    background?: Background.Attributes;
    bodyFont?: BodyFont.Attributes;
    border?: Border.Attributes;
    boxShadow?: BoxShadow.Attributes;
    transform?: Transform.Attributes;
    disabledOn?: DisabledOn.Attributes;
    filter?: Filters.Attributes;
    font?: Font.Attributes;
    headingFont?: HeadingFont.Attributes;
    overflow?: Overflow.Attributes;
    position?: Position.Attributes;
    scroll?: Scroll.Attributes;
    sizing?: Sizing.Attributes;
    spacing?: Spacing.Attributes;
    sticky?: Sticky.Attributes;
    zIndex?: ZIndex.Attributes;
  };
  export interface TransitionStyleProps extends StyleProps<Transition.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.TransitionStyleCssProperties>;
    attrs?: TransitionElementAttrs;
    important?: boolean;
    advancedStyles?: ElementAdvancedStyles[];
    transitionData?: {
      attrs: TransitionElementAttrs;
      props: Record<string, unknown>;
    };
  }

  // module/options/transfer/style
  export interface TransformStyleProps extends StyleProps<Transform.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.TransformStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.TransformStyleCssProperties>;
  }

  // module/options/disabled-on
  export interface DisabledModuleStyleProps extends Omit<StyleProps<DisabledOn.Attributes>, 'selectors'> {
    selectors?: {
      desktopAbove?: Partial<Record<AttrState, string>>;
      tabletOnly?: Partial<Record<AttrState, string>>;
      phone?: Partial<Record<AttrState, string>>;
    };
    propertySelectors?: FormatPropertySelectors<StyleLibrary.DisabledOnStyleCssProperties>;
    disabledModuleVisibility: AppPreferences.DisabledModule.Visibility;
  }

  // module/options/font-header-group/style
  export interface FontHeaderStyleProps extends StyleProps<HeadingFont.Attributes>{
    propertySelectors?: {
      h1?: FontStyleProps['propertySelectors'];
      h2?: FontStyleProps['propertySelectors'];
      h3?: FontStyleProps['propertySelectors'];
      h4?: FontStyleProps['propertySelectors'];
      h5?: FontStyleProps['propertySelectors'];
      h6?: FontStyleProps['propertySelectors'];
    };
    important?: boolean | {
      h1?: FontStyleProps['important'];
      h2?: FontStyleProps['important'];
      h3?: FontStyleProps['important'];
      h4?: FontStyleProps['important'];
      h5?: FontStyleProps['important'];
      h6?: FontStyleProps['important'];
    };
  }

  // module/options/icon/style
  export interface IconStyleProps extends StyleProps<Icon.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.IconStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.IconStyleCssProperties>;
  }

  // module/options/overflow/style
  export interface OverflowStyleProps extends StyleProps<Overflow.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.OverflowStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.OverflowStyleCssProperties>;
  }

  // module/options/position/style
  export interface PositionStyleProps extends StyleProps<Position.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.PositionStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.PositionStyleCssProperties>;
  }

  // module/options/sizing/style
  export interface SizingStyleProps extends StyleProps<Sizing.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.SizingStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.SizingStyleCssProperties>;
  }

  // module/options/spacing/style
  export interface SpacingStyleProps extends StyleProps<Spacing.Attributes> {
    propertySelectors?: FormatPropertySelectors<StyleLibrary.SpacingStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.SpacingStyleCssProperties>;
  }

  // module/options/z-index/style
  export interface ZIndexStyleProps extends StyleProps<ZIndex.Attributes> {
    important?: boolean | FormatStyleImportant<'z-index'>;
  }

  // module/options/button/style
  export interface ButtonStyleProps extends StyleProps<Button.Attributes> {
    affectingAttrs?: {
      spacing?: Spacing.Attributes
    };
    propertySelectors?: FormatPropertySelectors<StyleLibrary.ButtonStyleCssProperties>;
    important?: boolean | FormatStyleImportant<StyleLibrary.ButtonStyleCssProperties>;
    selectorFunctions?: {
      buttonHoverSelectorFunction?: SelectorFunction;
      buttonIconSelectorBeforeAfterFunction?: SelectorFunction;
      buttonIconSelectorFunction?: SelectorFunction;
      buttonIconSelectorHoverFunction?: SelectorFunction;
      buttonIconSelectorRightFunction?: SelectorFunction;
    };
  }

  // module/options/background
  export type BackgroundComponentProps = {
    attr: Background.Attributes;
    id: string;
    settings?: {
      mask?: boolean;
      parallax?: boolean;
      pattern?: boolean;
      video?: boolean;
    } | boolean;
  }

  // module/options/box-shadow
  export interface BoxShadowComponentProps {
    attr: BoxShadow.Attributes;
    id: string;
    settings?: {
      overlay?: boolean;
    } | boolean;
  }

  // module/options/divider/style
  type DividersStyleExcludedStyleProps = 'selectors' | 'selectorFunction';
  export interface DividersStyleProps extends Omit<
    StyleProps<FormatBreakpointStateAttr<Dividers.AttributeValue>>, DividersStyleExcludedStyleProps
  > {
    fullwidth?: boolean;
    placement: 'top' | 'bottom';
    backgroundColors?: {
      siblingBackgroundAttr?: Background.Attributes;
      moduleBackgroundAttr?: Background.Attributes,
      defaultColor?: string;
    }
    featureSelectors?: {
      top?: FormatBreakpointStateAttr<string>;
      bottom?: FormatBreakpointStateAttr<string>;
    };
    important?: boolean | FormatStyleImportant<StyleLibrary.DividersStyleCssProperties>;
  }

  // module/options/element/components
  export type BackgroundSettingsType = boolean | Omit<BackgroundComponentProps, 'attr' | 'id'>;
  export type BoxShadowSettingsType = boolean | Omit<BoxShadowComponentProps, 'attr' | 'id'>;
  export type ElementComponentsProps = {
    id?: string;
    attrs: Element.Decoration.Attributes;
    background?: BackgroundSettingsType;
    boxShadow?: BoxShadowSettingsType;
  }


  // module/options/element/style
  export type ElementAdvancedStyles = {
    props: CommonStyleProps<unknown>
    | CssStyleProps<string>
    | TextStyleProps
    | BackgroundStyleProps
    | DividersStyleProps;
    componentName: 'divi/common' | 'divi/text' | 'divi/css' | 'divi/background' | 'divi/dividers';
  }
  type ElementStyleExcludedStyleProps = 'attr' | 'asStyle' | 'mode' | 'orderClass';
  export type ElementStyleType = 'module' | 'element' | 'button';
  export type ElementStyleProps = {

    // General Props.
    selector?: string;
    attrs: Element.Decoration.Attributes;
    defaultPrintedStyleAttrs?: Element.Decoration.Attributes;
    attrsFilter?: (attrs: Element.Decoration.Attributes) => Element.Decoration.Attributes;
    type?: ElementStyleType;
    advancedStyles?: ElementAdvancedStyles[];
    orderClass?: string;

    // Style specific settings
    animation?: Omit<AnimationStyleProps, ElementStyleExcludedStyleProps>;
    background?: Omit<BackgroundStyleProps, ElementStyleExcludedStyleProps>;
    bodyFont?: Omit<FontBodyStyleProps, ElementStyleExcludedStyleProps>;
    border?: Omit<BorderStyleProps, ElementStyleExcludedStyleProps>;
    boxShadow?: Omit<BoxShadowStyleProps, ElementStyleExcludedStyleProps>;
    filters?: Omit<FiltersStyleProps, ElementStyleExcludedStyleProps>;
    transform?: Omit<TransformStyleProps, ElementStyleExcludedStyleProps>;
    transition?: Omit<TransitionStyleProps, ElementStyleExcludedStyleProps>;
    disabledOn?: Omit<DisabledModuleStyleProps, ElementStyleExcludedStyleProps>;
    font?: Omit<FontStyleProps, ElementStyleExcludedStyleProps>;
    headingFont?: Omit<FontHeaderStyleProps, ElementStyleExcludedStyleProps>;
    icon?: Omit<IconStyleProps, ElementStyleExcludedStyleProps>;
    overflow?: Omit<OverflowStyleProps, ElementStyleExcludedStyleProps>;
    position?: Omit<PositionStyleProps, ElementStyleExcludedStyleProps>;
    sizing?: Omit<SizingStyleProps, ElementStyleExcludedStyleProps>;
    spacing?: Omit<SpacingStyleProps, ElementStyleExcludedStyleProps>;
    zIndex?: Omit<ZIndexStyleProps, ElementStyleExcludedStyleProps>;
    button?: Omit<ButtonStyleProps, ElementStyleExcludedStyleProps>;
  }


  // module/layout/components/ elements --------------------------------------
  export type ModuleElementsComponentParams = {
    attrName?: string;
    styleComponentsProps?: Partial<ElementComponentsProps>;
  }
  export type ModuleElementsStyleParams = {
    attrName?: string;
    styleProps?: Partial<ElementStyleProps>;
  }
  export type ModuleElementsScriptDataParams = {
    attrName?: string;
    scriptDataProps?: Partial<UseElementScriptDataParams>;
    attrsResolver?: (attrs: Element.Decoration.Attributes) => Element.Decoration.Attributes;
  }
  export type ValueResolverArgs = {
    breakpoint:Breakpoint;
    state:AttrState;
    attrName:string;
  }
  export type ValueResolver = (value: string, resolverArgs: ValueResolverArgs) => string;
  export type ModuleElementsRenderParams = {
    allowEmptyValue?: boolean;
    allowHtml?: boolean;
    attrName: string;
    attrSubName?: string;
    elementProps?: Omit<ButtonComponentProps, 'buttonAttr' | 'className' | 'innerContent'>;
    elementType?: 'heading'
      | 'headingLink'
      | 'button'
      | 'image'
      | 'imageLink'
      | 'element';
    htmlAttributes?: InlineTextEditorAdditionalProps;
    inlineEditor?: 'plainText' | 'richText' | false;
    tagName?: keyof InlineTextEditorAllowedTags | keyof InlineRichTextEditorAllowedTags;
    dynamicDataLoaderProps?: DynamicDataLoaderProps;
    inlineTextEditorProps?: Partial<InlineTextEditorContainerProps>;
    className?: string;
    children?: ReactNode;
    valueResolver?: ValueResolver;
    imageWrapperClassName?: string;
  }

  export type ModuleElements = {
    attrs: Record<string, Module.Element.Attributes>;
    id: string;
    moduleMetadata: ModuleMetadata;
    breakpoint: Breakpoint;
    state: AttrState;
    orderClass: string;
    wrapperOrderClass: string;
    orderId: string;
    render: (param: ModuleElementsRenderParams) => ReactElement;
    setOrderClass: (orderClass: string) => void;
    setWrapperOrderClass: (wrapperOrderClass: string) => void;
    setOrderId: (orderId: string) => void;
    scriptData: ({
      attrName,
      scriptDataProps,
      attrsResolver,
    }:ModuleElementsScriptDataParams) => ReactElement;
    style: ({
      attrName,
      styleProps,
    }:ModuleElementsStyleParams) => ReactElement;
    styleComponents: ({ attrName, styleComponentsProps }: ModuleElementsComponentParams) => ReactElement;
  }

  // --------------------------------------

  export type StylesProps<TAttrs, TParentAttrs = unknown> = {
    id?: string;
    name?: string;
    attrs: TAttrs;
    parentAttrs?: TParentAttrs;
    siblingAttrs?: {
      previous?: TAttrs;
      next?: TAttrs;
    };
    defaultPrintedStyleAttrs?: TAttrs;
    orderClass: string;
    parentOrderClass?: string;
    wrapperOrderClass?: string;
    state?: AttrState;
    mode?: 'builder' | 'frontend';
    noStyleTag?: boolean;
    settings?: {
      disabledModuleVisibility?: AppPreferences.DisabledModule.Visibility;
    }
    pageSettings?: {
      sectionBackgroundColor?: string;
    }
    elements?: ModuleElements;
  }

  export type ModuleClassnames = {
    values: Record<string, boolean>;
    value: string;
    add: (classname: string | string[] | Record<string, boolean>, shouldBeAdded?: boolean) => void;
    remove: (classname: string) => void;
  }

  export type ModuleClassnamesParams<TAttrs, TParentAttrs = unknown> = {
    childrenIds?: string[];
    classnamesInstance: ModuleClassnames;
    attrs: TAttrs;
    parentAttrs?: TParentAttrs;
    hasModule?: boolean;
    name?: string;
    id?: string;
    moduleOrderIndex?: number;
    isFirst?: boolean;
    isLast?: boolean;
    state?: AttrState;
    breakpoint?: Breakpoint;
  }

  export type ModuleScriptDataProps<TAttrs> = {
    attrs?: TAttrs
    selector?: string;
    id?: string;
    elements?: ModuleElements;
  }

  export type ModuleBaseProps<TAttrs, TParentAttrs = unknown> = {
    id: string;
    name: string;
    className?: string;
    children?: ReactNode;
    childrenIds?: string[];
    hasModule?: boolean;
    moduleOrderIndex?: number;
    isLast?: boolean;
    isFirst?: boolean;
    hasModuleClassName?: boolean;
    hasHoveredClassName?: boolean;
    hasAnimatedClassName?: boolean;
    parentAttrs?: TParentAttrs;
    parentId?: string;
    parentName?: string;
    stylesComponent?: (props: StylesProps<TAttrs>) => ReactElement;
    scriptDataComponent?: (props: ModuleScriptDataProps<TAttrs>) => ReactElement;

    // feat(D5, Dev) :: To be removed
    moduleClassName?: string;
    moduleOrderClassName?: string;

    // feat(D5, Dev) :: Once all module has this, make this mandatory.
    classnamesFunction?: (params: ModuleClassnamesParams<TAttrs>) => void;
    wrapperClassnamesFunction?: (params: ModuleClassnamesParams<TAttrs>) => void;

    htmlAttributesFunction?: (params: IdClasses.Attributes) => Record<'id' | 'classNames', string>;
  }

  export type ModuleContainerProps<TAttrs> = Omit<ModuleBaseProps<TAttrs>, 'moduleClassName' | 'moduleOrderClassName'>
  & {
    attrs: TAttrs;
    parentAttrs?: TAttrs;
    defaultPrintedStyleAttrs?: TAttrs;
    domRef?: MutableRefObject<HTMLDivElement>;
    elements?: ModuleElements;
    htmlAttrs?: Record<string, string | number>;
    innerHTML?: string;
    originalName?: string;
    siblingAttrs?: {
      previous?: TAttrs;
      next?: TAttrs;
    };
    pageSettings?: {
      sectionBackgroundColor?: string;
    }
    tag?: ElementType;
    hasModuleWrapper?: boolean;
    wrapperTag?: ElementType;
    wrapperHtmlAttrs?: Record<string, string | number>;
    wrapperClassname?: string;
    cssPosition?: 'before' | 'after';
  }

  export type SettingsProps<
    TAttrs = ModuleAttrs,
    TParentAttrs = ModuleAttrs
  > = Module.Settings.Panel.Props<TAttrs, TParentAttrs>;

  // module/layout/component/inline-richtext-editor/toolbar
  export interface InlineEditorToolbarContainerProps {
    active: boolean;
    animated: boolean;
    isSelected: boolean;
    isSelectionChanged: boolean;
    moduleId: string;
    position: {
      x?: number;
      y?: number;
    }
    tinymceEditorId: string;
  }

  export interface InlineEditorToolbarProps extends InlineEditorToolbarContainerProps {
    colorPalette: string[];
    editingColorsAllowed: boolean;
    editingFontsAllowed: boolean;
    googleFonts?: Fonts.Google.Fonts,
    customFonts?: Fonts.Custom.Fonts;
    removedFonts?: Fonts.Removed.Fonts;
    viewport: {
      height: number;
      width: number;
    }
  }

  export type InlineRichTextEditorMenuItem = {
    key: string;
    buttonCommand?: string;
    formatBlock?: string;
    className?: string[];
    iconName?: string;
    iconColor?: string;
    linkToMenu?: string;
    tooltip?: string;
    label?: string;
    tooltipText?: string;
  }
  export interface InlineRichTextEditorMenuProps {
    animation?: boolean;
    colorPalette?: string[];
    tinymceEditorId: string;
    isSelectionChanged?: boolean;
    isWireframe?: boolean;
    items: InlineRichTextEditorMenuItem[];
    moduleId?: string;
    position?: {
      x: number;
      y: number;
    };
    googleFonts?: Fonts.Google.Fonts;
    customFonts?: Fonts.Custom.Fonts;
    removedFonts?: Fonts.Removed.Fonts;
    backToMainMenu?: () => void;
    onClick: (event: any, menuItem: InlineRichTextEditorMenuItem) => void;
  }

  export interface CurrentToolbarAreaName {
    mainMenu: 'mainMenu';
    linkSettingsMenu: 'linkSettingsMenu';
    listSettingsMenu: 'listSettingsMenu';
    textSettingsMenu: 'textSettingsMenu';
    textColorMenu: 'textColorMenu';
    textAlignmentMenu: 'textAlignmentMenu';
  }

  export interface RangeControlProps {
    id: string;
    name: string;
    max: number;
    min: number;
    step: number;
    value: number;
    onChange: (inputValue: string) => void;
  }

  export interface SelectControlProps {
    id: string;
    moduleId: string;
    value: string;
    googleFonts?: Fonts.Google.Fonts;
    customFonts?: Fonts.Custom.Fonts;
    tinymceEditorId: string;
    onChange: (fontName: string) => void;
  }

  export interface ColorControlProps {
    hideOnLoad: boolean;
    name: string;
    onChange: (value: string) => void;
    colorPalette: string[];
    type: string;
    value: string;
  }

  export interface ColorPickerUIParams {
    color: {
      toString: () => string;
    };
  }

  // --------------------------------------
}

/**
 * These types are declared at @divi/types instead of @divi/module because doing so
 * will cause circular dependency because these types usage are so wide. Placing
 * this on @divi/types avoids such problem because @divi/types has no internal dependencies.
 */

/**
 * TinyMCE editor instance.
 */
export type TinyMCEEditor = TinyMCEEditorReact['editor'] | null;

/**
 * Module attribute value type.
 *
 * @since ??
 */
export type ModuleDefinitionAttributeType = 'array' | 'boolean' | 'object' | 'null' | 'number' | 'string';

/**
 * Module drop zone position.
 *
 * @since ??
 */
export type DropZonePosition = 'top' | 'bottom' | 'inside';

/**
 * Module categorization.
 */
export * from './category';

/**
 * Element types & interfaces.
 */
export * from './element';

/**
 * All module option's attributes.
 */
export * from './options';

/**
 * All module library's attributes.
 */
export * from './library';

/**
 * All module state related types.
 */
export * from './state';

/**
 * Module metadata types.
 */
export * from './metadata';

/**
 * Dynamic Content types.
 */
export * from './dynamic-content';
