// @divi/types/src/module/settings/group/library/dividers/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Dividers {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Advanced.Dividers.Attributes;
    defaultTab?: Element.Advanced.Dividers.Placement;
    fields?: {
      tabNav?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      style?: GroupFieldProps & {
        defaultAttr?: Element.Advanced.Dividers.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select-divider']
      };
      color?: GroupFieldProps & {
        defaultAttr?: Element.Advanced.Dividers.Attributes;
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      height?: GroupFieldProps & {
        defaultAttr?: Element.Advanced.Dividers.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      repeat?: GroupFieldProps & {
        defaultAttr?: Element.Advanced.Dividers.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      flip?: GroupFieldProps & {
        defaultAttr?: Element.Advanced.Dividers.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      arrangement?: GroupFieldProps & {
        defaultAttr?: Element.Advanced.Dividers.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
    }
  }
}
