// @divi/types/src/module/library/post-title/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface PostTitleAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'conditions' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Title
  title?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
    advanced?: {
      showTitle?: FormatBreakpointStateAttr<OnOff>;
    };
  };

  // Meta
  meta?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
    advanced?: {
      showMeta?: FormatBreakpointStateAttr<OnOff>;
      showAuthor?: FormatBreakpointStateAttr<OnOff>;
      showDate?: FormatBreakpointStateAttr<OnOff>;
      dateFormat?: FormatBreakpointStateAttr<string>;
      showCategories?: FormatBreakpointStateAttr<OnOff>;
      showCommentsCount?: FormatBreakpointStateAttr<OnOff>;
    };
  };

  // Text Wrapper
  textWrapper?: {
    decoration?: Element.Decoration.PickedAttributes<'background'>;
    advanced?: {
      useBackground?: FormatBreakpointStateAttr<OnOff>;
    }
  }

  // Featured Image
  featuredImage?: {
    decoration?: Element.Decoration.PickedAttributes<
      'sizing'
    >;

    advanced?: {
      enabled?: FormatBreakpointStateAttr<OnOff>;
      forceFullwidth?: FormatBreakpointStateAttr<OnOff>;
      placement?: FormatBreakpointStateAttr<'below' | 'above' | 'background'>;
    };
  }
}
