// @divi/types/src/module/settings/group/library/font/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type Field } from '../../../field';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';
import { type TextShadow } from '../text-shadow';


export namespace Font {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Font.Attributes;
    fields?: {
      headingLevel?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      family?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        component?: FieldLibrary.ComponentMap['divi/select-font'];
      };
      weight?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      style?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        onChange?: Field.ContainerProps['onChange'];
        onChangeParams?: Field.ContainerProps['onChangeParams'];
        multipleChoices?: boolean;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      lineColor?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        visible?: Field.ContainerProps['visible'];
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      lineStyle?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        visible?: Field.ContainerProps['visible'];
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      textAlign?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      color?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      size?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      letterSpacing?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      lineHeight?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['font'];
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      textShadowGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/text-shadow',
          props: TextShadow.GroupProps;
        }
      };
      listType?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['list'];
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      listPosition?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['list'];
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      listItemIndent?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['list'];
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      quoteStylesLeftWidth?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['border'];
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      quoteStylesLeftColor?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.Attributes['border'];
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
    }
  }

  // Previously referred as `MultiFontGroupProps`
  export interface MultiGroupProps {
    attrName?: string;
    title?: string;
    grouped?: boolean;
  }
}
