// @divi/types/src/module/library/counters/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface BarCountersAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'conditions' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Title
  title?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };


  // percent
  barProgress?: {
    advanced?: {
      usePercentages?: FormatBreakpointStateAttr<OnOff>
    };
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Bar Counter
  barCounter?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'border' |
      'boxShadow'
    >;
  }

  children?: {
    barProgress?: {
      decoration?: Element.Decoration.PickedAttributes<'background'>;
    };
  }
}
