// @divi/types/src/module/settings/group/library/box-shadow/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace BoxShadow {
  export interface StyleOption {
    icon?: string,
    className?: string;
    label: string
  }

  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.BoxShadow.Attributes;
    fields?: {
      style?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.BoxShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      horizontal?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.BoxShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      vertical?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.BoxShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      blur?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.BoxShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      spread?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.BoxShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      color?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.BoxShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      position?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.BoxShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
    }
  }
}
