// @divi/types/src/module/settings/configuration/index.ts
import { type Element } from '../../element';
import { type Group } from '../group';


export namespace Configuration {
  export interface Item<
    TGroupName extends keyof Group.Library.ComponentMap = keyof Group.Library.ComponentMap
  > {
    priority: number;
    render: boolean;
    component: {

      // TODO this can be removed if group rendering is abstracted from <ModuleFields /> into its own component.
      type: 'group';
      name: TGroupName;
      props?: Partial<Group.Library.ComponentMap[TGroupName]['props']> & {

        // TODO maybe refactor this. Some group component has no `defaultGroupAttr` type which causes TS
        // to complain when `defaultGroupAttr` is manually set up for the rest of the group component.
        // The challenge here is there's no way to detect and only set `defaultGroupAttr` to group component
        // that has it only.
        defaultGroupAttr?: Element.AttributesPart;
      };
    }
  }

  export type Items = Record<string, Item>;

  // TODO Refactor this so panel is not hardcoded.
  export interface Value {
    content?: Items;
    design?: Items;
    advanced?: Items;
  }
}
