// @divi/types/src/module/element/decoration/position/index.ts
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';


export namespace Position {
  export interface AttributeValue {
    mode?: 'default' | 'relative' | 'absolute' | 'fixed';
    origin?: {
      relative?: 'top left' | 'top right' | 'bottom left' | 'bottom right';
      absolute?: 'top left' | 'top center' | 'top right' | 'center left' | 'center center' | 'center right' | 'bottom left' | 'bottom center' | 'bottom right';
      fixed?: 'top left' | 'top center' | 'top right' | 'center left' | 'center center' | 'center right' | 'bottom left' | 'bottom center' | 'bottom right';
    };
    offset?: {
      vertical?: string;
      horizontal?: string;
    }
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Style {

    // TODO replace `PositionStyleProps` with this
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.Position>;
      important?: boolean | FormatStyleImportant<Css.Properties.Position>;
    }

  }
}
