// @divi/types/src/module/settings/group/library/composite/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type FormatBreakpointStateAttr } from '../../../../../format-attr';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';
import { type GroupLibraryMapWithoutComposite } from '../map';


export namespace Composite {

  /**
   * Group component map without `divi/composite`'s group props.
   */
  type GroupComponentMap = {
    [TName in keyof GroupLibraryMapWithoutComposite]?: {
      type?: 'group';
      name?: TName;
      props?: Partial<GroupLibraryMapWithoutComposite[TName]['props']>;
    };
  }

  type FieldComponentProps = GroupFieldProps & {
    defaultAttr?: FormatBreakpointStateAttr<unknown>;
    component?: FieldLibrary.Component
  };

  type GroupComponentProps = GroupFieldProps & GroupComponentMap;

  export interface GroupProps extends GroupPropsSource {
    fields?: {
      [key: string]: FieldComponentProps | GroupComponentProps;
    }
  }
}
