// @divi/types/src/module/settings/group/library/button/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type Field } from '../../../field';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';
import { type Background } from '../background';
import { type Border } from '../border';
import { type BoxShadow } from '../box-shadow';
import { type ButtonIcon } from '../button-icon';
import { type Font } from '../font';
import { type Spacing } from '../spacing';


export namespace Button {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: {
      decoration?: Element.Decoration.Attributes;
    };
    showButtonStyleFields?: boolean;
    fields?: {
      enable?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Button.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      backgroundGroup?: Field.BaseProps & {
        component?: {
          type?: 'group';
          name?: 'divi/background';
          props?: Background.GroupProps;
        }
      };
      borderGroup?: Field.BaseProps & {
        component?: {
          type?: 'group';
          name?: 'divi/border';
          props?: Border.GroupProps;
        }
      };
      fontGroup?: Field.BaseProps & {
        component?: {
          type?: 'group';
          name?: 'divi/font';
          props?: Font.GroupProps;
        }
      };
      buttonIconGroup?: Field.BaseProps & {
        component?: {
          type?: 'group';
          name?: 'divi/button-icon';
          props?: ButtonIcon.GroupProps;
        }
      };
      spacingGroup?: Field.BaseProps & {
        component?: {
          type?: 'group';
          name?: 'divi/spacing';
          props?: Spacing.GroupProps;
        }
      };
      boxShadowGroup?: Field.BaseProps & {
        component?: {
          type?: 'group';
          name?: 'divi/box-shadow';
          props?: BoxShadow.GroupProps;
        }
      };
      alignment?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Button.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      }
    };
  }
}
