// @divi/types/src/module/element/decoration/box-shadow/index.ts
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';


export namespace BoxShadow {
  export interface AttributeValue {
    style?: 'none' | 'preset1' | 'preset2' | 'preset3' | 'preset4' | 'preset5' | 'preset6' | 'preset7';
    horizontal?: string;
    vertical?: string;
    blur?: string;
    spread?: string;
    color?: string;
    position?: 'outer' | 'inner';
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Style {

    // TODO replace `BoxShadowStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.BoxShadow>;
      important?: boolean | FormatStyleImportant<Css.Properties.BoxShadow>;
      useOverlay?: boolean;
    }
  }

  export namespace Component {

    // TODO replace `BoxShadowComponentProps` with this.
    export interface Props {
      attr: Attributes;
      id: string;
      settings?: {
        overlay?: boolean;
      } | boolean;
    }

  }
}
