// @divi/types/src/field-library/conditions/index.ts
import { type FormEvent, type FunctionComponent, type ReactNode } from 'react';


export namespace Conditions {

  /**
   * @typedef {object} Item
   * @property {string} condition
   * @property {string} field Field.
   * @property {string} [value] Value.
   */
  export interface Item {
    condition: string;
    field: string;
    value?: string;
  }

  /**
   * @typedef {object} Field
   * @property {string|ReactNode|FunctionComponent} label Label.
   * @property {string[]} values Values.
   * @property {string} [type] Type.
   */
  export interface Field {
    label: string | ReactNode | FunctionComponent;
    values: string[];
    type?: string;
  }

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {Event} [event] Event.
   * @property {object} [inputValue] Input value.
   * @property {Item[]} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: Item[];
  }

  /**
   * @typedef {object} Props
   * @property {string} [className] Class name.
   * @property {string} name Name.
   * @property {object} value Value.
   * @property {object} fields Fields.
   * @property {boolean} ["readonly"] Read only.
   * @property {boolean} [showDeleteButton] Show delete button.
   * @property {boolean} [showAddButton] Show add button.
   * @property {Function} onChange On change callback.
   */
  export interface Props {
    className?: string;
    name: string;
    value: Item[];
    fields: Record<string, Field>;
    readonly?: boolean;
    showDeleteButton?: boolean;
    showAddButton?: boolean;
    onChange: (value: OnChangeCallbackParams) => void;
  }
}
