// @divi/types/src/clipboard/item/index.ts
import { type EditPost } from '../../edit-post';
import {
  type ModuleAttr,
  type ModuleAttrs,
} from '../../module/library';
import { type Attrs } from '../attrs';


export namespace Item {
  export interface Base {

    // Module / element where this item gets added to clipboard.
    origin: string;
  }

  export interface Module extends Base {
    clipboardType: 'module';
    payload: EditPost.Clipboard.Item.Payload.Module;
  }

  export interface Modules extends Base {
    clipboardType: 'modules';
    payload: EditPost.Clipboard.Item.Payload.Modules;
  }

  export interface ModuleFieldAttribute extends Base {
    clipboardType: 'moduleFieldAttribute';
    payload: {
      attrName: string;
      subName?: string;
      moduleId: string;
      modalName: string;
      fieldName: string;
      fieldOptions?: Record<string, any>;
      fieldLabel: string;
      fieldAttr: ModuleAttr;
      groupName: string;
    }
  }

  export interface ModuleGroupAttribute extends Base {
    clipboardType: 'moduleGroupAttributes';
    payload: {
      moduleId: string;
      modalName: string;
      modalTab: string;
      groupId: string;
      groupLabel: string;
      groupName: string;
      attrsGroupNameMap: Attrs.GroupNameMap;
      attrsMap: Attrs.CategoryMapItem;
      attrs: ModuleAttrs;
    }
  }

  export interface ModuleGroupTabAttribute extends Base {
    clipboardType: 'moduleGroupTabAttributes';
    payload: {
      moduleId: string;
      modalName: string;
      groupId: string;
      groupTabLabel: string;
      groupTabName: string;
      attrsMap: Attrs.CategoryMapItem;
      attrs: ModuleAttrs;
      attrsGroupNameMap: Attrs.GroupNameMap;
    }
  }

  export interface ModuleAttributes extends Base {
    clipboardType: 'moduleAttributes';
    payload: {
      moduleId: string;
      moduleLabel: string;
      attrsMap: Attrs.CategoryMapItem;
      attrs: ModuleAttrs;
    }
  }

  export type All =
    Module |
    Modules |
    ModuleAttributes |
    ModuleGroupAttribute |
    ModuleGroupTabAttribute |
    ModuleFieldAttribute;

  export type Type = All['clipboardType'];
}
