// @divi/types/src/module/library/fullwidth-slider/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface FullwidthSliderCssAttr extends Css.AttributeValue {
  slideDescription?: string;
  slideTitle?: string;
  slideButton?: string;
  slideControllers?: string;
  slideActiveController?: string;
  slideImage?: string;
  slideArrows?: string;
}

export type FullwidthSliderCssGroupAttr = FormatBreakpointStateAttr<FullwidthSliderCssAttr>;

export interface FullwidthSliderOverlay {
  useOverlay?: string;
  color?: string;
}
export interface FullwidthSliderTextOverlay extends FullwidthSliderOverlay {
  radius?: string;
}

export interface FullwidthSliderAttrs extends InternalAttrs {

  // Button
  button?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'border' |
      'boxShadow' |
      'button' |
      'font' |
      'spacing'
    >;
  };

  css?: FullwidthSliderCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      auto?: FormatBreakpointStateAttr<OnOff>;
      autoSpeed?: FormatBreakpointStateAttr<string>;
      autoIgnoreHover?: FormatBreakpointStateAttr<OnOff>;
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'spacing' |
      'sizing' |
      'filters' |
      'transform' |
      'overflow' |
      'zIndex' |
      'transition' |
      'boxShadow' |
      'border' |
      'disabledOn' |
      'position' |
      'sticky' |
      'scroll'
    >;
  };

  image?: {
    advanced?: {
      showOnMobile?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<'border' | 'boxShadow'>;
  };

  title?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  content?: {
    decoration?: Element.Decoration.PickedAttributes<'bodyFont' | 'sizing'>;
  };

  arrows?: {
    advanced?: {
      show?: FormatBreakpointStateAttr<OnOff>;
      color?: FormatBreakpointStateAttr<string>;
    };
  };

  dotNav?: {
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  };

  pagination?: {
    advanced?: {
      show?: FormatBreakpointStateAttr<OnOff>;
    };
  };

  children?: {
    module?: {
      decoration?: Element.Decoration.PickedAttributes<'background'>;
    };
    slideOverlay?: {
      advanced?: {
        use?: FormatBreakpointStateAttr<OnOff>;
      };
      decoration?: Element.Decoration.PickedAttributes<'background'>;
    };
    contentOverlay?: {
      advanced?: {
        use?: FormatBreakpointStateAttr<OnOff>;
      };
      decoration?: Element.Decoration.PickedAttributes<'background' | 'border'>;
    };
    button?: {
      advanced?: {
        showOnMobile?: FormatBreakpointStateAttr<OnOff>;
      };
      innerContent?: Element.Types.Button.InnerContent.Attributes;
    };
    content?: {
      advanced?: {
        showOnMobile?: FormatBreakpointStateAttr<OnOff>;
      };
    };
  };
}
