// @divi/types/src/module/element/decoration/spacing/index.ts
import { type OnOff } from '../../../../attr';
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';


export namespace Spacing {
  export interface SidesVertical {
    top?: string;
    bottom?: string;
  }

  export interface SidesHorizontal {
    right?: string;
    left?: string;
  }

  export interface Syncs {
    syncHorizontal?: OnOff;
    syncVertical?: OnOff;
  }

  export type Sides = SidesHorizontal & SidesVertical;

  export type SyncNames = keyof Syncs;

  export type ItemValue = Sides & Syncs;

  export interface AttributeValue {
    margin?: ItemValue;
    padding?: ItemValue;
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Style {

    // TODO Replace `SpacingStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.Spacing>;
      important?: boolean | FormatStyleImportant<Css.Properties.Spacing>;
    }
  }
}
