// @divi/types/src/events/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Module } from '../../module';
import { type ModuleType } from '../../module/category';
import { type Cursor } from '../cursor';
import { type Dragged } from '../dragged';
import { type Hovered } from '../hovered';
import { type Selected } from '../selected';


export namespace Store {
  export interface State {
    parentHover: Record<string, string>;
    parentSelect: Record<string, string>;
    parentSelectMap: Record<string, string[]>;
    currentHover: Hovered.Module;
    cursorPosition: Cursor.Position;
    isDragging: boolean;
    dragLocation: Dragged.DragLocation;
    draggedModules: Dragged.Modules;
    draggedModuleDropZone: null | Module.DropZone.Props;
    selectedModules: Selected.Modules;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  };

  export namespace Log {
    export type Payload = {
      target: string,

      // @todo (D5) need to circle back and standardized this
      eventPayload: Record<string, unknown>,
      state?: {
        currentHover: ImmutableState['currentHover'],
        parentHover: ImmutableState['parentHover'],
      },
    }
  }

  export namespace Actions {
    export namespace ClearSelection {
      export type ReturnValue = {
        type: 'CLEAR_SELECTION';
      }
    }

    export namespace Deselect {
      export type ReturnValue = {
        type: 'DESELECT';
        id: string;
      }
    }

    export namespace EndDrag {
      export type ReturnValue = {
        type: 'END_DRAG';
      }
    }

    export namespace OnMouseEnter {
      export type Payload = {
        label?: string;
        anchestorId?: string;
        parentId?: string;
        anchestorHoveredArea?: 'top' | 'bottom' | 'inside';
      }

      export type ReturnValue = {
        type: 'MOUSE_ENTER';
        target: string;
        id: string;
        payload: Payload;
        area: Hovered.Area;
      }
    }

    export namespace OnMouseLeave {
      export interface ReturnValue {
        type: 'MOUSE_LEAVE';
        target: string;
        id: string;
        payload: OnMouseEnter.Payload;
      }
    }

    export namespace Select {
      export interface ReturnValue {
        type: 'SELECT';
        id: string;
        moduleType: ModuleType;
      }
    }

    export namespace SetCursorPosition {
      export interface ReturnValue {
        type: 'SET_CURSOR_POSITION';
        pageX: number;
        pageY: number;
        clientX: number;
        clientY: number;
      }
    }

    export namespace SetDraggedModuleDropZone {
      export interface ReturnValue {
        type: 'SET_DRAGGED_MODULE_DROP_ZONE';
        props: null | Module.DropZone.Props;
      }
    }

    export namespace SetHoveredArea {
      export interface ReturnValue {
        type: 'SET_HOVERED_AREA';
        area: Hovered.Area;
      }
    }

    export namespace StartDrag {
      export interface ReturnValue {
        type: 'START_DRAG';
        items: Dragged.Modules;
        dragLocation: Dragged.DragLocation;
      }
    }

    export type FunctionMap = {
      clearSelection: () => ClearSelection.ReturnValue,
      deselect: (id: string) => Deselect.ReturnValue,
      endDrag: () => EndDrag.ReturnValue,
      onMouseEnter: (
        target: string,
        id: string,
        payload: OnMouseEnter.Payload,
        area: Hovered.Area,
      ) => OnMouseEnter.ReturnValue,
      onMouseLeave: (
        target: string,
        id: string,
        payload: OnMouseEnter.Payload,
      ) => OnMouseLeave.ReturnValue,
      select: (
        id: string,
        moduleType: ModuleType,
      ) => Select.ReturnValue,
      setCursorPosition: (
        pageX: number,
        pageY: number,
        clientX: number,
        clientY: number,
      ) => SetCursorPosition.ReturnValue,
      setDraggedModuleDropZone: (props: null | Module.DropZone.Props) => SetDraggedModuleDropZone.ReturnValue,
      setHoveredArea: (area: Hovered.Area) => SetHoveredArea.ReturnValue,
      startDrag: (items: Dragged.Modules, dragLocation: Dragged.DragLocation) => StartDrag.ReturnValue,
    };

    export type ReturnValue =
      ClearSelection.ReturnValue |
      Deselect.ReturnValue |
      EndDrag.ReturnValue |
      OnMouseEnter.ReturnValue |
      OnMouseLeave.ReturnValue |
      Select.ReturnValue |
      SetCursorPosition.ReturnValue |
      SetDraggedModuleDropZone.ReturnValue |
      SetHoveredArea.ReturnValue |
      StartDrag.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getHoveredModule: () => Hovered.Module;
      isHovered: (id: string) => boolean;
      isChildHovered: (id: string) => boolean;
      isChildSelected: (id: string) => boolean;
      isDragged: (id: string, dragLocation: Dragged.DragLocation) => boolean;
      getDraggingStatus: (dragLocation: Dragged.DragLocation) => boolean;
      getDraggedModule: (id: string) => Dragged.Module;
      getDraggedModules: () => ImmutableState['draggedModules'];
      getCursorPosition: () => Cursor.Position;
      getDraggedModuleDropZone: () => null | Module.DropZone.Props;
      getSelectedModules: () => Store.ImmutableState['selectedModules'];
      isSelectedModule: (id: string) => boolean;
    }
  }

  export namespace Effects {
    export type FunctionMap = {
      // eslint-disable-next-line @typescript-eslint/naming-convention
      MOUSE_ENTER: (action: Actions.ReturnValue, store: Instance) => void;
      // eslint-disable-next-line @typescript-eslint/naming-convention
      MOUSE_LEAVE: (action: Actions.ReturnValue, store: Instance) => void;
      // eslint-disable-next-line @typescript-eslint/naming-convention
      SET_CURSOR_POSITION: (action: Actions.SetCursorPosition.ReturnValue, store: Instance) => void;
      // eslint-disable-next-line @typescript-eslint/naming-convention
      END_DRAG: () => void;
      // eslint-disable-next-line @typescript-eslint/naming-convention
      SELECT: (action: Actions.Select.ReturnValue) => void;
    }
  }
}
