// @divi/types/src/field-library/border-radius/index.ts
import { type OnOff } from '../../attr';


export namespace BorderRadius {

  /**
   * @typedef {object} BorderRadius.InputValue
   * @property {('topLeft'|'topRight'|'bottomLeft'|'bottomRight')} side Side.
   * @property {('on'|'off')} sync Sync.
   * @property {string} value Value.
   */
  export interface InputValue {
    side: keyof Omit<Value, 'sync'>;
    sync: Value['sync']
    value: string;
  }

  /**
   * @typedef {object} BorderRadius.Value
   * @property {string} [topLeft] Top left.
   * @property {string} [topRight] Top right.
   * @property {string} [bottomLeft] Bottom left.
   * @property {string} [bottomRight] Bottom right.
   * @property {('on'|'off')} [sync] Sync.
   */
  export interface Value {
    topLeft?: string;
    topRight?: string;
    bottomLeft?: string;
    bottomRight?: string;
    sync?: OnOff;
  }

  /**
   * @typedef {object} BorderRadius.Steps
   * @property {number} [topLeft] Top left.
   * @property {number} [topRight] Top right.
   * @property {number} [bottomLeft] Bottom left.
   * @property {number} [bottomRight] Bottom right.
   * @property {number} [sync] Sync.
   */
  export type Steps = Record<keyof Omit<Value, 'sync'>, number>;


  /**
   * @typedef {object} BorderRadius.OnChangeCallbackParams
   * @property {object} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    inputValue: Record<string, string>;
  }


  /**
   * @typedef {object} BorderRadius.Props
   * @property {string} name Name.
   * @property {string} [className] Class name.
   * @property {object} value Value.
   * @property {object} [defaultValue] Default value.
   * @property {Function} onChange On change.
   * @property {string[]} [allowedUnits] Allowed units.
   */
  export interface Props {
    name: string;
    className?: string;
    value: Value;
    defaultValue?: Value;
    onChange: (params: OnChangeCallbackParams) => void;
    allowedUnits?: string[];
  }
}
