// @divi/types/src/field-library/button-options/index.ts
import { type MouseEvent } from 'react';


export namespace ButtonOptions {

  /**
   * @typedef {string|string[]} MultiOptionsAttr
   */
  export type MultiOptionsAttr = string | string[];

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {MouseEvent} event Event.
   */
  export interface OnChangeCallbackParams {
    event: MouseEvent;
  }

  /**
   * @typedef {object} ButtonComponentProps
   * @property {string} [className] Class name.
   * @property {string} [icon] Icon.
   * @property {string} [label] Label.
   * @property {string} [tooltip] Tooltip.
   */
  export interface ButtonComponentProps {
    className?: string,
    icon?: string,
    label?: string,
    tooltip?: string,
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} [className] Class name.
   * @property {Function} onChange On change callback.
   * @property {object} options Options.
   * @property {string} [options.className] Class name.
   * @property {boolean} [multipleChoices] Multiple choices.
   * @property {string|string[]} [value] Value.
   * @property {number} [column] Column.
   * @property {boolean} [showLabel] Show label.
   * @property {boolean} [fullwidth] Full width.
   * @property {boolean} [tabUi] Tab UI.
   * @property {string|string[]} [defaultValue] Default value.
   * @property {boolean} [renderEmptyStringAsDefaultValue] Render empty string as default value.
   * @property {boolean} [allowClickToUnset] Allow click to unset.
   */
  export interface Props {
    name: string;
    className?: string;
    onChange: (params: OnChangeCallbackParams) => void;
    options: Record<string, ButtonComponentProps>;
    optionsClassName?: string;
    multipleChoices: boolean;
    value: MultiOptionsAttr;
    column?: number;
    showLabel?: boolean;
    fullwidth?: boolean;
    tabUi?: boolean;
    defaultValue?: MultiOptionsAttr;
    renderEmptyStringAsDefaultValue?: boolean;
    allowClickToUnset?: boolean;
  }
}
