// @divi/types/src/settings/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type BlockConfiguration } from '@wordpress/blocks';

import {
  type EmailMarketing,
  type GlobalData,
  type SpamProtection,
} from '@divi/types';
import { type PostStatus } from '@divi/types-external';

import { type AppUi } from '../../app-ui';
import { type Fonts } from '../../fonts';
import { type DynamicContentOptions } from '../../module/dynamic-content';
import { type ShortcodeModule } from '../../shortcode-module';
import {
  type PostObject,
  type Taxonomy,
} from '../../utils';


export namespace Store {

  /**
   * Interface of Acceptable CSS String Values.
   *
   * @since ??
   */
  export interface AcceptableCSSStringValues {
    'font-size'?: (string)[] | null;
    'line-height'?: (string)[] | null;
    'max-height'?: (string)[] | null;
    'max-width'?: (string)[] | null;
    'min-height'?: (string)[] | null;
    'z-index'?: (string)[] | null;
    height?: (string)[] | null;
    margin?: (string)[] | null;
    padding?: (string)[] | null;
    width?: (string)[] | null;
  }

  export interface Category {
    id: number;
    label: string;
    permalink: string;
  }

  export interface ColumnLayouts {
    regular: string[];
    columnInner: string[];
    columnInnerWide: string[];
    specialty: Record<number, { position: string; structure: string; }>
  }

  /**
   * Interface of Conditional Tags.
   *
   * @since ??
   */
  export interface ConditionalTags {
    /* eslint-disable @typescript-eslint/naming-convention */
    is_limited_mode: boolean;
    is_bfb: boolean;
    is_bfb_activated: boolean;
    is_tb: boolean;
    is_front_page: boolean;
    is_home_page: boolean;
    is_search: boolean;
    is_single: boolean;
    is_singular: boolean;
    is_singular_project: boolean;
    is_rtl: boolean;
    is_no_rtl: boolean;
    et_is_builder_plugin_active: boolean;
    is_user_logged_in: boolean;
    et_is_ab_testing_active: string;
    is_wrapped_styles: boolean;
    is_gutenberg: boolean;
    is_custom_post_type: boolean;
    is_layout_post_type: boolean;
    is_rich_editor: string;
    is_layout_block: string;
    /* eslint-enable @typescript-eslint/naming-convention */
  }

  export type GlobalLayouts = Record<number | string, PostObject>;

  export type PageSettings = {
    /* eslint-disable @typescript-eslint/naming-convention */
    et_pb_ab_bounce_rate_limit: string;
    et_pb_ab_current_shortcode: string;
    et_pb_ab_stats_refresh_interval: string;
    et_pb_ab_subjects: string[];
    et_pb_color_palette: string;
    et_pb_content_area_background_color: string;
    et_pb_custom_css: string;
    et_pb_dark_text_color: string;
    et_pb_enable_ab_testing: string;
    et_pb_enable_shortcode_tracking: string;
    et_pb_gc_palette: GlobalData.Store.State['globalColors'];
    et_pb_light_text_color: string;
    'et_pb_overflow-x': string;
    'et_pb_overflow-y': string;
    et_pb_page_gutter_width: string;
    et_pb_page_z_index: string;
    et_pb_post_settings_categories: string;
    et_pb_post_settings_excerpt: string;
    et_pb_post_settings_image: string;
    et_pb_post_settings_project_categories: string;
    et_pb_post_settings_project_tags: string;
    et_pb_post_settings_tags: string;
    et_pb_post_settings_title: string;
    et_pb_section_background_color: string;
    et_pb_static_css_file: string;
    global_colors_info: string;
    /* eslint-enable @typescript-eslint/naming-convention */
  };

  export interface PostData {
    id: string;
    title: string;
    content: string;
    type: string;
    status: PostStatus;
    url: string;
  }

  export type RoleSettings = {
    [key: string]: Record<string, string>;
  }

  export interface SettingsDataCurrentPage {
    url?: string;
    permalink?: string;
    backendBuilderUrl?: string;
    id?: number;
    title?: string;
    thumbnailUrl?: string;
    thumbnailId?: number;
    authorName?: string;
    authorUrl?: string;
    authorUrlTitle?: string;
    date?: number;
    categories?: Record<string, Category>;
    commentsPopup?: string;
    commentsCount?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    comments_popup_tb?: string;
    paged?: number;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    post_modified?: Date;
    lang?: string;
    blockId?: string;
    langCode?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    page_layout?: string;
    woocommerceTabs?: WooCommerceTabs;
    woocommerce?: WooCommerce;
  }

  export interface SettingsDataThemeBuilder {
    isLayout?: boolean;
    layoutPostTypes?: string[];
    bodyLayoutPostType?: string;
    postContentModules?: string[];
    hasValidBodyLayout?: boolean;
    themeBuilderAreas?: string[];
  }

  export interface WooCommerce {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    inactive_module_notice: string;
  }

  export interface WooCommerceTabs {
    description?: WooCommerceTabsDescription;
    reviews?: WooCommerceTabsDescription;
  }

  export interface WooCommerceTabsDescription {
    value?: string;
    label?: string;
  }

  export type StoreStateItem = string | number | string[] | Record<string, Category> | ColumnLayouts;

  export interface State {
    urls: {
      admin: string;
      ajax: string;
      adminOptionsGeneralUrl: string;
      currentPageUrl: string;
      forgotPasswordUrl: string;
      homeUrl: string;
      loginFormUrl: string;
    };
    acceptableCSSStringValues: AcceptableCSSStringValues;
    allFontFormats?: (string)[] | null;
    blocks: Record<string, BlockConfiguration>;
    cookiePath: string;
    columnLayouts: ColumnLayouts;
    conditionalTags: ConditionalTags;
    customFonts?: Fonts.Custom.Fonts;
    currentUserRole: string;
    dynamicContent?: {
      options?: DynamicContentOptions;
    }
    fontIcons: string[];
    fontIconsDown: string[];
    globalLayouts: GlobalLayouts;
    googleAPIKey: string;
    googleMapsScriptNotice: boolean;
    googleFonts: Fonts.Google.Fonts;
    layoutCategory: Array<Taxonomy>;
    layoutTag: Array<Taxonomy>;
    layoutType: string;
    nonces: Record<string, Record<string, string>>;
    pageSettingsValues: {
      // eslint-disable-next-line @typescript-eslint/naming-convention
      post_content: PageSettings;
    };
    post: PostData;
    preferences: {
      appColorMode: AppUi.ColorMode;
      appColorScheme: AppUi.ColorScheme;
      toolbarClick: boolean;
      toolbarDesktop: boolean;
      toolbarGrid: boolean;
      toolbarHover: boolean;
      toolbarPhone: boolean;
      toolbarTablet: boolean;
      toolbarWireframe: boolean;
      toolbarZoom: boolean;
      builderAnimation: boolean;
      builderEnableDummyContent: boolean;
      hideDisabledModules: boolean;
      eventMode: AppUi.Mode.Type;
      viewMode: AppUi.View.Type;
      historyIntervals: number;
      pageCreationFlow: string;
      quickActionsAlwaysStartWith: string;
      quickActionsShowRecentQueries: string;
      quickActionsRecentQueries: string;
      quickActionsRecentCategory: string;
      modalPreference: string;
      builderDisplayModalSettings: boolean;
    };
    removedFonts: Fonts.Removed.Fonts;
    roleSettings: RoleSettings;
    shortcodeModuleDefinitions: Record<string, ShortcodeModule.Module.Definition>;
    shortcodeTags: string[];
    styles: {
      body: {
        fontHeight: string;
        fontSize: string;
      }
    }
    themeBuilder?: SettingsDataThemeBuilder;
    tinymce: {
      skinUrl: string;
      cssFiles: string;
      plugins: string[];
    };
    version: string;
    siteSettings: {
      gmtOffsetString?: string;
    }
    themeSettings?: {
      widgetAreas?: Record<string, { label: string; }>;
      defaultArea?: string;
    };
    currentPage?: SettingsDataCurrentPage;
    commentsModuleMarkup?: string;
    emailServices?: Record<string, EmailMarketing.Service.Provider>;
    spamProtectionServices?: Record<string, SpamProtection.Service.Provider>;
    navMenusOptions?: Record<string, string>;
    projectCategories?: Array<Taxonomy>;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  };

  export namespace Actions {
    export namespace Add {
      export type ReturnValue<T1, T2> = {
        type: 'ADD';
        settingKey: T1;
        payload: T2;
      }
    }

    export type FunctionMap = {

      // add<T1 extends keyof ImmutableState> (settingKey: T1, payload: ImmutableState[T1]) => Add.ReturnValue,
      add: <T1 extends keyof ImmutableState>(
        settingKey: T1,
        payload: ImmutableState[T1]
      ) => Add.ReturnValue<T1, ImmutableState[T1]>
    };

    export type ReturnValue = Add.ReturnValue<keyof ImmutableState, ImmutableState[keyof ImmutableState]>;
  };

  export namespace Selectors {

    /**
     * Interface for getSetting function overloading.
     *
     * @since ??
     */
    export interface GetSetting<T = ImmutableState> {
      // eslint-disable-next-line max-len
      <T1 extends keyof T, T2 extends keyof T[T1], T3 extends keyof T[T1][T2], T4 extends keyof T[T1][T2][T3]>(settingPath: [T1, T2, T3, T4], defaultValue?: T[T1][T2][T3][T4]): T[T1][T2][T3][T4];
      // eslint-disable-next-line max-len
      <T1 extends keyof T, T2 extends keyof T[T1], T3 extends keyof T[T1][T2]>(settingPath: [T1, T2, T3], defaultValue?: T[T1][T2][T3]): T[T1][T2][T3];
      <T1 extends keyof T, T2 extends keyof T[T1]>(settingPath: [T1, T2], defaultValue?: T[T1][T2]): T[T1][T2];
      <T1 extends keyof T>(settingPath: [T1], defaultValue?: T[T1]): T[T1];
      <T1 extends keyof T>(settingPath: T1, defaultValue?: T[T1]): T[T1];
      (settingPath: string | string[], defaultValue?: StoreStateItem): ImmutableObject<StoreStateItem>;
    }

    export type FunctionMap = {
      checkRolePermission: (
        permission: string,
      ) => boolean,
      getSetting: GetSetting,
    }
  }
}
