// @divi/types/src/field-library/select-font/index.ts
import { type FormEvent } from 'react';

import { type KeyboardShortcuts } from '../../keyboard-shortcuts';


export namespace SelectFont {

  /**
   * @typedef {object} OpenModalParams
   * @property {string} name Name.
   * @property {boolean} centered Centered.
   */
  export interface OpenModalParams {
    name: 'divi/upload-font';
    centered: boolean;
  }

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {Event} [event] Form Event.
   * @property {string} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: string;
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} [className] Class name.
   * @property {string} [defaultValue] Default value.
   * @property {string} value Value.
   * @property {string[]} googleFonts Google fonts.
   * @property {object} customFonts Custom fonts.
   * @property {Function} onChange On change callback.
   * @property {Function} onDeleteFont On delete font callback.
   * @property {Function} onHoverFont On hover font callback.
   * @property {Function} openModal Open modal callback.
   * @property {Function} [registerShortcuts] Register shortcuts callback.
   * @property {Function} [unregisterShortcuts] Unregister shortcuts callback.
   */
  export interface Props {
    name: string;
    className?: string;
    defaultValue?: string;
    value: string;
    googleFonts: string[],
    customFonts: Record<string, unknown>;
    onChange: (params: OnChangeCallbackParams) => void;
    onDeleteFont: (fontName: string) => void;
    onHoverFont: (fontName: string) => void;
    openModal: (params: OpenModalParams) => void;
    registerShortcuts?: (shortcuts: KeyboardShortcuts.Shortcut.Item[]) => void,
    unregisterShortcuts?: (names: string[]) => void;
  }
}
