// @divi/types/src/module/library/fullwidth-menu/index.ts
/* eslint-disable @typescript-eslint/naming-convention */
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface FullwidthMenuAttrs extends InternalAttrs {

    // CSS options is used across multiple elements inside the module thus it deserves its own top property.
    css?: Css.Attributes;

    module?: {
        meta?: Element.Meta.Attributes;
        advanced?: {
            htmlAttributes?: Element.Advanced.IdClasses.Attributes;
            link?: Element.Advanced.Link.Attributes;
            text?: Element.Advanced.Text.Attributes;
        };
        decoration?: Element.Decoration.Attributes;
    };

    logo?: {
        innerContent?: Element.Types.ImageLink.Attributes['innerContent'];
        decoration?: Element.Decoration.PickedAttributes<
        'sizing' |
        'filters' |
        'boxShadow' |
        'border'>;
    }

    menu?: {
        advanced?: {
            menuId?: FormatBreakpointStateAttr<string>;
            style?: FormatBreakpointStateAttr<
            'left_aligned' |
            'centered' |
            'inline_centered_logo'>;
            activeLinkColor?: FormatBreakpointStateAttr<string>;
            fullwidth?: FormatBreakpointStateAttr<OnOff>;
        };
        decoration?: Element.Decoration.PickedAttributes<
        'font' |
        'bodyFont'>;
    };

    menuDropdown?: {
        advanced?: {
            direction?: FormatBreakpointStateAttr<
            'downwards' |
            'upwards'>;
            activeLinkColor?: FormatBreakpointStateAttr<string>;
            lineColor?: FormatBreakpointStateAttr<string>;
            animation?: FormatBreakpointStateAttr<string>;
        };
        decoration?: Element.Decoration.PickedAttributes<
        'font' |
        'background'>;
    };

    menuMobile?: {
        decoration?: Element.Decoration.PickedAttributes<
        'font' |
        'background'>;
    };

    cartQuantity?: {
        advanced?: {
            show?: FormatBreakpointStateAttr<OnOff>;
        };

        decoration?: Element.Decoration.PickedAttributes<'font'>;
    };

    cartIcon?: {
        advanced?: {
            show?: FormatBreakpointStateAttr<OnOff>;
        };

        decoration?: Element.Decoration.PickedAttributes<'font'>;
    };

    searchIcon?: {
        advanced?: {
            show?: FormatBreakpointStateAttr<OnOff>;
        };

        decoration?: Element.Decoration.PickedAttributes<'font'>;
    };

    hamburgerMenuIcon?: {
        decoration?: Element.Decoration.PickedAttributes<'font'>;
    };
}
