// @divi/types/src/clipboard/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Attrs } from '../attrs';
import { type Component } from '../component';
import { type Item } from '../item';


export namespace Store {
  export interface State {
    items: Item.All[];
    componentAttrsMaps: Record<string, Component.AttrsMaps.Item>;
    componentAttrsMapsActive: string;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  };

  export namespace Actions {
    export namespace AddComponentAttrsMaps {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        payload: Component.AttrsMaps.Item;
        type: 'ADD_COMPONENT_ATTRS_MAPS';
      }
    }

    export namespace AddItem {
      export type Params = Omit<ReturnValue, 'type'>;

      export interface ReturnValue {
        type: 'ADD_ITEM';
        clipboardType: Item.All['clipboardType'];
        origin: Item.All['origin'];
        payload: Item.All['payload'];
      }
    }

    export namespace ResetState {
      export type ReturnValue = {
        type: 'RESET_STATE';
      }
    }

    export namespace SetComponentAttrsMapsValue {
      export type Params = Omit<ReturnValue, 'type'>;

      export interface ReturnValue {
        type: 'SET_COMPONENT_ATTRS_MAPS_VALUE';
        id: string;
        attrsMaps: Attrs.Maps;
        resetActiveComponent: boolean;
      }
    }

    export type FunctionMap = {
      addComponentAttrsMaps: (params: AddComponentAttrsMaps.Params) => AddComponentAttrsMaps.ReturnValue;
      addItem: (params: AddItem.Params) => AddItem.ReturnValue;
      resetState: () => ResetState.ReturnValue;
      setComponentAttrsMapsValue: (params: SetComponentAttrsMapsValue.Params) => SetComponentAttrsMapsValue.ReturnValue;
    };

    export type ReturnValue =
      AddComponentAttrsMaps.ReturnValue |
      AddItem.ReturnValue |
      ResetState.ReturnValue |
      SetComponentAttrsMapsValue.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getItems: () => Item.All[];
      getLastItem: (clipboardType: Item.Type | Item.Type[] | 'all') => ImmutableObject<Item.All>;
      getComponentAttrsMaps: (id: string) => ImmutableObject<Component.AttrsMaps.Item>;
      getActiveComponentAttrsMaps: () => string;
    }
  }
}
