// @divi/types/src/keyboard-shortcuts/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Shortcut } from '../shortcut';


export namespace Store {
  export interface State {
    enableListeners: boolean;
    pressedKeys: Record<string, string>;
    pressedWithMeta: Record<string, string>;
    meta: boolean;
    do: Shortcut.Item;
    registry: Record<string, Shortcut.Item>;
    keymap: Record<string, string[]>;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace DoShortcut {
      export type ReturnValue = {
        type: 'DO';
        shortcut: ImmutableObject<Shortcut.Item>;
      }
    }

    export namespace KeyDown {
      export type ReturnValue = {
        type: 'KEYDOWN';
        physicalKeyCode: string;
        key: string;
      }
    }

    export namespace KeyUp {
      export type ReturnValue = {
        type: 'KEYUP';
        physicalKeyCode: string;
        key: string;
      }
    }

    export namespace RegisterShortcut {
      export type ReturnValue = {
        type: 'REGISTER';
        shortcut: Shortcut.Item;
      }
    }

    export namespace ResetDoShortcut {
      export type ReturnValue = {
        type: 'RESET_DO';
      }
    }

    export namespace ResetState {
      export type ReturnValue = {
        type: 'RESET_STATE';
      }
    }

    export namespace ToggleShortcutListeners {
      export type ReturnValue = {
        type: 'TOGGLE_LISTENERS';
        enableListeners: boolean;
      }
    }

    export namespace UnregisterShortcut {
      export type ReturnValue = {
        type: 'UNREGISTER';
        name: string;
      }
    }

    export namespace WindowBlur {
      export type ReturnValue = {
        type: 'WINDOW_BLUR';
      }
    }

    export type FunctionMap = {
      doShortcut: (shortcut: DoShortcut.ReturnValue['shortcut']) => DoShortcut.ReturnValue;
      keyDown: (physicalKeyCode: string, key: string) => KeyDown.ReturnValue;
      keyUp: (physicalKeyCode: string, key: string) => KeyUp.ReturnValue;
      registerShortcut: (shortcut: Shortcut.Item) => RegisterShortcut.ReturnValue;
      resetDoShortcut: () => ResetDoShortcut.ReturnValue;
      resetState: () => ResetState.ReturnValue;
      toggleShortcutListeners: (enableListeners: ToggleShortcutListeners.ReturnValue['enableListeners']) => ToggleShortcutListeners.ReturnValue;
      unregisterShortcut: (name: UnregisterShortcut.ReturnValue['name']) => UnregisterShortcut.ReturnValue;
      windowBlur: () => WindowBlur.ReturnValue;
    };


    export type ReturnValue =
      DoShortcut.ReturnValue |
      KeyDown.ReturnValue |
      KeyUp.ReturnValue |
      RegisterShortcut.ReturnValue |
      ResetDoShortcut.ReturnValue |
      ResetState.ReturnValue |
      ToggleShortcutListeners.ReturnValue |
      UnregisterShortcut.ReturnValue |
      WindowBlur.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getPressedKeys: () => ImmutableState['pressedKeys'];
      getMatchingShortcut: (keys: string[]) => false | ImmutableObject<Shortcut.Item>;
      getCurrentShortcut: () => ImmutableState['do'];
      getShortcuts: (groupBy?: Shortcut.Group) => ImmutableObject<any> | Shortcut.Items;
    }
  }
}
