// @divi/types/src/field-library/email-service-account/index.ts
import { type FormEvent } from 'react';
import { type ImmutableObject } from 'seamless-immutable';

import {
  type EmailMarketing,
} from '../../email-marketing';
import {
  type AttrState,
  type Breakpoint,
} from '../../format-attr';
import {
  type EmailServiceGroupAttr,
} from '../../module/options/email-service';


export namespace EmailServiceAccount {

  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: string;
  }

  export interface OnRemoveCallbackParams {
    provider:string;
    account:string;
  }

  export interface OnAddCallbackParams extends OnRemoveCallbackParams {
    data:Record<string, string>;
  }

  export type OnFetchCallbackParams =  OnRemoveCallbackParams;

  export type OnChangeCallback = (params: OnChangeCallbackParams) => void;
  export type OnRemoveCallback = (params:OnRemoveCallbackParams) => void;
  export type OnAddCallback = (params:OnAddCallbackParams) => void;
  export type OnFetchCallback = (params:OnFetchCallbackParams) => void;

  export interface ContainerProps {
    appBreakpoint?: Breakpoint;
    appState?: AttrState;
    attr?: EmailServiceGroupAttr;
    defaultAttr?: Partial<EmailServiceGroupAttr>;
    name?: string;
    onChange?: OnChangeCallback;
  }

  export interface Props {
    defaultValue?: string;
    name: string;
    onAdd: OnAddCallback;
    onChange: OnChangeCallback;
    onFetch: OnFetchCallback;
    onRemove: OnRemoveCallback;
    value: string;
    provider: ImmutableObject<EmailMarketing.Service.Provider>;
  }
}
