// @divi/types/src/module/library/audio/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface AudioHTML {
  html?: string;
}

export interface AudioAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'conditions' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  image?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
      alt?: string;
      titleText?: string
    }>;
    decoration?: Element.Decoration.PickedAttributes<
      'border' |
      'boxShadow' |
      'filters'
    >;
  };

  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  caption?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  artistName?: {
    innerContent?: FormatBreakpointStateAttr<string>;
  };

  albumName?: {
    innerContent?: FormatBreakpointStateAttr<string>;
  };

  audio?: {
    innerContent?: FormatBreakpointStateAttr<string>;
  };
}
