// @divi/types/src/module/element/decoration/button/index.ts
import { type OnOff } from '../../../../attr';
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';
import { type InnerContent } from '../../types/button/inner-content';
import { type Spacing } from '../spacing';


export namespace Button {
  export interface AttributeValue {
    enable?: OnOff;
    alignment?: 'left' | 'center' | 'right';
    icon?: {
      enable?: OnOff;
      settings?: {
        unicode?: string,
        type?: string,
        weight?: '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900';
        color?: string;
        useSize?: OnOff;
        size?: string;
        show?: OnOff;
      };
      color?: string;
      placement?: 'right' | 'left';
      onHover?: OnOff;
    }
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Style {

    // TODO Replace `ButtonStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      affectingAttrs?: {
        spacing?: Spacing.Attributes
      };
      propertySelectors?: FormatPropertySelectors<Css.Properties.Button>;
      important?: boolean | FormatStyleImportant<Css.Properties.Button>;
      selectorFunctions?: {
        buttonHoverSelectorFunction?: Components.Style.SelectorFunction,
        buttonIconSelectorBeforeAfterFunction?: Components.Style.SelectorFunction,
        buttonIconSelectorFunction?: Components.Style.SelectorFunction,
        buttonIconSelectorHoverFunction?: Components.Style.SelectorFunction,
        buttonIconSelectorRightFunction?: Components.Style.SelectorFunction,
      };
    }
  }

  export namespace Component {

    // TODO replace `ButtonComponentProps` with this.
    export interface Props {
      allowEmptyUrl?: boolean;
      buttonAttr?: Attributes;
      className?: string;
      hasWrapper?: boolean;
      id?: string;
      innerContent?: InnerContent.Attributes;
      name?: string;
      renderIconAsDataAttribute?: boolean;
      type?: 'link' | 'button';
      dynamicDataLoaderProps?: Components.DynamicDataLoader.Props;
    }

  }
}
