// @divi/types/src/module/settings/group/library/background/index.ts
import { type MouseEvent } from 'react';

import { type FieldLibrary } from '../../../../../field-library';
import { type AttrState, type Breakpoint } from '../../../../../format-attr';
import { type Element } from '../../../../element';
import { type Field } from '../../../field';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Background {
  export namespace FieldsProps {
    export interface Color {
      color?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/color-picker']
      }
    }

    export interface Gradient {
      stops?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/gradient'];
      };
      enabled?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      type?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      direction?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      directionRadial?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      repeat?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      length?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      overlaysImage?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
    }

    export interface Image {
      url?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/upload'];
      };
      parallaxEnabled?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      parallaxMethod?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      size?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      width?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      height?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      position?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      horizontalOffset?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      verticalOffset?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      repeat?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      blend?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      }
    }

    export interface Video {
      mp4?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/upload'];
      };
      webm?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/upload'];
      };
      width?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/text'];
      };
      height?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/text'];
      };
      allowPlayerPause?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      pauseOutsideViewport?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
    }

    export interface Pattern {
      style?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select-pattern'];
      };
      enabled?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      color?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      transform?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      size?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      width?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      height?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      repeatOrigin?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      horizontalOffset?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      verticalOffset?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      repeat?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      blend?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
    }

    export interface Mask {
      style?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select-mask'];
      };
      enabled?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      color?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      transform?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      aspectRatio?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      size?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      width?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      height?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      position?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      horizontalOffset?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      verticalOffset?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      blend?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Background.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
    }
  }

  export interface Panel {
    attrName: string;
    fieldLabel?: string;
    panelVisible?: boolean;
    groupName?: string;
    defaultGroupAttr?: Element.Decoration.Background.Attributes;
  }

  export namespace PanelProps {
    export interface Color extends Panel {
      fields?: FieldsProps.Color;
    }

    export interface Gradient extends Panel {
      fields?: FieldsProps.Gradient;
    }

    export interface Image extends Panel {
      fields?: FieldsProps.Image;
    }

    export interface Video extends Panel {
      fields?: FieldsProps.Video;
    }

    export interface Pattern extends Panel {
      fields?: FieldsProps.Pattern;
    }

    export interface Mask extends Panel {
      fields?: FieldsProps.Mask;
    }
  }

  export interface Option {
    label: string
  }

  export namespace FieldCallback {

    // Previously `BackgroundFieldCallbackParams`
    export type Params = Field.CallbackParams<unknown, Element.Decoration.Background.Attributes>;

    // Previously `BackgroundFieldOnChangeCallbackParams`
    export interface OnChangeParams {
      event: CustomEvent | MouseEvent;
      moduleId: string;
      attrName: string;
      attrSubName?: string;
      responsiveMode: string;
      stateMode: string;
      inputValue?: string;
    }

    // Previously `BackgroundFieldOnResetCallbackParams`
    export interface OnResetParams {
      attrName: string;
      attrSubName?: string;
      moduleId: string;
      responsiveMode: Breakpoint;
      stateMode: AttrState;
    }

    // Previously `BackgroundFieldOnChangeCallbackParams`
    export interface GradientOnChangeParams {
      event: CustomEvent | MouseEvent;
      moduleId: string;
      attrName: string;
      attrSubName?: string;
      responsiveMode: string;
      stateMode: string;
      inputValue?: Element.Decoration.Background.GradientStop[];
    }
  }

  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Background.Attributes;
    defaultPanel?: 'color' | 'gradient' | 'image' | 'video' | 'pattern' | 'mask';
    hidePanels?: string[];
    fields?: {
      color?: FieldsProps.Color;
      gradient?: FieldsProps.Gradient;
      image?: FieldsProps.Image;
      video?: FieldsProps.Video;
      pattern?: FieldsProps.Pattern;
      mask?: FieldsProps.Mask;
    }
  }
}
