// @divi/types/src/module/settings/group/library/sticky/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Sticky {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Sticky.Attributes;
    fields?: {
      position?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sticky.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      offsetTop?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sticky.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      offsetBottom?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sticky.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      limitTop?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sticky.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      limitBottom?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sticky.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      offsetSurrounding?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sticky.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      transition?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Sticky.Attributes;
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
    }
  }
}
