// @divi/types/src/spam-protection/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Service } from '../service';


export namespace Store {

  export interface State {
    services?: Record<string, Service.Provider>;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  };

  export namespace Actions {
    export namespace SetService {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_SPAM_PROTECTION_SERVICE';
        provider: string;
        service: Service.Provider;
      }
    }

    export namespace SetServices {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_SPAM_PROTECTION_SERVICES';
        services: Record<string, Service.Provider>;
      }
    }

    export type FunctionMap = {
      setService: (params: SetService.Params) => SetService.ReturnValue;
      setServices: (params: SetServices.Params) => SetServices.ReturnValue;
    };

    export type ReturnValue =
      SetService.ReturnValue |
      SetServices.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getService: (provider: string) => ImmutableObject<Service.Provider>;
      getServices: () => ImmutableObject<Record<string, Service.Provider>>;
    }
  }
}
