// @divi/types/src/modal-library/store/index.ts
import { type ReactNode } from 'react';
import {
  type ImmutableArray,
  type ImmutableObject,
} from 'seamless-immutable';

import { type Help } from '../../help';
import { type Modal as TypePackageModal } from '../../modal';


export namespace Store {
  export type ModalDefinition = {
    name: string;
    label?: string;
    component: ReactNode;
    type: TypePackageModal.Type;
    videos?: Help.Video.Items;
    dimension?: {
      width: number;
      height: number;
    };
    dimensionMinimum?: {
      width?: number;
      height?: number;
    };
    layersView?: {
      expandAll?: boolean;
      modules: Record<string, LayerModuleData>;
    };

    // TO DO TO BE REMOVED
    snappable?: boolean;
  }

  export type Modal = ModalDefinition & {
    isActive: boolean;
    owner: string;
    offset: TypePackageModal.Offset;
    centered?: boolean;
    attributes?: Record<string, any>;
    tab: string;
    group: Record<string, string>;
    groupTab: Record<string, string>;
    panelSize?: number;
    bodySiblingHeight?: number;
    isResizing?: boolean;
    dragging?: boolean

    // Modal that is being overwritten by this modal when this modal is opened / activated.
    // This is only used for `singleInstanceModal` type.
    overwrittenModalName?: string;

    // TO DO TO BE REMOVED
    syncToLastClosed?: boolean;
    expand: boolean;
  }

  export type Modals = Record<string, Modal>;

  export type State = Modals;

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  }

  export namespace Actions {
    export namespace AddModal {
      export type ReturnValue = {
        type: 'ADD_MODAL';
        modal: ModalDefinition;
      }
    }

    export namespace Close {
      export type Params = {
        name: string;
      }

      export type ReturnValue = {
        type: 'CLOSE';
        params: Params;
      }
    }

    export namespace Expand {
      export type Params = {
        name: string;
        expand: boolean;
      }

      export type ReturnValue = {
        type: 'EXPAND',
        params: Params;
      }
    }

    export namespace Open {
      export type Params<TAttributes = Record<string, any>> = {
        name: string;
        owner?: string;
        dimension?: TypePackageModal.Dimension;
        offset?: TypePackageModal.Offset;
        centered?: boolean;

        // Modal specific configuration.
        attributes?: TAttributes;
      }

      export type ReturnValue<TAttributes = Record<string, any>> = {
        type: 'OPEN';
        params: Params<TAttributes>;
      }
    }

    export namespace OpenGroup {
      export type Params = {
        name: string;
        tabName: string;
        groupId: string;
        groupIds?: string[];
      }

      export type ReturnValue = {
        type: 'OPEN_GROUP';
        params: Params;
      }
    }

    export namespace OpenGroupTab {
      export type Params = {
        name: string;
        groupId: string;
        groupTabName: string;
      }

      export type ReturnValue = {
        type: 'OPEN_GROUP_TAB';
        params: Params;
      }
    }

    export namespace OpenTab {
      export type Params = {
        name: string;
        tabName: string;
      }

      export type ReturnValue = {
        type: 'OPEN_TAB';
        params: Params;
      }
    }

    export namespace RemoveModal {
      export type ReturnValue = {
        type: 'REMOVE_MODAL';
        name: string;
      }
    }

    export namespace SetAttributes {
      export type Params = {
        name: string;
        attributes: Record<string, any>;
        merge?: boolean;
      }

      export type ReturnValue = {
        type: 'SET_ATTRIBUTES';
        params: Params;
      }
    }

    export namespace SetBodySiblingHeight {
      export type Params = {
        name: string;
        bodySiblingHeight: number;
      }

      export type ReturnValue = {
        type: 'SET_BODY_SIBLING_HEIGHT';
        params: Params;
      }
    }

    export namespace SetDimension {
      export type Params = {
        name: string;
        dimension: TypePackageModal.Dimension;
        minimum?: TypePackageModal.Dimension;
      }

      export type ReturnValue = {
        type: 'SET_DIMENSION';
        params: Params;
      }
    }

    export namespace SetDraggingState {
      export type Params = {
        name: string;
        status: boolean;
      }

      export type ReturnValue = {
        type: 'SET_DRAGGING_STATE';
        params: Params;
      }
    }

    export namespace SetIsResizing {
      export type Params = {
        name: string;
        isResizing: boolean;
      }

      export type ReturnValue = {
        type: 'SET_IS_RESIZING';
        params: Params;
      }
    }

    export namespace SetOffset {
      export type Params = {
        name: string;
        offset: TypePackageModal.Offset;
      }

      export type ReturnValue = {
        type: 'SET_OFFSET';
        params: Params;
      }
    }

    export namespace SetPanelSize {
      export type Params = {
        name: string;
        panelSize: number;
      }

      export type ReturnValue = {
        type: 'SET_PANEL_SIZE';
        params: Params;
      }
    }

    export namespace SetLayersViewModule {

      // TODO feat(D5, Layers View) The `moduleInfo` is there as temporary solution to collect
      // all module info. So, we don't need to pass modal name when running dispatch. There is
      // a task to move Layers View to its own component.
      // @link https://github.com/elegantthemes/Divi/issues/36253.
      export type Params = {
        name: string,
        moduleInfo: {
          moduleId: string,
          childrenModuleIds?: ImmutableArray<string>;
          checkChildren?: boolean,
          data?: Record<string, unknown>,
        }
      }

      export type ReturnValue = {
        type: 'SET_LAYERS_VIEW_MODULE';
        params: Params;
      }
    }

    export namespace SetLayersViewExpandAll {
      export type Params = {
        name: string,
        expandAll: boolean,
      }

      export type ReturnValue = {
        type: 'SET_LAYERS_VIEW_EXPAND_ALL';
        params: Params;
      }
    }

    export type FunctionMap = {
      addModal: (modal: ModalDefinition) => AddModal.ReturnValue;
      close: (params: Close.Params) => Close.ReturnValue;
      expand: (params: Expand.Params) => Expand.ReturnValue;
      open: (params: Open.Params) => Open.ReturnValue;
      openGroup: (params: OpenGroup.Params) => OpenGroup.ReturnValue;
      openGroupTab: (params: OpenGroupTab.Params) => OpenGroupTab.ReturnValue;
      openTab: (params: OpenTab.Params) => OpenTab.ReturnValue;
      removeModal: (name: string) => RemoveModal.ReturnValue;
      setAttributes: (params: SetAttributes.Params) => SetAttributes.ReturnValue;
      setBodySiblingHeight: (params: SetBodySiblingHeight.Params) => SetBodySiblingHeight.ReturnValue;
      setDimension: (params: SetDimension.Params) => SetDimension.ReturnValue;
      setDraggingState: (params: SetDraggingState.Params) => SetDraggingState.ReturnValue;
      setIsResizing: (params: SetIsResizing.Params) => SetIsResizing.ReturnValue;
      setOffset: (params: SetOffset.Params) => SetOffset.ReturnValue;
      setPanelSize: (params: SetPanelSize.Params) => SetPanelSize.ReturnValue;
      setLayersViewExpandAll: (expandAll: boolean) => SetLayersViewExpandAll.ReturnValue;
      setLayersViewModule: (moduleInfo: SetLayersViewModule.Params['moduleInfo']) => SetLayersViewModule.ReturnValue;
    }

    export type ReturnValue =
      AddModal.ReturnValue |
      Close.ReturnValue |
      Expand.ReturnValue |
      Open.ReturnValue |
      OpenGroup.ReturnValue |
      OpenGroupTab.ReturnValue |
      OpenTab.ReturnValue |
      RemoveModal.ReturnValue |
      SetAttributes.ReturnValue |
      SetBodySiblingHeight.ReturnValue |
      SetDimension.ReturnValue |
      SetDraggingState.ReturnValue |
      SetIsResizing.ReturnValue |
      SetLayersViewExpandAll.ReturnValue |
      SetLayersViewModule.ReturnValue |
      SetOffset.ReturnValue |
      SetPanelSize.ReturnValue;
  }

  export namespace Selectors {
    export type FunctionMap = {

      // TODO To be removed. `getActiveModal` is no longer used in enhanced design store.
      getActiveModal<T>(modalType: T): T extends ('multi' | 'confirmation') ? Record<string, TypePackageModal.ModalStates> : TypePackageModal.ModalStates;

      getLayersViewExpandAll: () => boolean;
      getLayersViewModule: (moduleId: string) => LayerModuleData;
      getModal: (name: string) => Modal;
      getModalLabelsByNames: (modalIds: string[]) => Record<string, string>;
      getModals: () => State;
      getModalOwner: (modalName: string) => string;
      isModalActive: (modalName: string) => boolean;
      isModalDragging: (modalName: string) => boolean;
      isModuleEdited: (moduleId: string) => boolean;
    }
  }

  export interface LayerModuleData {
    id: string;
    expand?: boolean;
    active?: boolean;
  }
}
