// @divi/types/src/shortcode-module/advanced-field/button/index.ts
/* eslint-disable @typescript-eslint/naming-convention */
export interface ShortcodeModuleAdvancedFieldDefinitionButton {
  border_width?: {
    default?: string;
  };
  box_shadow?: boolean | {
    css?: {
      important?: boolean;
      main?: string;
    }
  }
  css?: {
    alignment?: string;

    // Looks like typo but this exist on et_pb_fullwidth_slider module.
    linited_main?: string;
    limited_main?: string;
    important?: string | boolean;
    main?: string;
    show_if?: Record<string, string>;
  };
  hide_icon?: boolean;
  hide_custom_padding?: boolean;
  label?: string;
  margin_padding?: boolean | {
    css?: {
      main?: string;
      important?: string | string[];
    };
    custom_padding?: {
      default?: string;
    }
  };
  no_rel_attr?: boolean;
  option_category?: string;
  show_if?: Record<string, string>;
  tab_slug?: string;
  text_size?: {
    default?: string;
  };
  toggle_priority?: number;
  toggle_slug?: string;
  use_alignment?: boolean;
  use_icon?: boolean;
}
