// @divi/types/src/module/components/inline-text-editor/index.ts
export namespace InlineTextEditor {

  // TODO Replace `InlineTextEditorAdditionalProps` with this.
  export interface AdditionalProps {
    onClick?: (event: MouseEvent) => void;
    href?: string;
    title?: string;
    target?: string;
    style?: Record<string, string>;
  }

  // TODO Replace `InlineTextEditorContainerProps` with this.
  export type AllowedTags = Pick<HTMLElementTagNameMap, 'div'>;

  // TODO Replace `InlineTextEditorContainerProps` with this.
  export interface ContainerProps {
    attrName: string;
    moduleId: string;
    tagName:  keyof AllowedTags;
    text: string;

    // Optional props.
    additionalProps?: AdditionalProps;
    allowHTML?: boolean;
    attrSubName?: string;
    callbackEditingStatusChange?: (event: FocusEvent, isEditing: boolean) => void;
    defaultText?: string;
    className?: string;
    handleBlur?: (event: FocusEvent) => void;
    handleClick?: (event: MouseEvent) => void;
    handleDoubleClick?: (event: MouseEvent) => void;
    handleKeyDown?: (event: KeyboardEvent) => void;
    handleKeyUp?: (event: KeyboardEvent) => void;
    handleMouseDown?: (event: MouseEvent) => void;
    handleMouseUp?: (event: MouseEvent) => void;
    handlePaste?: (event: ClipboardEvent) => void;
    handleShortcode?: (content: string) => string;
  }
}
