// @divi/types/src/cloud-app/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Library } from '../library';


export namespace Store {
  export interface State {
    libraryData: Library.Data.Fetched;
    libraryItem: Library.Item.Fetched;
    libraryTerms: Library.Term.Fetched;
    fetchLog: Library.Data.FetchLog;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace FetchLibraryData {
      export type ReturnValue = {
        type: 'FETCH_LIBRARY_DATA';
        context: Library.Data.Type[];
      }
    }

    export namespace FetchLibraryDataLog {
      export type ReturnValue = {
        type: 'FETCH_LIBRARY_DATA_LOG';
        context: Library.Data.Type;
        inProgress: boolean;
      }
    }

    export namespace FetchLibraryItem {
      export type ProgressCallback = (completed: number) => void;

      export type SuccessCallback = (response: ImmutableObject<Library.Item.Item>) => void;

      export type Callbacks = {
        progress?: ProgressCallback;
        success?: SuccessCallback;
      };

      export type ReturnValue = {
        type: 'FETCH_LIBRARY_ITEM';
        context: Library.Data.Type;
        id: number;
        callbacks?: Callbacks;
      }
    }

    export namespace PurgeLibraryData {
      export type ReturnValue = {
        type: 'PURGE_LIBRARY_DATA';
      }
    }

    export namespace SetLibraryData {
      export type ReturnValue = {
        type: 'SET_LIBRARY_DATA';
        context: Library.Data.Type;
        data: Library.Data.Item;
      }
    }

    export namespace SetLibraryItem {
      export type ReturnValue = {
        type: 'SET_LIBRARY_ITEM';
        id: number;
        data: Library.Item.Item;
      }
    }

    export namespace SetLibraryTerms {
      export type ReturnValue = {
        type: 'SET_LIBRARY_TERMS';
        termsType: Library.Term.Type;
        termsData: Library.Term.Item[];
      }
    }

    export namespace UpdateLibraryTerms {
      export type SuccessCallback = (response: Response) => void;

      export interface Data {
        updateType: 'remove' | 'rename' | 'add';
        filterType: string;
        id: string;
        newName?: string;
        location: string;
      }

      export interface ResponseNewFilter {
        name: string;
        id: number;
        location: string;
      }

      export interface Response {
        newFilters: ResponseNewFilter[];
        filterType: string;
        localLibraryTerms: {
          [key in Library.Term.Type]?: Library.Term.Item[];
        }
      }

      export type Callbacks = {
        success?: SuccessCallback;
      };

      export type ReturnValue = {
        type: 'UPDATE_LIBRARY_TERMS';
        data: Data[];
        callbacks?: Callbacks;
      }

    }

    export type FunctionMap = {
      fetchLibraryData: (context: Library.Data.Type[]) => FetchLibraryData.ReturnValue;
      fetchLibraryDataLog: (context: Library.Data.Type, inProgress: boolean) => FetchLibraryDataLog.ReturnValue;
      fetchLibraryItem: (
        context: Library.Data.Type,
        id: number,
        callbacks?: FetchLibraryItem.Callbacks
      ) => FetchLibraryItem.ReturnValue;
      purgeLibraryData: () => PurgeLibraryData.ReturnValue;
      setLibraryData: (
        context: Library.Data.Type,
        data: Library.Data.Item
      ) => SetLibraryData.ReturnValue;
      setLibraryItem: (
        id: number,
        data: Library.Item.Item
      ) => SetLibraryItem.ReturnValue;
      setLibraryTerms: (
        termsType: Library.Term.Type,
        termsData: Library.Term.Item[]
      ) => SetLibraryTerms.ReturnValue;
      updateLibraryTerms: (
        data: UpdateLibraryTerms.Data[],
        callbacks?: UpdateLibraryTerms.Callbacks
      ) => UpdateLibraryTerms.ReturnValue;
    };

    export type ReturnValue =
      FetchLibraryData.ReturnValue |
      FetchLibraryDataLog.ReturnValue |
      FetchLibraryItem.ReturnValue |
      PurgeLibraryData.ReturnValue |
      SetLibraryData.ReturnValue |
      SetLibraryItem.ReturnValue |
      SetLibraryTerms.ReturnValue |
      UpdateLibraryTerms.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getFetchLibraryDataLog: () => ImmutableObject<Library.Data.FetchLog>;
      getLibraryData: () => ImmutableObject<Library.Data.Fetched>;
      getLibraryItem: () => ImmutableObject<Library.Item.Fetched>;
      getLibraryTerms: () => ImmutableObject<Library.Term.Fetched>;
      isFetchingLibraryData: (context: Library.Data.Type) => boolean;
    }
  }
}
