// @divi/types/src/module/settings/group/library/text/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type Field } from '../../../field';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';
import { type TextShadow } from '../text-shadow';


export namespace Text {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Advanced.Text.Attributes;
    fields?: {
      orientation?: GroupFieldProps & {
        defaultAttr?: Element.Advanced.Text.Attributes['text'];
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      color?: GroupFieldProps & {
        defaultAttr?: Element.Advanced.Text.Attributes['text'];
        component?: FieldLibrary.ComponentMap['divi/select']
      };
      textShadowGroup?: Field.BaseProps & {
        component?: {
          type?: 'group';
          name?: 'divi/text-shadow';
          props: TextShadow.GroupProps;
        }
      };
    }
  }
}
