// @divi/types/src/module/settings/group/library/transform/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Transform {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Transform.Attributes;
    defaultTab?: keyof Element.Decoration.Transform.AttributeValue;
    fields?: {
      tabNav?: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      scale?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Transform.Attributes;
        component?: FieldLibrary.ComponentMap['divi/transform-scale'];
      };
      translate?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Transform.Attributes;
        component?: FieldLibrary.ComponentMap['divi/transform-translate'];
      };
      rotate?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Transform.Attributes;
        component?: FieldLibrary.ComponentMap['divi/transform-rotate'];
      };
      skew?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Transform.Attributes;
        component?: FieldLibrary.ComponentMap['divi/transform-skew'];
      };
      origin?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Transform.Attributes;
        component?: FieldLibrary.ComponentMap['divi/transform-origin'];
      };
    }
  }
}
