// @divi/types/src/module/settings/group/library/text-shadow/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace TextShadow {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Font.TextShadow.Attributes;
    fields?: {
      style?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.TextShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      horizontal?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.TextShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      vertical?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.TextShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      blur?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.TextShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      color?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Font.TextShadow.Attributes;
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      }
    }
  }
}
