// @divi/types/src/module-library/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type ModuleAttrs } from '../../module/library';
import { type ChildModules } from '../child-modules';
import { type Folder } from '../folder';
import { type Module } from '../module';


export namespace Store {
  export interface State {
    folders: Folder.Collections;
    modules: Module.Collections;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  };

  export namespace Actions {
    export namespace AddFolder {
      export type ReturnValue = {
        type: 'ADD_FOLDER';
        folder: Folder.Definition;
      }
    }

    export namespace AddModule {
      export type ReturnValue = {
        type: 'ADD_MODULE';
        module: Module.Definition;
      }
    }

    export namespace RemoveFolder {
      export type ReturnValue = {
        type: 'REMOVE_FOLDER';
        name: string;
        path?: string;
      }
    }

    export namespace RemoveModule {
      export type ReturnValue = {
        type: 'REMOVE_MODULE';
        moduleName: string;
      }
    }

    export type FunctionMap = {
      addModule: (module: Module.Definition) => Actions.AddModule.ReturnValue;
      removeModule: (moduleName: string) => Actions.RemoveModule.ReturnValue;
      addFolder: (folder: Folder.Definition) => Actions.AddFolder.ReturnValue;
      removeFolder: (folderName: string, folderPath?: string) => Actions.RemoveFolder.ReturnValue;
    };

    export type ReturnValue =
      AddFolder.ReturnValue |
      AddModule.ReturnValue |
      RemoveFolder.ReturnValue |
      RemoveModule.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getModules: () => Module.Collections;
      getModule: (moduleName: string) => Module.Settings;
      getChildModules: (moduleParams: ChildModules.Params) => Module.Collections;
      getModuleCategory: (moduleName: string) => Module.Category;
      getFolders: (folderName?: string, moduleAttrs?: ModuleAttrs) => Folder.Collections;
      isValidChild: (moduleName: string, moduleAttrs: ModuleAttrs, childModuleName: string) => boolean;
      isFullwidthModule: (moduleName: string) => boolean;
      isChildModule: (moduleName: string) => boolean;
      isParentModule: (moduleName: string) => boolean;
    }
  }
}
