// @divi/types/src/edit-post/store/index.ts
import { type ImmutableArray, type ImmutableObject } from 'seamless-immutable';

import { type CloudApp } from '../../cloud-app';
import { type GlobalData } from '../../global-data';
import {
  type Module,
  type ModuleAttr,
  type ModuleAttrs,
} from '../../module';
import { type ModuleType } from '../../module/category';
import {
  type ModuleFlatObject,
  type ModuleFlatObjects,
} from '../../module/state';
import { type Clipboard } from '../clipboard';


export namespace Store {
  export type Caller = 'user' | 'global-layout';

  export type NewModulePosition = 'before' | 'after' | 'inside';

  export type NewSpecialtySectionPosition = 'before' | 'after';

  export type ImportOptionsValues = {
    replaceLayout: string,
    importBackUp: string,
    includeGlobalPresets: string,
  }

  export type FilteredModulesByIds<T> = T extends true ? ModuleFlatObjects : ImmutableObject<ModuleFlatObjects>;

  export interface AncestorModule {
    id: string;
    name: string;
  }

  export type AncestorModules = AncestorModule[];

  export interface State<TModuleFlatObjects = ModuleFlatObjects> {
    content: TModuleFlatObjects;

    // Presets Prototype. Marking this as optional to avoid breaking the existing code.
    // This presets data keep tracks of all the presets that is currently used in this page.
    presets?: {
      module: Record<GlobalData.Presets.Module.Name, GlobalData.Presets.Module.Id[]>;
    }
  }

  export type ImmutableState<TState = State> = ImmutableObject<TState>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace AddColumnViaDivider {
      export type ReturnValue = {
        type: 'ADD_COLUMN_VIA_DIVIDER',
        id: string,
        dividerPosition: number,
        caller: Caller,
      }
    }

    export namespace AddModule {
      export type ReturnValue = {
        type: 'ADD_MODULE';
        id: string;
        name: string;
        props: ModuleFlatObject['props'];
        position: NewModulePosition;
        newId: string;
        caller: Caller;
      }
    }

    export namespace AddRow {
      export type ReturnValue = {
        type: 'ADD_ROW';
        id: string;
        position: 'before' | 'after' | 'inside';
        columnStructure: string;
        inner: boolean;
        caller: Caller;
      }
    }

    export namespace AddSpecialtySection {
      export type ReturnValue = {
        type: 'ADD_SPECIALTY_SECTION';
        id: string;
        position: NewSpecialtySectionPosition;
        columnStructure: string;
        innerPosition: string;
      }
    }

    export namespace ChangeColumnStructure {
      export type ReturnValue = {
        type: 'CHANGE_COLUMN_STRUCTURE';
        id: string;
        columnStructure: string;
        caller: Caller;
      }
    }

    export namespace ClearLayout {
      export type ReturnValue = {
        type: 'CLEAR_LAYOUT';
      }
    }

    export namespace CloneColumn {
      export type ReturnValue = {
        type: 'CLONE_COLUMN';
        id: string;
      }
    }

    export namespace CloneModule {
      export type ReturnValue = {
        type: 'CLONE_MODULE';
        id: string;
        caller: Caller;
      }
    }

    export namespace CopyModule {
      export type ReturnValue = {
        type: 'COPY_MODULE';
        id: string;
        position: NewModulePosition;
        targetId: string;
      }
    }

    export namespace CopyModuleFromLibrary {
      export type ReturnValue = {
        type: 'COPY_MODULE_FROM_LIBRARY';
        payload: ImmutableObject<CloudApp.Library.Item.Layout>;
        position: NewModulePosition;
        targetId: string;
        caller: Caller;
      }
    }

    export namespace CopyModuleFromPayload {
      export type ReturnValue = {
        type: 'COPY_MODULE_FROM_PAYLOAD';
        payload: Clipboard.Item.Payload.Module;
        position: NewModulePosition;
        targetId: string;
        caller: Caller;
      }
    }

    export namespace CopyModulesFromPayload {
      export type ReturnValue = {
        type: 'COPY_MODULES_FROM_PAYLOAD';
        payload: Clipboard.Item.Payload.Modules;
        position: NewModulePosition;
        targetId: string;
        caller: Caller;
      }
    }

    export namespace DisableGlobalModule {
      export type ReturnValue = {
        type: 'DISABLE_GLOBAL_MODULE';
        id: string;
      }
    }

    export namespace EditChildren {
      export type ReturnValue = {
        type: 'EDIT_CHILDREN';
        id: string;
        children: string[];
      }
    }

    export namespace EditModuleAttribute {
      export type ReturnValue = {
        type: 'EDIT_MODULE_ATTRIBUTE';
        id: string;
        attrName: string;
        value: ImmutableObject<ModuleAttr> | string;
        caller?: Caller;
      }
    }

    export namespace EditModuleProp {
      export type ReturnValue = {
        type: 'EDIT_MODULE_PROP';
        id: string;
        propName: keyof ModuleFlatObject['props'];
        propValue: ModuleFlatObject['props'][keyof ModuleFlatObject['props'] ];
      }
    }

    export namespace EditUnsyncAttrs {
      export type ReturnValue = {
        type: 'EDIT_UNSYNC_ATTRS';
        id: string;
        attribute: Module.Unsync.Attr;
        isSynced: boolean;
        caller: Caller;
      }
    }

    export namespace ImportContent {
      export type ReturnValue = {
        type: 'IMPORT_CONTENT',
        content: string,
        importOptionsValues: ImportOptionsValues,
      }
    }

    export namespace MergeGlobalModules {
      export type ReturnValue = {
        type: 'MERGE_GLOBAL_MODULES';
        content: ModuleFlatObjects;
        owner: string;
      }
    }

    export namespace MoveModule {
      export type ReturnValue = {
        type: 'MOVE_MODULE';
        id: string;
        position: NewModulePosition;
        targetId: string;
        caller: Caller;
      }
    }

    export namespace MoveModules {
      export type ReturnValue = {
        type: 'MOVE_MODULES';
        ids: string[];
        position: NewModulePosition;
        targetId: string;
      }
    }

    export namespace RemoveColumn {
      export type ReturnValue = {
        type: 'REMOVE_COLUMN';
        id: string;
        caller: Caller;
      }
    }

    export namespace RemoveGlobalModuleAttribute {
      export type ReturnValue = {
        type: 'REMOVE_GLOBAL_MODULE_ATTRIBUTE';
        id: string;
      }
    }

    export namespace RemoveGlobalParentAttribute {
      export type ReturnValue = {
        type: 'REMOVE_GLOBAL_PARENT_ATTRIBUTE';
        id: string;
      }
    }

    export namespace RemoveModule {
      export type ReturnValue = {
        type: 'REMOVE_MODULE';
        id: string;
        caller: Caller;
      }
    }

    export namespace RemoveModuleAttribute {
      export type ReturnValue = {
        type: 'REMOVE_MODULE_ATTRIBUTE';
        id: string;
        attrName: string;
        caller: Caller;
      }
    }

    export namespace ReplaceModuleAttributes {
      export type ReturnValue = {
        type: 'REPLACE_MODULE_ATTRIBUTES';
        id: string;
        attrs: ImmutableObject<ModuleAttr>;
        caller: Caller;
      }
    }

    export namespace SetGlobalModuleAttribute {
      export type ReturnValue = {
        type: 'SET_GLOBAL_MODULE_ATTRIBUTE';
        id: string;
        value: ImmutableObject<ModuleAttr> | string;
      }
    }

    export namespace SetGlobalParentAttribute {
      export type ReturnValue = {
        type: 'SET_GLOBAL_PARENT_ATTRIBUTE';
        id: string;
        value: ImmutableObject<ModuleAttr> | string;
      }
    }

    export namespace SetLayoutState {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_LAYOUT_STATE';
        location: keyof State;
        layoutState: ModuleFlatObjects | ImmutableObject<ModuleFlatObjects>;
      }
    }

    export namespace SyncGlobalLayoutAttributes {
      export type ReturnValue = {
        type: 'SYNC_GLOBAL_LAYOUT_ATTRIBUTES';
        id: string;
        attrs: ImmutableObject<ModuleAttr>;
        unsyncAttrs: Module.Unsync.Attrs;
      }
    }

    export namespace SyncModuleInlineFont {
      export type ReturnValue = {
        type: 'SYNC_MODULE_INLINE_FONT';
      }
    }

    export type FunctionMap = {
      addColumnViaDivider: (
        id: string,
        dividerPosition: number,
        caller?: Caller,
      ) => AddColumnViaDivider.ReturnValue,
      addModule: (
        id: string,
        name: string,
        props?: AddModule.ReturnValue['props'],
        position?: AddModule.ReturnValue['position'],
        newId?: string,
        caller?: Caller,
      ) => AddModule.ReturnValue,
      addRow: (
        id: string,
        position: AddRow.ReturnValue['position'],
        columnStructure: string,
        inner?: boolean,
        caller?: Caller,
      ) => AddRow.ReturnValue,
      addSpecialtySection: (
        id: string,
        position: AddSpecialtySection.ReturnValue['position'],
        columnStructure: string,
        innerPosition: string,
      ) => AddSpecialtySection.ReturnValue,
      changeColumnStructure: (
        id: string,
        columnStructure: string,
        caller?: Caller,
      ) => ChangeColumnStructure.ReturnValue,
      clearLayout: () => ClearLayout.ReturnValue,
      cloneColumn: (
        id: string,
      ) => CloneColumn.ReturnValue,
      cloneModule: (
        id: string,
        caller?: Caller,
      ) => CloneModule.ReturnValue,
      copyModule: (
        id: CopyModule.ReturnValue['id'],
        position : CopyModule.ReturnValue['position'],
        targetId: CopyModule.ReturnValue['targetId'],
      ) => CopyModule.ReturnValue,
      copyModuleFromLibrary: (
        payload: CopyModuleFromLibrary.ReturnValue['payload'],
        position: CopyModuleFromLibrary.ReturnValue['position'],
        targetId: string,
        caller?: Caller,
      ) => CopyModuleFromLibrary.ReturnValue,
      copyModuleFromPayload: (
        payload: CopyModuleFromPayload.ReturnValue['payload'],
        position: CopyModuleFromPayload.ReturnValue['position'],
        targetId: string,
        caller?: Caller,
      ) => CopyModuleFromPayload.ReturnValue,
      copyModulesFromPayload: (
        payload: CopyModulesFromPayload.ReturnValue['payload'],
        position: CopyModulesFromPayload.ReturnValue['position'],
        targetId: string,
        caller?: Caller,
      ) => CopyModulesFromPayload.ReturnValue,
      disableGlobalModule: (
        id: string,
      ) => DisableGlobalModule.ReturnValue,
      editChildren: (
        id: string,
        children: string[],
      ) => EditChildren.ReturnValue,
      editModuleAttribute: (
        id: string,
        attrName: string,
        value: ImmutableObject<ModuleAttr> | string,
        caller?: Caller,
      ) => EditModuleAttribute.ReturnValue,
      editModuleProp: (
        id: EditModuleProp.ReturnValue['id'],
        propName: EditModuleProp.ReturnValue['propName'],
        propValue: EditModuleProp.ReturnValue['propValue'],
      ) => EditModuleProp.ReturnValue,
      editUnsyncAttrs: (
        id: string,
        attribute: EditUnsyncAttrs.ReturnValue['attribute'],
        isSynced: boolean,
        caller?: Caller,
      ) => EditUnsyncAttrs.ReturnValue,
      importContent: (
        content: string,
        importOptionsValues: ImportContent.ReturnValue['importOptionsValues'],
      ) => ImportContent.ReturnValue,
      mergeGlobalModules: (
        content: MergeGlobalModules.ReturnValue['content'],
        owner: MergeGlobalModules.ReturnValue['owner'],
      ) => MergeGlobalModules.ReturnValue,
      moveModule: (
        id: string,
        position: MoveModule.ReturnValue['position'],
        targetId: string,
        caller?: Caller,
      ) => MoveModule.ReturnValue,
      moveModules: (
        ids: MoveModules.ReturnValue['ids'],
        position: MoveModules.ReturnValue['position'],
        targetId: MoveModules.ReturnValue['targetId'],
      ) => MoveModules.ReturnValue,
      removeColumn: (
        id: RemoveColumn.ReturnValue['id'],
        caller?: RemoveColumn.ReturnValue['caller'],
      ) => RemoveColumn.ReturnValue,
      removeGlobalModuleAttribute: (
        id: RemoveGlobalModuleAttribute.ReturnValue['id'],
      ) => RemoveGlobalModuleAttribute.ReturnValue,
      removeGlobalParentAttribute: (
        id: RemoveGlobalParentAttribute.ReturnValue['id'],
      ) => RemoveGlobalParentAttribute.ReturnValue,
      removeModule: (
        id: RemoveModule.ReturnValue['id'],
        caller?: RemoveModule.ReturnValue['caller'],
      ) => RemoveModule.ReturnValue,
      removeModuleAttribute: (
        id: RemoveModuleAttribute.ReturnValue['id'],
        attrName: RemoveModuleAttribute.ReturnValue['attrName'],
        caller?: RemoveModuleAttribute.ReturnValue['caller'],
      ) => RemoveModuleAttribute.ReturnValue,
      replaceModuleAttributes: (
        id: ReplaceModuleAttributes.ReturnValue['id'],
        attrs: ReplaceModuleAttributes.ReturnValue['attrs'],
        caller?: ReplaceModuleAttributes.ReturnValue['caller'],
      ) => ReplaceModuleAttributes.ReturnValue,
      setGlobalModuleAttribute: (
        id: SetGlobalModuleAttribute.ReturnValue['id'],
        value: SetGlobalModuleAttribute.ReturnValue['value'],
      ) => SetGlobalModuleAttribute.ReturnValue,
      setGlobalParentAttribute: (
        id: SetGlobalParentAttribute.ReturnValue['id'],
        value: SetGlobalParentAttribute.ReturnValue['value'],
      ) => SetGlobalParentAttribute.ReturnValue,
      setLayoutState: (
        params: SetLayoutState.Params,
      ) => SetLayoutState.ReturnValue,
      syncGlobalLayoutAttributes: (
        id: SyncGlobalLayoutAttributes.ReturnValue['id'],
        attrs: SyncGlobalLayoutAttributes.ReturnValue['attrs'],
        unsyncAttrs: SyncGlobalLayoutAttributes.ReturnValue['unsyncAttrs'],
      ) => SyncGlobalLayoutAttributes.ReturnValue,
      syncModuleInlineFont: () => SyncGlobalLayoutAttributes.ReturnValue,
    };

    export type ReturnValue =
      AddColumnViaDivider.ReturnValue |
      AddModule.ReturnValue |
      AddRow.ReturnValue |
      AddSpecialtySection.ReturnValue |
      ChangeColumnStructure.ReturnValue |
      ClearLayout.ReturnValue |
      CloneColumn.ReturnValue |
      CloneModule.ReturnValue |
      CopyModule.ReturnValue |
      CopyModuleFromLibrary.ReturnValue |
      CopyModuleFromPayload.ReturnValue |
      CopyModulesFromPayload.ReturnValue |
      DisableGlobalModule.ReturnValue |
      EditChildren.ReturnValue |
      EditModuleAttribute.ReturnValue |
      EditModuleProp.ReturnValue |
      EditUnsyncAttrs.ReturnValue |
      ImportContent.ReturnValue |
      MergeGlobalModules.ReturnValue |
      MoveModule.ReturnValue |
      MoveModules.ReturnValue |
      RemoveColumn.ReturnValue |
      RemoveGlobalModuleAttribute.ReturnValue |
      RemoveGlobalParentAttribute.ReturnValue |
      RemoveModule.ReturnValue |
      RemoveModuleAttribute.ReturnValue |
      ReplaceModuleAttributes.ReturnValue |
      SetGlobalModuleAttribute.ReturnValue |
      SetGlobalParentAttribute.ReturnValue |
      SetLayoutState.ReturnValue |
      SyncGlobalLayoutAttributes.ReturnValue |
      SyncModuleInlineFont.ReturnValue;
  };

  export namespace Effects {
    export namespace MaybeCloseModuleSettings {
      export type Action = Actions.RemoveModule.ReturnValue |
      Actions.RemoveColumn.ReturnValue;
    }
  }

  export namespace Selectors {
    export type FunctionMap = {
      getAncestorModules: (
        id: string,
        sort?: 'top-to-bottom' | 'bottom-to-top',
      ) => AncestorModules,
      getContent: () => ImmutableState['content'],
      getGlobalLayoutId: (id: string) => string | null,
      getGlobalLayoutOwners: (id: string) => ImmutableArray<string>,
      getModule: <TAttrs = ModuleAttrs>(id: string) => ImmutableObject<ModuleFlatObject<TAttrs>>;
      getModuleAttr: (
        id: string,
        attrName: string,
      ) => ImmutableObject<ModuleAttr>,
      getModuleAttrs: (
        id: string,
        defaultValue?: ImmutableObject<ModuleAttrs>
      ) => ImmutableObject<ModuleAttrs>,
      getModuleName: (id: string) => string,
      getModuleStructureIds: (id: string) => string[],
      getModuleType: (id: string) => ModuleType,
      getModuleWithChildren: (id: string) => ModuleFlatObjects,
      getModulesByIds: <T extends boolean>(ids: string[], mutable: T) => FilteredModulesByIds<T>,
      getParentModule: (id: string) => ImmutableObject<ModuleFlatObject>,
      getParentModuleType: (id: string) => ModuleType,
      getRootModuleIds: () => string[],
      getSiblingModuleIds: (id: string) => string[],
      getUsedPresets: () => State['presets'],
      isContentEmpty: () => boolean,
    }
  }
}
