// @divi/types/src/field-library/common-numeric-spinner/index.ts
import { type NumericInput } from '../common-numeric-input';


export namespace NumericSpinner {

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} value Value.
   * @property {Function} onChange On change callback.
   * @property {string} [defaultValue] Default value.
   * @property {string} [defaultUnit] Default unit.
   * @property {string} [label] Label.
   * @property {number} [step] Step.
   * @property {number} [min] Min value.
   * @property {number} [max] Max value.
   * @property {number} [speed] Speed.
   * @property {boolean} ["readonly"] Read only.
   * @property {string} [className] Class name.
   * @property {string[]} [allowedUnits] Allowed units.
   * @property {string[]} [allowedValues] Allowed values.
   * @property {Function} [formatter] Formatter.
   * @property {object[]} [valueMapping] Value mapping.
   */
  export interface Props {
    name: string;
    value: string;
    onChange: (newValue: string, updateSource: string) => void;
    defaultValue?: string;
    defaultUnit?: string;
    label?: string;
    step?: number;
    min?: number | null;
    max?: number | null;
    speed?: number;
    readonly?: boolean;
    className?: string;
    allowedUnits?: string[];
    allowedValues?: string[];
    formatter?: NumericInput.Formatter;
    valueMapping?: NumericInput.ValueMapping[];
  }
}
