// @divi/types/src/module/library/pricing-table/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface PricingTableCssAttr extends Css.AttributeValue {
  pricingHeading?: string;
  pricingTitle?: string;
  pricingSubtitle?: string;
  pricingTop?: string;
  price?: string;
  currency?: string;
  frequency?: string;
  pricingContent?: string;
  pricingItem?: string;
  pricingItemExcluded?: string;
  pricingButton?: string;
}

export type PricingTableCssGroupAttr = FormatBreakpointStateAttr<PricingTableCssAttr>;

export interface PricingTableAttrs extends InternalAttrs {

  // Button
  button?: Element.Types.Button.Attributes;

  css?: PricingTableCssGroupAttr;

  // Module
  module?: {
    advanced?: {
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
      featured?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'spacing' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };


  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };

  price?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font' | 'border' | 'background'>;
  };

  currencyFrequency?: {
    innerContent?: FormatBreakpointStateAttr<{
      currency?: string;
      per?: string;
    }>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  subtitle?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  }

  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    advanced?: {
      bulletColor?: FormatBreakpointStateAttr<string>;
    }
    decoration?: Element.Decoration.PickedAttributes<'bodyFont'>;
  };

  excluded?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };
}
