// @divi/types/src/field-library/spacing/index.ts
import { type OnOff } from '../../attr';
import { type Spacing as SpacingType } from '../../module/element/decoration/spacing';


export namespace Spacing {
  export type SideName = keyof SpacingType.Sides;

  export type SideNameVertical = keyof SpacingType.SidesVertical;

  export type SideNameHorizontal = keyof SpacingType.SidesHorizontal;

  export type Steps = Record<SideName, number>;

  /**
   * @typedef {object} InputValue
   * @property {string} side Side.
   * @property {'on' | 'off'} syncHorizontal Sync horizontal.
   * @property {'on' | 'off'} syncVertical Sync vertical.
   * @property {string} value Value.
   */
  export interface InputValue {
    side: keyof SpacingType.Sides;
    syncHorizontal: OnOff;
    syncVertical: OnOff;
    value: string;
  }

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {Record<string, string>} inputValue Input value.
   */
  export interface OnChangeCallbackParams {
    inputValue: Record<string, string>;
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} [className] Class name.
   * @property {Function} onChange On change callback.
   * @property {object} value Value.
   * @property {object} [defaultValue] Default value.
   * @property {number} [min] Min value.
   * @property {string[]} allowedUnits Allowed units.
   */
  export interface Props {
    name: string;
    className?: string;
    onChange: (params: OnChangeCallbackParams) => void;
    value: SpacingType.ItemValue;
    defaultValue?: SpacingType.ItemValue;
    min?: number | null;
    allowedUnits: string[];
  }
}
