// @divi/types/src/module/library/circle-counter/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface CircleBackgroundAttrs {
  color?: string;
  opacity?: string;
}
export interface CircleAttrs {
  color?: FormatBreakpointStateAttr<string>;
  background?: FormatBreakpointStateAttr<CircleBackgroundAttrs>
}

export interface CircleCounterCssAttr extends Css.AttributeValue {
  percent?: string;
  circleCounterTitle?: string;
  percentText?: string;
}

export type CircleCounterCssGroupAttr = FormatBreakpointStateAttr<CircleCounterCssAttr>;

export interface CircleCounterAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: CircleCounterCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };
  number?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    advanced?: {
      percentSign?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };
  circle?: {
    advanced?: {

      // The below attributes are looks like they are for CSS. But they are not. It's for the circle counter's JS.
      color?: FormatBreakpointStateAttr<string>;
      background?: FormatBreakpointStateAttr<CircleBackgroundAttrs>;
    }
  };
}
