// @divi/types/src/module/library/contact-field/index.ts
import { type OnOff } from '../../../attr';
import { type SortableList } from '../../../field-library/sortable-list';
import {
  type FormatBreakpointStateAttr,
} from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import {
  type FieldElementAttr,
} from '../../options';
import {
  type InternalAttrs,
} from '../internal';


interface ContactFieldCondition {
  condition: string;
  field: string;
  value?: string;
}

export interface ContactFieldAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.AttributeValue;

  module?: {
    advanced?: {
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    },
    decoration?: Pick<Element.Decoration.Attributes,
      'background' |
      'conditions' |
      'disabledOn' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex' |
      'border' |
      'boxShadow' |
      'filters'
    >;
  },

  // Contact Field Item.
  fieldItem?: {
    advanced?: {
      fullwidth?: FormatBreakpointStateAttr<OnOff>;
      required?: FormatBreakpointStateAttr<OnOff>;
      type?: FormatBreakpointStateAttr<string>; // Field Options Type.
      id?: FormatBreakpointStateAttr<string>; // Field ID.
      minLength?: FormatBreakpointStateAttr<string>; // Min Input Field Length.
      maxLength?: FormatBreakpointStateAttr<string>; // Max Input Field Length.
      allowedSymbols?: FormatBreakpointStateAttr<string>; // Allowed Input Symbols.
      checkboxOptions?: FormatBreakpointStateAttr<SortableList.Value>; // Checkbox Options.
      radioOptions?: FormatBreakpointStateAttr<SortableList.Value>; // Radio Options.
      selectOptions?: FormatBreakpointStateAttr<SortableList.Value>; // Select Options.
      booleanCheckboxOptions?: FormatBreakpointStateAttr<SortableList.Value>; // Boolean checkbox Options.
      checkboxChecked?: FormatBreakpointStateAttr<string>; // Checkbox Checked.
    },
    innerContent?: FormatBreakpointStateAttr<string>, // Field Title.
  },

  // Field: Normal, Focused, Placeholder.
  field?: FieldElementAttr;

  conditionalLogic?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>; // Enable conditional logic rules.
      relation?: FormatBreakpointStateAttr<OnOff>; // Conditional logic relation (all/any).
    },
    innerContent?: FormatBreakpointStateAttr<ContactFieldCondition[]>, // Conditional logic rules.
  },
}
