// @divi/types/src/module/element/advanced/dividers/index.ts
import { type FormatBreakpointStateAttr } from '../../../../format-attr';


export namespace Dividers {
  export type Arrangement = 'above' | 'below';

  export namespace Component {
    export interface ContainerProps {
      placement: Placement;
      attr: Attr;
      id: string;
    }

    export interface Props {
      placement: Placement;
      style: Style;
    }

    export type Attr = FormatBreakpointStateAttr<AttributeValue>;
  }

  export type Placement = 'top' | 'bottom';

  export type Style = 'none' |
    'arrow' |
    'arrow2' |
    'arrow3' |
    'asymmetric' |
    'asymmetric2' |
    'asymmetric3' |
    'asymmetric4' |
    'clouds' |
    'clouds2' |
    'curve' |
    'curve2' |
    'graph' |
    'graph2' |
    'graph3' |
    'graph4' |
    'mountains' |
    'mountains2' |
    'ramp' |
    'ramp2' |
    'slant' |
    'slant2' |
    'triangle' |
    'wave' |
    'wave2' |
    'waves' |
    'waves2';


  export interface AttributeValue {
    style?: Style;
    color?: string;
    height?: string;
    repeat?: string;
    flip?: string[];
    arrangement?: Arrangement;
  }

  export type Attributes = Partial<Record<Placement, FormatBreakpointStateAttr<AttributeValue>>>;

}
