// @divi/types/src/module/library/button/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface ButtonAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
      alignment?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'boxShadow' |
      'button' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // button
  // buttonText
  // buttonRel
  button?: Element.Types.Button.Attributes;
}
