// @divi/types/src/module/library/testimonial/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface TestimonialAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  quoteIcon?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'icon'
    >;
  };

  portrait?: {
    innerContent?: FormatBreakpointStateAttr<{
      url?: string;
    }>
    decoration?: Element.Decoration.PickedAttributes<
      'border' |
      'boxShadow' |
      'filters' |
      'sizing'
    >;
  };

  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<
      'bodyFont' |
      'inlineFont'
    >;
  };

  author?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  jobTitle?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  company?: {
    innerContent?: FormatBreakpointStateAttr<{
      text?: string;
      linkUrl?: string;
      linkTarget?: 'off' | 'on';
    }>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };
}
