// @divi/types/src/module/settings/group/library/form-field/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type Field  } from '../../../field';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';
import { type Border } from '../border';
import { type BoxShadow } from '../box-shadow';
import { type Font } from '../font';
import { type Spacing } from '../spacing';


export namespace FormField {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Types.Field.Attributes;
    fields?: {
      placeholderFontGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/font';
          props?: Font.GroupProps;
        }
      };
      backgroundColor?: GroupFieldProps & {
        defaultAttr?: Element.Types.Field.Attributes['decoration']['background'];
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      focusBackgroundColor?: GroupFieldProps & {
        defaultAttr?: Element.Types.Field.Attributes['advanced']['focus']['background'];
        component?: FieldLibrary.ComponentMap['divi/color-picker'];
      };
      focusFontGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/font';
          props?: Font.GroupProps;
        }
      };
      fontTopGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/font';
          props?: Font.GroupProps;
        }
      };
      spacingGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/spacing';
          props?: Spacing.GroupProps;
        }
      };
      boxShadowGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/box-shadow';
          props?: BoxShadow.GroupProps;
        }
      };
      borderGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/border';
          props?: Border.GroupProps;
        }
      };
      fontBottomGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/font';
          props?: Font.GroupProps;
        }
      };
      useFocusBorder?: GroupFieldProps & {
        defaultAttr?: Element.Types.Field.Attributes['advanced']['focusUseBorder'];
        component?: FieldLibrary.ComponentMap['divi/toggle'];
      };
      focusBorderGroup?: Field.BaseProps & {
        component?: {
          type: 'group';
          name: 'divi/border';
          props?: Border.GroupProps;
        }
      };
    }
  }
}
