// @divi/types/src/module/library/blurb/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Icon } from '../../../icon';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface BlurbCssAttr extends Css.AttributeValue {
  blurbImage?: string;
  blurbTitle?: string;
  blurbContent?: string;
}

export type BlurbCssGroupAttr = FormatBreakpointStateAttr<BlurbCssAttr>;


export type ImageIconAnimation = 'left' | 'right' | 'top' | 'bottom' | 'off';

export interface Title extends Element.Advanced.Link.AttributeValue {
  text?: string;
}

export interface BlurbAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: BlurbCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'conditions' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  imageIcon?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
      alt?: string;
      useIcon?: string;
      icon?: Icon.Font.AttributeValue;
      animation?: ImageIconAnimation,
    }>;

    advanced?: {
      color?: FormatBreakpointStateAttr<string>;
      placement?: FormatBreakpointStateAttr<string>;
      width?: FormatBreakpointStateAttr<{
        icon?: string;
        image?: string;
      }>;
      alignment?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'border' |
      'boxShadow' |
      'filters' |
      'spacing'
    >;
  };

  title?: Element.Types.TitleLink.Attributes;

  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<
      'bodyFont' |
      'inlineFont'
    >;
  };

  contentContainer?: Element.Types.Wrapper.Attributes<'sizing'>;
}
