// @divi/types/src/page-settings/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Module } from '../../module';


export namespace Store {
  export interface State {
    abBounceRateLimit: string;
    abCurrentShortcode: string;
    abStatsRefreshInterval: string;
    contentAreaBackgroundColor: string;
    customCss: string;
    darkTextColor: string;
    enableAbTesting: string;
    enableShortcodeTracking: string;
    lightTextColor: string;
    overflowX: Module.Element.Decoration.Overflow.OverflowValue;
    overflowY: Module.Element.Decoration.Overflow.OverflowValue;
    pageGutterWidth: string;
    pageZIndex: string;
    postCategories: string;
    postExcerpt: string;
    postImage: string;
    postProjectCategories: string;
    postProjectTags: string;
    postTags: string;
    postTitle: string;
    sectionBackgroundColor: string;
    staticCssFile: string;
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace Update {
      export type ReturnValue = {
        type: 'UPDATE';
        settingKey: string;
        payload: string;
      }
    }

    export type FunctionMap = {
      update: (settingKey: string, payload: string) => Update.ReturnValue;
    };

    export type ReturnValue = Update.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getSetting: (settingPath: string | string[], defaultValue?: string) => string;
      getSettings: () => ImmutableState;
    }
  }
}
