// @divi/types/src/app-preferences/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type AppUi } from '../../app-ui';
import { type DisabledModule } from '../disabled-module';


export namespace Store {
  export interface State {
    app: {
      mode: AppUi.Mode.Type;
      view: AppUi.View.Type;
      colorMode: AppUi.ColorMode;
      colorScheme: AppUi.ColorScheme;
    };
    history: {
      interval: '1' | '10' | '20' | '30' | '40';
    };

    // Previously used, no longer used, but might be used again.
    // interface: {};
    modal: {

      // TODO Revisit this because this might no longer be relevant due to D5 design.
      preference: 'default' | 'minimum' | 'fullscreen' | 'left' | 'right';
      alwaysCollapseGroups: boolean;
    };
    module: {
      dummyContent: boolean;
      disabled: DisabledModule.Visibility;
    };
    pageCreationFlow: {
      onStart: 'default' | 'buildFromScratch' | 'choosePremadeLayout' | 'cloneExistingPage';
    };

    // @todo replace the type with more specific types once Quick Actions feature has been developed.
    quickActions: {
      startWith: string;
      showRecentQueries: string;
      recentQueries: string;
      recentQueriesMaxLength: number;
      recentCategories: string;
      recentCategoriesMaxLength: number;
    };
  };

  export type ImmutableState = ImmutableObject<State & {
    defaults: State
  }>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  };

  export namespace Actions {
    export type FunctionMap = {
      set: <TKey1 extends keyof State, TKey2 extends keyof State[TKey1]>(
        path: [TKey1, TKey2],
        value: State[TKey1][TKey2]
      ) => {
        type: 'SET';
        path: [TKey1, TKey2];
        value: State[TKey1][TKey2];
      }
    };

    export type ReturnValue = ReturnType<FunctionMap['set']>;
  };

  export namespace Selectors {

    /**
     * Interface for getAll function overloading.
     *
     * @since ??
     */
    interface GetAll {
      (): ImmutableState;
      (includeDefault?: false): ImmutableState;
      (includeDefault?: true): ImmutableState;
    }

    export type FunctionMap = {
      get: <TKey1 extends keyof ImmutableState, TKey2 extends keyof ImmutableState[TKey1]>
        (path: [TKey1, TKey2]) => ImmutableState[TKey1][TKey2];
      getDefault: <TKey1 extends keyof ImmutableState['defaults'], TKey2 extends keyof ImmutableState['defaults'][TKey1]>
        (path: [TKey1, TKey2]) => ImmutableState['defaults'][TKey1][TKey2];
      getAll: GetAll;
    }
  }
}
