// @divi/types/src/style-library/index.ts
export namespace StyleLibrary {
  export type TextStyleCssProperties = 'text-align';
  export type TextShadowStyleCssProperties = 'text-shadow';
  export type AnimationStyleCssProperties =
    'animation-delay' |
    'animation-duration' |
    'animation-timing-function' |
    'opacity' |
    'transform';
  export type BackgroundStyleCssProperties =
    'background-blend-mode' |
    'background-color' |
    'background-image' |
    'background-position' |
    'background-repeat' |
    'background-size';
  export type FontStyleCssProperties =
    'color' |
    'font-family' |
    'font-weight' |
    'font-size' |
    'font-style' |
    'font-variant' |
    'letter-spacing' |
    'line-height' |
    'text-align' |
    'text-decoration-line' |
    'text-decoration-color' |
    'text-decoration-style' |
    'text-transform';
  export type ListFontStyleCssProperties = FontStyleCssProperties |
    'list-style-type' |
    'list-style-position' |
    'padding-left';
  export type BlockquoteFontStyleCssProperties = FontStyleCssProperties |
    'border-left-width' |
    'border-left-style' |
    'border-left-color';
  export type BorderStyleShorthandCssProperties =
    'border' |
    'border-radius' |
    'border-style';
  export type BorderStyleCssProperties =
    BorderStyleShorthandCssProperties |
    'border-top-left-radius' |
    'border-top-right-radius' |
    'border-bottom-left-radius' |
    'border-bottom-right-radius' |
    'border-top-style' |
    'border-right-style' |
    'border-bottom-style' |
    'border-left-style' |
    'border-color' |
    'border-top-color' |
    'border-right-color' |
    'border-bottom-color' |
    'border-left-color' |
    'border-width' |
    'border-top-width' |
    'border-right-width' |
    'border-bottom-width' |
    'border-left-width';
  export type BoxShadowCssProperties = 'box-shadow';
  export type FiltersStyleCssProperties = 'filter' | 'mix-blend-mode';
  export type TransformStyleCssProperties =
    'transform' |
    'transform-origin' |
    'transition';
  export type TransitionStyleCssProperties = 'z-index';
  export type DisabledOnStyleCssProperties = 'display' | 'opacity';
  export type IconStyleCssProperties =
    'font-family' |
    'font-size' |
    'font-weight' |
    'line-height' |
    'content' |
    'color' |
    'margin-top' |
    'margin-left';
  export type OverflowStyleCssProperties = 'overflow-x' | 'overflow-y';
  export type PositionStyleCssProperties =
    'position' |
    'top' |
    'right' |
    'bottom' |
    'left' |
    'transform';
  export type SizingStyleCssProperties =
    'width' |
    'max-width' |
    'margin-left' |
    'margin-right' |
    'min-height' |
    'height' |
    'max-height';
  export type SpacingStyleShorthandCssProperties =
    'margin' |
    'padding';
  export type SpacingStyleCssProperties =
    SpacingStyleShorthandCssProperties |
    'margin-top' |
    'margin-right' |
    'margin-bottom' |
    'margin-left' |
    'padding-top' |
    'padding-right' |
    'padding-bottom' |
    'padding-left';
  export type ButtonStyleCssProperties = 'text-align' |
    'left' |
    'margin-left' |
    'right' |
    'margin-right' |
    'color' |
    'content' |
    'display' |
    'font-family' |
    'font-size' |
    'font-weight' |
    'line-height'

     |
    'opacity';
  export type DividersStyleCssProperties =
    'background-color' |
    'background-image' |
    'background-position' |
    'background-size' |
    'bottom' |
    'height' |
    'top' |
    'transform' |
    'z-index';
}
