// @divi/types/src/module/library/pricing-tables/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface PricingTablesCssAttr extends Css.AttributeValue {
  pricingHeading: string;
  pricingTitle: string;
  pricingSubtitle: string;
  pricingTop: string;
  price: string;
  currency: string;
  frequency: string;
  pricingContent: string;
  pricingItem: string;
  pricingItemExcluded: string;
  pricingButton: string;
  featuredTable: string;
}

export type PricingTablesCssGroupAttr = FormatBreakpointStateAttr<PricingTablesCssAttr>;

export interface PricingTablesAttrs extends InternalAttrs {

  // Button
  button?: {
    decoration?: Element.Types.Button.Decoration.Attributes;
  };


  css?: PricingTablesCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'spacing' |
      'transform' |
      'transition' |
      'zIndex' |
      'sizing' |
      'sticky'
    >;
  };

  price?: {
    decoration?: Element.Decoration.PickedAttributes<
      'font' |
      'border' |
      'background'
    >;
  };

  title?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };

  subtitle?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  content?: {
    advanced?: {
      bulletColor?: FormatBreakpointStateAttr<string>;
      showBullet?: FormatBreakpointStateAttr<OnOff>;
    }
    decoration?: Element.Decoration.PickedAttributes<
      'bodyFont' |
      'inlineFont'
    >;
  };

  excluded?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  currencyFrequency?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  featuredTable?: {
    decoration?: Element.Decoration.PickedAttributes<'background'>;
    advanced?: {
      showDropShadow?: FormatBreakpointStateAttr<OnOff>;
    };
  };

  featuredContent?: {
    advanced?: {
      bulletColor?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  featuredTitle?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };

  featuredSubtitle?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  featuredPrice?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };

  featuredCurrencyFrequency?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  featuredExcluded?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Setup button attributes for child pricing tables.
  children?: {
    button?: {
      innerContent?: FormatBreakpointStateAttr<{
        rel?: Element.Types.Button.InnerContent.AttributeValue['rel'];
      }>
    };
  };
}
