// @divi/types/src/module/library/comments/index.ts
import { type OnOff } from '../../../attr';
import {
  type FormatBreakpointStateAttr,
} from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import {
  type FieldElementAttr,
} from '../../options';
import {
  type InternalAttrs,
} from '../internal';


export interface CommentsAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
      showReply?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'animation' |
      'background' |
      'conditions' |
      'disabledOn' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex' |
      'boxShadow' |
      'border' |
      'filters'
    >;
    meta?: Element.Meta.Attributes;
  };

  // Form Field: normal, focused, placeholder.
  field?: FieldElementAttr;

  image?: {
    advanced?: {
      showAvatar?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'boxShadow' |
      'border' |
      'filters'
    >;
  };

  // Button
  button?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'background' |
      'border' |
      'boxShadow' |
      'button' |
      'font' |
      'spacing'
    >;
  };

  // D4 - Comment Count.
  commentCount?: {
    advanced?: {
      showCount?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  // D4 - Form Title.
  formTitle?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  // D4 - Meta.
  meta?: {
    advanced?: {
      showMeta?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  // D4 - Comment.
  commentText?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };
}
