// @divi/types/src/module/element/advanced/index.ts
import { Dividers as DividersNamespace } from './dividers';
import { EmailService as EmailServiceNamespace } from './email-service';
import { Gutter as GutterNamespace } from './gutter';
import { IdClasses as IdClassesNamespace } from './id-classes';
import { Link as LinkNamespace } from './link';
import { SpamProtection as SpamProtectionNamespace } from './spam-protection';
import { Text as TextNamespace } from './text';


export namespace Advanced {
  export import Dividers = DividersNamespace;
  export import Text = TextNamespace;
  export import Gutter = GutterNamespace;
  export import IdClasses = IdClassesNamespace;
  export import Link = LinkNamespace;
  export import SpamProtection = SpamProtectionNamespace;
  export import EmailService = EmailServiceNamespace;

  export type KnownAttributes = {
    dividers?: Dividers.Attributes;
    text?: Text.Attributes;
    gutter?: Gutter.Attributes;
    htmlAttributes?: IdClasses.Attributes;
    link?: Link.Attributes;
    spamProtection?: SpamProtection.Attributes;
    emailService?: EmailService.Attributes;
  }

  export type Attributes<TAttr = KnownAttributes & Record<string, unknown>> = TAttr;
}
