// @divi/types/src/module/library/gallery/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Icon } from '../../../icon';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface GalleryAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      text?: Element.Advanced.Text.Attributes;
      fullwidth?: FormatBreakpointStateAttr<string>;
      postsNumber?: FormatBreakpointStateAttr<string>;
      showTitleAndCaption?: FormatBreakpointStateAttr<string>;
      auto?: FormatBreakpointStateAttr<OnOff>;
      autoSpeed?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'animation' |
      'background' |
      'conditions' |
      'disabledOn' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex' |
      'boxShadow' |
      'filters'
    >;
  };
  title?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  caption?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  pagination?: {
    advanced?: {
      showPagination?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  item?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'border'
    >;
  };

  image?: {
    advanced?: {
      galleryCaptions?: FormatBreakpointStateAttr<string>;
      galleryIds?: FormatBreakpointStateAttr<string>;
      galleryOrderby?: FormatBreakpointStateAttr<string>;
      orientation?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'border' |
      'boxShadow' |
      'filters'
    >;
  };

  overlay?: {
    innerContent?: FormatBreakpointStateAttr<{
      icon?: Icon.Font.AttributeValue;
    }>;
    advanced?: {
      zoomIconColor?: FormatBreakpointStateAttr<string>;
      hoverOverlayColor?: FormatBreakpointStateAttr<string>;
    };
  };
}
