// @divi/types/src/app-ui/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import {
  type AttrState,
  type Breakpoint,
} from '../../format-attr';
import { type Modal } from '../../modal';
import { type ModalLibrary } from '../../modal-library';
import { type Docking } from '../docking';
import { type Element } from '../element';
import { type Mode } from '../mode';
import {
  type Sidebar,
  type SidebarRow,
} from '../sidebar';
import { type View } from '../view';
import { type Window } from '../window';


export namespace Store {
  export interface State {
    view: View.Type;
    breakpoint: Breakpoint;
    mode: Mode.Type;
    views: View.Items;
    viewsOrder: string[];
    modes: Mode.Items;
    modeInteractions: Record<Mode.InteractionLocation, Mode.Interaction>;
    attributeState: AttrState;
    window: {
      width: Window.PropertyValues;
      height: Window.PropertyValues;
      pageXOffset: Window.PropertyValues;
      pageYOffset: Window.PropertyValues;
      documentWidth: Window.PropertyValues;
      documentHeight: Window.PropertyValues;
    };
    windowAutoScroll: Window.AutoScroll;
    adminBarHeight: {
      top: number;
      app: number;
    };
    globalPreload: number;
    elements: Element.Items;

    /**
     * How layout are rendered in the AppFrame.
     * `fullscreen`: layout is rendered and use maximum available dimension after sidebars and builderBar are rendered.
     * `responsive`: layout is rendered in the area of pre-defined window dimension.
     */
    frameMode: 'fullscreen' | 'responsive';
    sidebarRows?: Sidebar.RowItems;

    sidebarHoveredDockArea?: false | {
      id: Docking.Area['id'];
      type: Docking.Area['type'];
    };

    // State of modal that is currently dragged to be undocked from sidebar.
    sidebarModalDraggedToUndock?: false | {
      sidebarRowId: string;
      modalName: string;
    },
    moduleInteraction: {
       status: 'on' | 'off';
       event: EventTarget | null;
    }
  }

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  }

  export namespace Actions {
    export namespace AddSidebarRow {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'ADD_SIDEBAR_ROW';
        modal: string;
        order: number;
        sidebarPosition: Sidebar.Position;
        newSidebarRowId?: string;
      }
    }

    export namespace DockModalToSidebar {
      export type Params = Omit<ReturnValue, 'type'>;

      /**
       * Interface of `dockModalToSidebar()`'s returned value.
       */
      export type ReturnValue = {
        type: 'DOCK_MODAL_TO_SIDEBAR';
        modalName: string;
        modalType: Modal.Type;
        sidebarName: Sidebar.Name;
        newSidebarRowId?: string;
      }
    }

    export namespace DockModalToSidebarRow {
      export type Params = Omit<ReturnValue, 'type'>;

      /**
       * Interface of `dockModalToSidebarRow()`'s returned value.
       */
      export type ReturnValue = {
        type: 'DOCK_MODAL_TO_SIDEBAR_ROW';
        modalName: string;
        modalType: Modal.Type;
        newSidebarRowId?: string;
        position: 'above' | 'inside' | 'below';
        sidebarRowId: string;
      }
    }

    export namespace RemoveSidebarRow {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'REMOVE_SIDEBAR_ROW';
        sidebarName: Sidebar.Name;
        sidebarRowId: string;
      }
    }

    export namespace ReplaceSidebarModal {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'REPLACE_SIDEBAR_MODAL';
        modalName: string;
        overwrittenModalName: string;
      }
    }

    export namespace ResetState {
      export type ReturnValue = {
        type: 'RESET_STATE';
      }
    }

    export namespace SetAdminBarHeight {
      export type ReturnValue = {
        type: 'SET_ADMIN_BAR_HEIGHT';
        value: number;
        windowLocation: Window.Location;
      }
    }

    export namespace SetAttributeState {
      export type ReturnValue = {
        type: 'SET_ATTRIBUTE_STATE';
        attributeState: AttrState;
      }
    }

    export namespace SetDraggedToUndockSidebarModal {

      /**
       * Interface of `setDraggedToUndockSidebarModal()`'s returned value.
       */
      export type ReturnValue = {
        type: 'SET_DRAGGED_TO_UNPIN_SIDEBAR_MODAL';
        modalName: false | string;
      }
    }

    export namespace SetElementProperty {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_ELEMENT_PROPERTY';
        elementName: Element.Name;
        propertyGroup: Element.StyleProperty.Group;
        propertyName: Element.StyleProperty.Name;
        value: Element.StyleProperty.Value;
      }
    }

    export namespace SetElementVisibility {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_ELEMENT_VISIBILITY';
        elementName: keyof Element.Items;
        visibility: boolean;
      }
    }

    export namespace SetGlobalPreload {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_GLOBAL_PRELOAD';
        globalPreload: boolean;
      }
    }

    export namespace SetHoveredSidebarDockingArea {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_HOVERED_SIDEBAR_DOCKING_AREA';
        sidebarDockingArea: false | {
          id: Docking.Area['id'];
          type: Docking.Area['type'];
        }
      }
    }

    export namespace SetMode {
      export type ReturnValue = {
        type: 'SET_MODE';
        mode: Mode.Type;
      }
    }

    export namespace SetSidebarRowActiveModal {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_SIDEBAR_ROW_ACTIVE_MODAL';
        activeModal: string;
        sidebarRowId: string;
      }
    }

    export namespace SetSidebarRowHeight {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_SIDEBAR_ROW_HEIGHT';
        height: number;
        sidebarRowId: string;
      }
    }

    export namespace SetSidebarRowResizingStatus {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_SIDEBAR_ROW_RESIZING_STATUS';
        sidebarRowId: string;
        status: boolean;
      }
    }

    export namespace SetView {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_VIEW';
        view: string;
      }
    }

    export namespace SetWindow {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'SET_WINDOW';
        attr: string;
        value: number;
        windowLocation: Window.Location;
      }
    }

    export namespace SetWindowAutoScroll {
      export type WindowScrollDirection = false | 'top' | 'bottom';

      export type ReturnValue = {
        type: 'SET_WINDOW_AUTO_SCROLL';
        direction: WindowScrollDirection;
        distance: number;
        location: string;
      }
    }

    export namespace UndockModalFromSidebarRow {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'UNDOCK_MODAL_FROM_SIDEBAR_ROW';
        modalName: string;
        sidebarRowId: string;
      }
    }

    export namespace UpdateSidebarRowLimits {
      export type ReturnValue = {
        type: 'UPDATE_SIDEBAR_ROW_LIMITS';
        sidebarRowId: string;
      }
    }

    export namespace ToggleModuleInteraction {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'TOGGLE_MODULE_INTERACTION';
        status: 'on' | 'off';
      }
    }

    export namespace OnModuleInteractionEvent {
      export type Params = Omit<ReturnValue, 'type'>;

      export type ReturnValue = {
        type: 'ON_MODULE_INTERACTION_EVENT';
        event: EventTarget | null;
      }
    }

    export type FunctionMap = {
      addSidebarRow: (params: AddSidebarRow.Params) => AddSidebarRow.ReturnValue;
      dockModalToSidebar: (params: DockModalToSidebar.Params) => DockModalToSidebar.ReturnValue;
      dockModalToSidebarRow: (params: DockModalToSidebarRow.Params) => DockModalToSidebarRow.ReturnValue;
      onModuleInteractionEvent: (event: OnModuleInteractionEvent.Params['event']) => OnModuleInteractionEvent.ReturnValue;
      removeSidebarRow: (params: RemoveSidebarRow.Params) => RemoveSidebarRow.ReturnValue;
      replaceSidebarModal: (params: ReplaceSidebarModal.Params) => ReplaceSidebarModal.ReturnValue;
      resetState: () => ResetState.ReturnValue;
      setAdminBarHeight: (
        value: number,
        windowLocation: Window.Location,
      ) => SetAdminBarHeight.ReturnValue;
      setAttributeState: (attributeState: AttrState) => SetAttributeState.ReturnValue;
      setDraggedToUndockSidebarModal: (
        modalName: SetDraggedToUndockSidebarModal.ReturnValue['modalName'],
      ) => SetDraggedToUndockSidebarModal.ReturnValue;
      setElementProperty: (params: SetElementProperty.Params) => SetElementProperty.ReturnValue;
      setElementVisibility: (params: SetElementVisibility.Params) => SetElementVisibility.ReturnValue;
      setGlobalPreload: (isActive: boolean) => SetGlobalPreload.ReturnValue;
      setHoveredSidebarDockingArea: (
        params: SetHoveredSidebarDockingArea.Params
      ) => SetHoveredSidebarDockingArea.ReturnValue;
      setMode: (mode: Mode.Type) => SetMode.ReturnValue;
      setSidebarRowActiveModal: (params: SetSidebarRowActiveModal.Params) => SetSidebarRowActiveModal.ReturnValue;
      setSidebarRowHeight: (params: SetSidebarRowHeight.Params) => SetSidebarRowHeight.ReturnValue;
      setSidebarRowResizingStatus: (
        params: SetSidebarRowResizingStatus.Params
      ) => SetSidebarRowResizingStatus.ReturnValue;
      setView: (view: string) => SetView.ReturnValue;
      setWindow: (params: SetWindow.Params) => SetWindow.ReturnValue;
      setWindowAutoScroll: (
        direction: SetWindowAutoScroll.WindowScrollDirection,
        distance?: number,
        location?: string,
      ) => SetWindowAutoScroll.ReturnValue;
      toggleModuleInteraction: (status: ToggleModuleInteraction.Params['status']) => ToggleModuleInteraction.ReturnValue;
      undockModalFromSidebarRow: (params: UndockModalFromSidebarRow.Params) => UndockModalFromSidebarRow.ReturnValue;
      updateSidebarRowLimits: (sidebarRowId: string) => UpdateSidebarRowLimits.ReturnValue;
    };

    export type ReturnValue =
      AddSidebarRow.ReturnValue |
      DockModalToSidebar.ReturnValue |
      DockModalToSidebarRow.ReturnValue |
      OnModuleInteractionEvent.ReturnValue |
      RemoveSidebarRow.ReturnValue |
      ReplaceSidebarModal.ReturnValue |
      ResetState.ReturnValue |
      SetAdminBarHeight.ReturnValue |
      SetAttributeState.ReturnValue |
      SetDraggedToUndockSidebarModal.ReturnValue |
      SetElementProperty.ReturnValue |
      SetElementVisibility.ReturnValue |
      SetGlobalPreload.ReturnValue |
      SetHoveredSidebarDockingArea.ReturnValue |
      SetMode.ReturnValue |
      SetSidebarRowActiveModal.ReturnValue |
      SetSidebarRowHeight.ReturnValue |
      SetSidebarRowResizingStatus.ReturnValue |
      SetView.ReturnValue |
      SetWindow.ReturnValue |
      SetWindowAutoScroll.ReturnValue |
      ToggleModuleInteraction.ReturnValue |
      UndockModalFromSidebarRow.ReturnValue |
      UpdateSidebarRowLimits.ReturnValue;
  };

  export namespace Selectors {
    export namespace GetElementProperty {
      export type Params = {
        elementName: Element.Name;
        propertyGroup: Element.StyleProperty.Group;
        propertyName: Element.StyleProperty.Name;
      }

      export type ReturnValue = Element.StyleProperty.Value;
    }

    export type FunctionMap = {
      getAdminBarHeight: (windowLocation: 'top' | 'app') => number;
      getAttributeState: () => AttrState;
      getBreakpoint: () => Breakpoint;
      getElementProperty: (params: GetElementProperty.Params) => GetElementProperty.ReturnValue;
      getElements: () => Element.Items;
      getFrameMode: () => Store.State['frameMode'];
      getGlobalPreload: () => boolean;
      getHoveredDockingArea: (
        modalOffset: Sidebar.Offset,
        snapAreas: Docking.Areas
      ) => false | Docking.Area;
      getHoveredSidebarDockingAreaId: () => false | string;
      getMode: () => Mode.Type;
      getModes: () => Mode.Items;
      getSidebarModalDimension: (modalName: string, modalType: ModalLibrary.Store.ModalDefinition['type']) => Sidebar.Modal.Dimension;
      getSidebarModalInfo: (modalName: string) => Sidebar.Modal.Info;
      getDockingAreas: (draggedModal: {
        type: Modal.Type;
        width: number;
      }) => Docking.Areas;
      getDraggedToUndockSidebarModal: () => Store.ImmutableState['sidebarModalDraggedToUndock'];
      getModuleInteractionEventTarget: () => ImmutableState['moduleInteraction']['event'];
      getModuleInteractionStatus: () => ImmutableState['moduleInteraction']['status'];
      getSidebarProps: (position: Sidebar.Position) => {
        rowIds: Sidebar.Props['rowIds'];
        dimension: Element.Item['dimension'];
        offset: Element.Item['offset'];
        dimensionMinimum?: Element.Item['dimension'];
      };
      getSidebarRow: (sidebarRowId: string) => Sidebar.RowItem;
      getSidebarRowHeight: (rowId: string) => number;
      getSidebarRowDimensionLimits: (sidebarRowId: string, property: keyof Sidebar.Dimension) => {
        max: number;
        min: number;
      };
      getSidebarRowOffset: (sidebarRowId: string) => Sidebar.Offset;
      getSidebarRowsHeight: (rowIds: string[]) => Record<string, number | null>;
      getSidebarRowSidebar: (sidebarRowId: string) => Sidebar.Name;
      getHoveredSidebarRowDockingArea: (sidebarRowId: string) => SidebarRow.DockingArea;
      getView: () => View.Type;
      getViewItem: (view: View.Type) => View.Item;
      getViews: () => View.Items;
      getViewsOrder: () => View.Type[];
      getWindow: (attr: string, windowLocation: string, defaultValue?: string | number) => number;
      getWindowAutoScroll: () => Window.AutoScroll;
      hasModeInteraction: (interaction: [Mode.InteractionLocation, keyof Mode.Interaction]) => boolean;
      isSidebarModal: (modalName: string) => boolean;
    }
  }
}
