// @divi/types/src/help/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Video } from '../video';


export namespace Store {
  export interface State {
    overlayMode: boolean,
    currentListName: string,
    currentVideo: number,
    videos: Video.Registry,
  };

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  }

  export namespace Actions {
    export namespace AddVideoList {

      /**
       * Action type for adding a video list to the store.
       *
       * @since ??
       *
       * @typedef {object} Help.Store.Actions.AddVideoList.ReturnValue
       *
       * @property {string} type - The action type.
       * @property {string} listName - The name of the list.
       * @property {Video.Items} payload - The video list.
       */
      export type ReturnValue = {
        type: 'ADD_VIDEO_LIST';
        listName: string;
        payload: Video.Items;
      }
    }

    export namespace SetCurrentListName {

      /**
       * Action type for setting the current video list name.
       *
       * @since ??
       *
       * @typedef {object} Help.Store.Actions.SetCurrentListName.ReturnValue
       *
       * @property {string} type - The action type.
       * @property {string} payload - The name of the current video list.
       */
      export type ReturnValue = {
        type: 'SET_CURRENT_LIST_NAME';
        payload: State['currentListName'];
      }
    }

    export namespace SetCurrentVideo {

      /**
       * Action type for setting the current video.
       *
       * @since ??
       *
       * @typedef {object} Help.Store.Actions.SetCurrentVideo.ReturnValue
       *
       * @property {string} type The action type.
       * @property {number} payload The id of the current video.
       */
      export type ReturnValue = {
        type: 'SET_CURRENT_VIDEO';
        payload: State['currentVideo'];
      }
    }

    export namespace SetJumpToVideo {

      /**
       * Set jump to video action type.
       *
       * @since ??
       *
       * @typedef {object} Help.Store.Actions.SetJumpToVideo.ReturnValue
       *
       * @property {string} type - The action type.
       * @property {number} payload - The video jumps.
       */
      export type ReturnValue = {
        type: 'SET_JUMP_TO_VIDEO';
        payload: number;
      }
    }

    export namespace SetOverlayMode {

      /**
       * Action type for setting the overlay mode.
       *
       * @since ??
       *
       * @typedef {object} Help.Store.Actions.SetOverlayMode.ReturnValue
       *
       * @property {string} type The action type.
       * @property {boolean} payload The status of the overlay.
       */
      export type ReturnValue = {
        type: 'SET_OVERLAY_MODE';
        payload: State['overlayMode'];
      }

    }

    export type FunctionMap = {
      addVideoList: (
        listName: AddVideoList.ReturnValue['listName'],
        payload: AddVideoList.ReturnValue['payload'],
      ) => AddVideoList.ReturnValue,
      setCurrentListName: (
        payload: SetCurrentListName.ReturnValue['payload'],
      ) => SetCurrentListName.ReturnValue,
      setCurrentVideo: (
        payload: SetCurrentVideo.ReturnValue['payload'],
      ) => SetCurrentVideo.ReturnValue,
      setJumpToVideo: (
        payload: SetJumpToVideo.ReturnValue['payload'],
      ) => SetJumpToVideo.ReturnValue,
      setOverlayMode: (
        payload: SetOverlayMode.ReturnValue['payload'],
      ) => SetOverlayMode.ReturnValue,
    }

    export type ReturnValue =
      AddVideoList.ReturnValue |
      SetCurrentListName.ReturnValue |
      SetCurrentVideo.ReturnValue |
      SetJumpToVideo.ReturnValue |
      SetOverlayMode.ReturnValue;
  }

  export namespace Selectors {
    export type FunctionMap = {
      getCurrentListName: () => ImmutableState['currentListName'];
      getCurrentVideo: () => ImmutableState['currentVideo'];
      getOverlayMode: () => ImmutableState['overlayMode'];
      getVideos: () => ImmutableState['videos'];
    }
  }
}
