// @divi/types/src/field-library/common-numeric-range/index.ts
import { type NumericInput } from '../common-numeric-input';


export namespace NumericRange {

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string[]} [allowedUnits] Allowed units.
   * @property {string} [className] Class name.
   * @property {string} [defaultUnit] Default unit.
   * @property {string} [defaultValue] Default value.
   * @property {number} [maxRange] Max range.
   * @property {number} [minRange] Min range.
   * @property {number} [maxNumeric] Max numeric.
   * @property {number} [minNumeric] Min numeric.
   * @property {boolean} ["readonly"] Read only.
   * @property {number} [step] Step.
   * @property {object} value Value.
   * @property {Array} [valueMapping] Value mapping.
   * @property {Function} [formatter] Formatter.
   * @property {Function} [onChange] On change.
   */
  export interface Props {
    name: string;
    allowedUnits?: string[];
    className?: string;
    defaultUnit?: string;
    defaultValue?: string;
    maxRange?:number | null;
    minRange?:number | null;
    maxNumeric?:number | null;
    minNumeric?:number | null;
    readonly?: boolean;
    step?:number;
    value: string;
    valueMapping?: NumericInput.ValueMapping[];
    formatter?: NumericInput.Formatter;
    onChange?: (valueNew: string, updateSource: string) => void;
  }
}
