// @divi/types/src/module/element/style/index.ts
import { type AdvancedStyle } from '../advanced-styles';
import { type Decoration } from '../decoration';


export namespace Style {

  // TODO replace `ElementStyleExcludedStyleProps` with this.
  export type ExcludedProps = 'attr' | 'asStyle' | 'mode';


  // TODO Replace `ElementType` with this
  export type Type = 'module' | 'element' | 'button';

  // TODO replace `ElementStyleProps` with this.
  export interface Props {

    // General Props.
    selector?: string;
    attrs: Decoration.Attributes;
    defaultPrintedStyleAttrs?: Decoration.Attributes;
    attrsFilter?: (attrs: Decoration.Attributes) => Decoration.Attributes;
    type?: Type;
    advancedStyles?: AdvancedStyle.Item[];

    // Style specific settings
    animation?: Omit<Decoration.Animation.Style.Props, ExcludedProps>;
    background?: Omit<Decoration.Background.Style.Props, ExcludedProps>;
    bodyFont?: Omit<Decoration.BodyFont.Style.Props, ExcludedProps>;
    border?: Omit<Decoration.Border.Style.Props, ExcludedProps>;
    boxShadow?: Omit<Decoration.BoxShadow.Style.Props, ExcludedProps>;
    filters?: Omit<Decoration.Filters.Style.Props, ExcludedProps>;
    transform?: Omit<Decoration.Transform.Style.Props, ExcludedProps>;
    transition?: Omit<Decoration.Transition.Style.Props, ExcludedProps>;
    disabledOn?: Omit<Decoration.DisabledOn.Style.Props, ExcludedProps>;
    font?: Omit<Decoration.Font.Style.Props, ExcludedProps>;
    headingFont?: Omit<Decoration.HeadingFont.Style.Props, ExcludedProps>;
    icon?: Omit<Decoration.Icon.Style.Props, ExcludedProps>;
    overflow?: Omit<Decoration.Overflow.Style.Props, ExcludedProps>;
    position?: Omit<Decoration.Position.Style.Props, ExcludedProps>;
    sizing?: Omit<Decoration.Sizing.Style.Props, ExcludedProps>;
    spacing?: Omit<Decoration.Spacing.Style.Props, ExcludedProps>;
    zIndex?: Omit<Decoration.ZIndex.Style.Props, ExcludedProps>;
    button?: Omit<Decoration.Button.Style.Props, ExcludedProps>;
  }
}
