// @divi/types/src/field-library/index.ts
import { BorderRadius as BorderRadiusNamespace } from './border-radius';
import { BorderStylesPreview as BorderStylesPreviewNamespace } from './border-styles-preview';
import { ButtonOptions as ButtonOptionsNamespace } from './button-options';
import { Checkboxes as CheckboxesNamespace } from './checkboxes';
import { Code as CodeNamespace } from './code';
import { ColorPicker as ColorPickerNamespace } from './color-picker';
import { NumericInput as NumericInputNamespace } from './common-numeric-input';
import { NumericRange as NumericRangeNamespace } from './common-numeric-range';
import { NumericSpinner as NumericSpinnerNamespace } from './common-numeric-spinner';
import { Preview as PreviewNamespace } from './common-preview';
import { Conditions as ConditionsNamespace } from './conditions';
import { CornerPicker as ConerPickerNamespace } from './corner-picker';
import { CornerSidePicker as CornerSidePickerNamespace } from './cornerside-picker';
import { DraggableList as DraggableListNamespace } from './draggable-list';
import { EmailServiceAccount as EmailServiceAccountNamespace } from './email-service-account';
import { EmailServiceProvider as EmailServiceProviderNamespace } from './email-service-provider';
import { TextGoogleAPIKey as TextGoogleAPIKeyNamespace } from './google-api-key-field';
import { Gradient as GradientNamespace } from './gradient';
import { IconPicker as IconPickerNamespace } from './icon-picker';
import { Map as MapNamespace } from './map';
import { Range as RangeNamespace } from './range';
import { RichText as RichTextNamespace } from './richtext';
import { ScrollEffect as ScrollEffectNamespace } from './scroll-effect';
import { Select as SelectNamespace } from './select';
import { SelectColumn as SelectColumnNamespace } from './select-column';
import { SelectDivider as SelectDividerNamespace } from './select-divider';
import { SelectFont as SelectFontNamespace } from './select-font';
import { SelectMask as SelectMaskNamespace } from './select-mask';
import { SelectMenu as SelectMenuNamespace } from './select-menu';
import { SelectPattern as SelectPatternNamespace } from './select-pattern';
import { SelectWidgetArea as SelectWidgetAreaNamespace } from './select-widget-area';
import { SortableList as SortableListNamespace } from './sortable-list';
import { Spacing as SpacingNamespace } from './spacing';
import { SpamProtectionAccount as SpamProtectionAccountNamespace } from './spam-protection-account';
import { SpamProtectionProvider as SpamProtectionProviderNamespace } from './spam-protection-provider';
import { Text as TextNamespace } from './text';
import { TextArea as TextAreaNamespace } from './textarea';
import { Timepicker as TimepickerNamespace } from './timepicker';
import { Toggle as ToggleNamespace } from './toggle';
import { TransformOrigin as TransformOriginNamespace } from './transform-origin';
import { TransformRotate as TransformRotateNamespace } from './transform-rotate';
import { TransformScale as TransformScaleNamespace } from './transform-scale';
import { TransformSkew as TransformSkewNamespace } from './transform-skew';
import { TransformTranslate as TransformTranslateNamespace } from './transform-translate';
import { Upload as UploadNamespace } from './upload';
import { UploadGallery as UploadGalleryNamespace } from './upload-gallery';
import { Warning as WarningNamespace } from './warning';


export namespace FieldLibrary {
  export import BorderRadius = BorderRadiusNamespace;
  export import BorderStylesPreview = BorderStylesPreviewNamespace;
  export import ButtonOptions = ButtonOptionsNamespace;
  export import Checkboxes = CheckboxesNamespace;
  export import Code = CodeNamespace;
  export import ColorPicker = ColorPickerNamespace;
  export import Conditions = ConditionsNamespace;
  export import CornerPicker = ConerPickerNamespace;
  export import CornerSidePicker = CornerSidePickerNamespace;
  export import DraggableList = DraggableListNamespace;
  export import EmailServiceAccount = EmailServiceAccountNamespace;
  export import EmailServiceProvider = EmailServiceProviderNamespace;
  export import Gradient = GradientNamespace;
  export import IconPicker = IconPickerNamespace;
  export import MapField = MapNamespace;
  export import NumericInput = NumericInputNamespace;
  export import NumericRange = NumericRangeNamespace;
  export import NumericSpinner = NumericSpinnerNamespace;
  export import Preview = PreviewNamespace;
  export import Range = RangeNamespace;
  export import RichText = RichTextNamespace;
  export import ScrollEffect = ScrollEffectNamespace;
  export import Select = SelectNamespace;
  export import SelectColumn = SelectColumnNamespace;
  export import SelectDivider = SelectDividerNamespace;
  export import SelectFont = SelectFontNamespace;
  export import SelectMask = SelectMaskNamespace;
  export import SelectMenu = SelectMenuNamespace;
  export import SelectPattern = SelectPatternNamespace;
  export import SelectWidgetArea = SelectWidgetAreaNamespace;
  export import SortableList = SortableListNamespace;
  export import Spacing = SpacingNamespace;
  export import SpamProtectionAccount = SpamProtectionAccountNamespace;
  export import SpamProtectionProvider = SpamProtectionProviderNamespace;
  export import Text = TextNamespace;
  export import TextArea = TextAreaNamespace;
  export import TextGoogleAPIKey = TextGoogleAPIKeyNamespace;
  export import Timepicker = TimepickerNamespace;
  export import Toggle = ToggleNamespace;
  export import TransformOrigin = TransformOriginNamespace;
  export import TransformRotate = TransformRotateNamespace;
  export import TransformScale = TransformScaleNamespace;
  export import TransformSkew = TransformSkewNamespace;
  export import TransformTranslate = TransformTranslateNamespace;
  export import Upload = UploadNamespace;
  export import UploadGallery = UploadGalleryNamespace;
  export import Warning = WarningNamespace;

  export interface Map {
    'divi/border-radius': {
      props: BorderRadius.Props;
      steps: BorderRadius.Steps;
      inputValue: BorderRadius.InputValue;
    };
    'divi/border-styles-preview': {
      props: BorderStylesPreview.Props;
      property: BorderStylesPreview.Property;
      edge: BorderStylesPreview.Edge;
      edgeAll: BorderStylesPreview.EdgeAll;
      style: BorderStylesPreview.Style;
    };
    'divi/button-options': {
      props: ButtonOptions.Props;
      onChangeCallbackParams: ButtonOptions.OnChangeCallbackParams;
    };
    'divi/checkboxes': {
      props: Checkboxes.Props;
    };
    'divi/code': {
      props: Code.Props;
    };
    'divi/color-picker': {
      props: ColorPicker.Props;
      containerProps: ColorPicker.ContainerProps;
    };
    'divi/common-numeric-input': {
      props: NumericInput.Props;
      valueNumberUnit: NumericInput.ValueNumberUnit;
      valueMapping: NumericInput.ValueMapping;
    };
    'divi/common-numeric-range': {
      props: NumericRange.Props;
    };
    'divi/common-numeric-spinner': {
      props: NumericSpinner.Props;
    };
    'divi/common-preview': {
      props: Preview.Props;
    };
    'divi/conditions': {
      props: Conditions.Props;
    };
    'divi/corner-picker': {
      props: CornerPicker.Props;
    };
    'divi/cornerside-picker': {
      props: CornerSidePicker.Props;
      optionValue: CornerSidePicker.OptionValue;
    };
    'divi/draggable-list': {
      props: DraggableList.Props;
      value: DraggableList.Value;
      onChangeCallbackParams: DraggableList.OnChangeCallbackParams;
    };
    'divi/email-service-account': {
      props: EmailServiceAccount.Props;
    };
    'divi/email-service-provider': {
      props: EmailServiceProvider.Props;
    };
    'divi/gradient': {
      containerProps: Gradient.ContainerProps;
      props: Gradient.Props;
    };
    'divi/icon-picker': {
      props: IconPicker.Props;
      actionArgs: IconPicker.ActionArgs;
      shortcutsArgs: IconPicker.ShortcutsArgs;
    };
    'divi/map': {
      props: MapField.Props;
    };
    'divi/range': {
      props?: Range.Props;
    };
    'divi/richtext': {
      props?: RichText.Props;
    };
    'divi/scroll-effect': {
      props: ScrollEffect.Props;
    };
    'divi/select': {
      option: Select.Option;
      props: Select.Props;
      onChangeCallbackParams: Select.OnChangeCallbackParams;
      updateCallbackParams: Select.UpdateCallbackParams;
    };
    'divi/select-column': {
      props: SelectColumn.Props;
    };
    'divi/select-divider': {
      props: SelectDivider.Props;
      containerProps: SelectDivider.ContainerProps;
    };
    'divi/select-font': {
      props: SelectFont.Props;
    };
    'divi/select-mask': {
      props: SelectMask.Props;
      containerProps: SelectMask.ContainerProps;
    };
    'divi/select-menu': {
      props: SelectMenu.Props;
    };
    'divi/select-pattern': {
      props: SelectPattern.Props;
      containerProps: SelectPattern.ContainerProps;
    };
    'divi/select-widget-area': {
      props: SelectWidgetArea.Props;
    };
    'divi/spacing': {
      props: Spacing.Props;
      sideName: Spacing.SideName;
      sideNameHorizontal: Spacing.SideNameHorizontal;
      sideNameVertical: Spacing.SideNameVertical;
      steps: Spacing.Steps;
      inputValue: Spacing.InputValue;
    };
    'divi/sortable-list': {
      props: SortableList.Props;
    };
    'divi/spam-protection-account': {
      props: SpamProtectionAccount.Props;
    };
    'divi/spam-protection-provider': {
      props: SpamProtectionProvider.Props;
    };
    'divi/text': {
      props: Text.Props;
      onChangeCallbackParams: Text.OnChangeCallbackParams;
    };
    'divi/textarea': {
      props: TextArea.Props;
    };
    'divi/google-api-key-field': {
      props: TextGoogleAPIKey.Props;
    };
    'divi/timepicker': {
      props: Timepicker.Props;
    };
    'divi/toggle': {
      props: Toggle.Props;
      onChangeCallbackParams: Toggle.OnChangeCallbackParams;
    };
    'divi/transform-origin': {
      props: TransformOrigin.Props;
      axis: TransformOrigin.Axis;
      step: TransformOrigin.Step;
      value: TransformOrigin.Value;
      valueMappings: TransformOrigin.ValueMappings;
      valuePreset: TransformOrigin.ValuePreset;
      onChangeCallbackParams: TransformOrigin.OnChangeCallbackParams;
    };
    'divi/transform-rotate': {
      props: TransformRotate.Props;
      axis: TransformRotate.Axis;
      step: TransformRotate.Step;
      value: TransformRotate.Value;
      onChangeCallbackParams: TransformRotate.OnChangeCallbackParams;
    };
    'divi/transform-scale': {
      props: TransformScale.Props;
      value: TransformScale.Value;
      step: TransformScale.Step;
      axis: TransformScale.Axis;
      onChangeCallbackParams: TransformScale.OnChangeCallbackParams;
      dragBoxSize: TransformScale.DragBoxSize;
    };
    'divi/transform-skew': {
      props: TransformSkew.Props;
      axis: TransformSkew.Axis;
      value: TransformSkew.Value;
      step: TransformSkew.Step;
      onChangeCallbackParams: TransformSkew.OnChangeCallbackParams;
    };
    'divi/transform-translate': {
      props: TransformTranslate.Props;
      axis: TransformTranslate.Axis;
      value: TransformTranslate.Value;
      step: TransformTranslate.Step;
      onChangeCallbackParams: TransformTranslate.OnChangeCallbackParams;
    };
    'divi/upload': {
      props: Upload.Props;
      containerProps: Upload.ContainerProps;
      onChangeCallback: Upload.OnChangeCallback;
      onChangeCallbackParams: Upload.OnChangeCallbackParams;
      attributeSize: Upload.AttributeSize;
      attributeSizes: Upload.AttributeSizes;
    };
    'divi/upload-gallery': {
      props: UploadGallery.Props;
      onChangeCallbackParams: UploadGallery.OnChangeCallbackParams;
    };
    'divi/warning': {
      props: Warning.Props;
    };
  }

  /**
   * Type of field library component in form of maps.
   * Provide `name` property to be used as Discriminating Unions in `FieldLibrary.Component` later.
   * `type` prop allows another type of component (eg., `group`) to be used in the same loop as
   * field (eg. Font Options).
   */
  export type ComponentMap = {
    [TName in keyof Map]?: {
      type?: 'field';
      name?: TName;
      props?: Partial<Map[TName]['props']>;
    };
  };

  export type Names = keyof Map;

  export type Component = ComponentMap[Names];
}
