// @divi/types/src/field-library/transform-rotate/index.ts
import { type NumericInput } from '../common-numeric-input';


export namespace TransformRotate {

  /**
   * @typedef {('x'|'y'|'z')} Axis
   */
  export type Axis = 'x' | 'y' | 'z';

  /**
   * @typedef {object} Value
   * @property {string} [x] X value.
   * @property {string} [y] Y value.
   * @property {string} [z] Z value.
   */
  export interface Value {
    x?: string;
    y?: string;
    z?: string;
  }

  /**
   * @typedef {object} Step
   * @property {number} x X value.
   * @property {number} y Y value.
   * @property {number} z Z value.
   */
  export interface Step {
    x: number;
    y: number;
    z: number;
  }

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {object} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    inputValue?: Value;
  }

  /**
   * @typedef {object} Props
   * @property {string[]} [allowedUnits] Allowed units.
   * @property {string} [className] Class name.
   * @property {string} [defaultUnit] Default unit.
   * @property {object} [defaultValue] Default value.
   * @property {Function} [formatter] Formatter.
   * @property {number} [max] Max value.
   * @property {number} [min] Min value.
   * @property {number} [maxResetStepValue] Max reset step value.
   * @property {number} [minResetStepValue] Min reset step value.
   * @property {string} [name] Name.
   * @property {boolean} ["readonly"] Read only.
   * @property {object} value Value.
   * @property {Function} onChange On change callback.
   */
  export interface Props {
    allowedUnits?: string[];
    className?: string;
    defaultUnit?: string;
    defaultValue?: Value;
    formatter?: NumericInput.Formatter;
    max?: number;
    min?: number;
    maxResetStepValue?: number;
    minResetStepValue?: number;
    name?: string;
    readonly?: boolean;
    value: Value;
    onChange: (params: OnChangeCallbackParams) => void;
  }
}
