// @divi/types/src/module/element/decoration/heading-font/index.ts
import { type Components } from '../../../components';
import { type Font } from '../font';


export namespace HeadingFont {
  export interface Attributes {
    h1?: Font.Attributes;
    h2?: Font.Attributes;
    h3?: Font.Attributes;
    h4?: Font.Attributes;
    h5?: Font.Attributes;
    h6?: Font.Attributes;
  }

  export namespace Style {

    // TODO replace `FontHeaderStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes>{
      propertySelectors?: {
        h1?: Font.Style.Props['propertySelectors'];
        h2?: Font.Style.Props['propertySelectors'];
        h3?: Font.Style.Props['propertySelectors'];
        h4?: Font.Style.Props['propertySelectors'];
        h5?: Font.Style.Props['propertySelectors'];
        h6?: Font.Style.Props['propertySelectors'];
      };
      important?: boolean | {
        h1?: Font.Style.Props['important'];
        h2?: Font.Style.Props['important'];
        h3?: Font.Style.Props['important'];
        h4?: Font.Style.Props['important'];
        h5?: Font.Style.Props['important'];
        h6?: Font.Style.Props['important'];
      };
    }
  }
}
