// @divi/types/src/field-library/richtext/index.ts
import { type MouseEvent } from 'react';


export namespace RichText {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {MouseEvent} [event] Event.
   * @property {string} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: MouseEvent;
    inputValue?: string;
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} [className] Class name.
   * @property {string} value Value.
   * @property {string} [defaultValue] Default value.
   * @property {string} [toggleSlug] Toggle slug.
   * @property {string[]} [tinyMCEPlugins] TinyMCE plugins.
   * @property {boolean} isTinyMCEAllowed Is TinyMCE allowed.
   * @property {string} tinymceCSSFiles TinyMCE CSS files.
   * @property {string} tinymceSkinUrl TinyMCE skin URL.
   * @property {Function} onChange On change callback.
   * @property {Function} [onModeSwitch] On mode switch callback.
   */
  export interface Props {
    name: string;
    className?: string;
    value: string;
    defaultValue?: string;
    toggleSlug?: string;
    tinyMCEPlugins?: string[];
    isTinyMCEAllowed: boolean;
    tinymceCSSFiles:  string;
    tinymceSkinUrl:   string;
    onChange: (params: OnChangeCallbackParams) => void;
    onModeSwitch?: (mode: string) => void;
  }
}
