// @divi/types/src/fonts/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';

import { type Custom } from '../custom';
import { type Google } from '../google';
import { type Loaded } from '../loaded';


export namespace Store {
  export interface State {
    customFonts: Custom.Fonts;
    googleFonts: Google.Fonts;
    loadedFonts: Loaded.Fonts;
  };

  export type StateItem = string | number | Record<string, unknown>;

  export type ImmutableState = ImmutableObject<State>;

  export interface Instance {
    dispatch: () => void;
    getState: () => any;
  };

  export namespace Actions {
    export namespace LoadFont {
      export type ReturnValue = {
        type: 'LOAD_FONT';
        name: string;
      }
    }

    export type FunctionMap = {
      loadFont: (name: string) => LoadFont.ReturnValue;
    };

    export type ReturnValue = LoadFont.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getCustomFonts: () => Custom.Fonts;
      getGoogleFonts: () => Google.Fonts;
      getLoadedFonts: () => Loaded.Fonts;
      isLoaded: (name: string) => boolean;
    }
  }
}
