// @divi/types/src/field-library/checkboxes/index.ts
import {
  type ReactNode,
} from 'react';


export namespace Checkboxes {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {string[]} inputValue Input value.
   */
  export interface OnChangeCallbackParams {
    inputValue: string[];
  }

  /**
   * @typedef {object} Options
   * @property {ReactNode} label Label.
   * @property {string} value Value.
   * @property {string} [className] Class name.
   */
  export interface Options {
    label: ReactNode,
    value: string,
    className?: string,
  }

  /**
   * @typedef {object} Props
   * @property {string} [name] Name.
   * @property {string} [className] Class name.
   * @property {string[]} [defaultValue] Default value.
   * @property {Function} [onChange] On change callback.
   * @property {object[]} options Options.
   * @property {string[]} [value] Value.
   * @property {object} [valueMap] Value map.
   * @property {string} [valueMap.checked] Checked value.
   * @property {string} [valueMap.unchecked] Unchecked value.
   */
  export interface Props {
    name?: string;
    className?: string;
    defaultValue?: string[];
    onChange?: (params: OnChangeCallbackParams) => void;
    options: Options[];
    value?: string[];
    valueMap?: {
      checked: string,
      unchecked: string,
    };
  }
}
