// @divi/types/src/field-library/select/index.ts
import {
  type FunctionComponent,
  type ReactNode,
} from 'react';

import { type KeyboardShortcuts } from '../../keyboard-shortcuts';


export namespace Select {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {string} inputValue Input value.
   */
  export interface OnChangeCallbackParams {
    inputValue: string;
  }

  export type OptionLabel = string | ReactNode | FunctionComponent;

  /**
   * @typedef {object} Option
   * @property {object} label Label.
   * @property {string} [name] Name.
   * @property {object} [options] Options.
   * @property {boolean} [isStatic] Is static.
   * @property {string} [marker] Marker.
   * @property {string} [className] Class name.
   * @property {Function} [markerAction] Marker action.
   * @property {Function} [customContent] Custom content.
   */
  export interface Option {
    label: OptionLabel;
    name?: string;
    options?: Record<string, Option>;
    isStatic?: boolean;
    marker?: string;
    className?: string;
    markerAction?: () => void;
    customContent?: () => void;
  }

  export type Options = Record<string, Option>;

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {Function} onChange On change callback.
   * @property {object} options Options.
   * @property {string} [value] Value.
   * @property {boolean} [activeOnLoad] Active on load.
   * @property {object} [additionalOptions] Additional options.
   * @property {Function} [afterList] After list.
   * @property {Function} [beforeList] Before list.
   * @property {string} [cookiePath] Cookie path.
   * @property {string} [className] Class name.
   * @property {string} [defaultValue] Default value.
   * @property {boolean} [displayCheckMark] Display check mark.
   * @property {boolean} [displayRecent] Display recent.
   * @property {string} [emptyLabel] Empty label.
   * @property {boolean} [groupSeparator] Group separator.
   * @property {Function} [labelFilter] Label filter.
   * @property {Function} [onClose] On close callback.
   * @property {Function} [onOpen] On open callback.
   * @property {Function} [onHoverOption] On hover option callback.
   * @property {boolean} ["readonly"] Read only.
   * @property {string} [recentCategory] Recent category.
   * @property {boolean} [searchable] Searchable.
   * @property {string} [searchPlaceholder] Search placeholder.
   * @property {Function} [valuesToUpdateCallback] Values to update callback.
   * @property {Function} [registerShortcuts] Register shortcuts.
   * @property {Function} [unregisterShortcuts] Unregister shortcuts.
   * @property {boolean} [updateMenuPositionOnLoad] Update menu position on load.
   *
   */
  export interface Props {
    name: string;
    onChange: (params: OnChangeCallbackParams) => void;
    options: Options;

    value?: string;
    activeOnLoad?: boolean;
    additionalOptions?: Options;
    afterList?: () => ReactNode | void;
    beforeList?: () => ReactNode | void;
    cookiePath?: string;
    className?: string;
    defaultValue?: string;
    displayCheckMark?: boolean;
    displayRecent?: boolean;
    emptyLabel?: string;
    groupSeparator?: boolean;
    labelFilter?: (label: string | ReactNode, value: string) => string;
    onClose?: () => void;
    onOpen?: () => void;
    onHoverOption?: (value: string) => void;
    readonly?: boolean;
    recentCategory?: string;
    searchable?: boolean;
    searchPlaceholder?: string;
    valuesToUpdateCallback?: any;
    registerShortcuts?: (shortcuts: KeyboardShortcuts.Shortcut.Item[]) => void,
    unregisterShortcuts?: (names: string[]) => void;
    updateMenuPositionOnLoad?: boolean;
  }

  /**
   * @typedef {object} UpdateCallbackParams
   * @property {number} menuPosition Menu position.
   * @property {number} menuWidth Menu width.
   * @property {number} menuHeight Menu height.
   */
  export interface UpdateCallbackParams {
    menuPosition: number;
    menuWidth: number;
    menuHeight: number;
  }
}
