// @divi/types/src/module/element/decoration/transform/index.ts
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';


export namespace Transform {
  export interface ScaleValue {
    x?: string;
    y?: string;
    linked?: string;
  }

  export interface TranslateValue {
    x?: string;
    y?: string;
    linked?: string;
  }

  export interface RotateValue {
    x?: string;
    y?: string;
    z?: string;
  }

  export interface SkewValue {
    x?: string;
    y?: string;
    linked?: string;
  }

  export interface OriginValue {
    x: string;
    y: string;
  }

  export interface AttributeValue {
    scale?: ScaleValue;
    translate?: TranslateValue;
    rotate?: RotateValue;
    skew?: SkewValue;
    origin?: OriginValue;
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Style {

    // TODO replace `TransformStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.Transform>;
      important?: boolean | FormatStyleImportant<Css.Properties.Transform>;
    }
  }
}
