// @divi/types/src/field-library/sortable-list/index.ts
import { type FormEvent } from 'react';


export namespace SortableList {
  export interface ItemValue {
    value?: string;
    checked?: string;
    dragID?: string;
    link?: {
      url?: string;
      text?: string;
    };
  }

  export type Value = ItemValue[];

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {FormEvent} [event]
   * @property {Array} [inputValue]
   */
  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: Value;
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   */
  export interface Props {
    value?: Value;
    buttonLabel?: string;
    useAddNewButton?: boolean;
    allowEmpty?: boolean;
    isRadio?: boolean;
    isCheckbox?: boolean;
    addFirstOnMount?: boolean;
    firstValue?: string;
    actions?: {
      left?: Actions[];
      right?: Actions[];
    };
    onChange: (params: OnChangeCallbackParams) => void;
  }

  type Actions = 'move' | 'copy' | 'delete' | 'link';

  export interface ItemProps {
    index?: number;
    dragID?: string;
    className?: string;
    isRadio?: boolean;
    isCheckbox?: boolean;
    isChecked?: boolean;
    onChange?: (
      index: number,
      valueKey: string,
      inputValue: string | {
        url?: string,
        text?: string
      }
    ) => void;
    onAdd?: (index: number) => void;
    onCopy?: (index: number) => void;
    onDelete?: (index: number) => void;
    onMove?: (id: string, atIndex: number) => void;
    onFind?: (id: string) => number;
    onRadioChecked?: (index: number, inputValue: string) => void;
    value?: string;
    link?: {
      url?: string;
      text?: string;
    };
    actions?: {
      left?: Actions[];
      right?: Actions[];
    };
  }
}
