// @divi/types/src/module/components/style/index.ts
import {
  type ReactElement,
  type ReactNode,
} from 'react';

import {
  type AttrState,
  type FormatBreakpointStateAttr,
} from '../../../format-attr';


export namespace Style {

  // TODO Replace `StyleProps` with this.
  export type Mode = 'builder' | 'frontend';

  // TODO Replace `SelectorFunctionParams` with this.
  export interface SelectorFunctionParams<TAttr> {
    selector: string,
    breakpoint: string,
    state: string,
    attr: TAttr
  }

  // TODO Replace `StyleImportantProp` with this.
  export interface ImportantProp {
    [key: string]: boolean;
  }

  // TODO Replace `SelectorFunction` with this.
  export type SelectorFunction = <TParams extends SelectorFunctionParams<TParams['attr']>>(params: TParams) => string;

  // TODO Replace `StyleProps` with this.
  export interface Props<TAttr> {
    selector?: string;
    selectors?: FormatBreakpointStateAttr<string>;
    selectorFunction?: SelectorFunction;
    attr: TAttr;
    defaultPrintedStyleAttr?: TAttr;
    asStyle?: boolean;

    // @todo set as required once mode is passed everywhere.
    mode?: Mode;
    state?: AttrState;
  }

  // TODO Replace `StyleWrapperProps` with this.
  export interface WrapperProps<TAttr> {
    attr: TAttr;
    children: ReactNode;
    asStyle: boolean;
  }

  // TODO Replace `StyleContainerProps` with this.
  export interface ContainerProps {
    children?: ReactElement | ReactElement[];
    mode: Mode;
    state: AttrState;
    noStyleTag?: boolean;
    noEmptyStyle?: boolean;
  }
}
