// @divi/types/src/module/library/video/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';

// Video Attr.
export interface VideoAttr {
  src?: string;
  webm?: string;
  thumbnail?: {
    src?: string;
    overlayColor?: string;
  }
}

export interface VideoAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  // Module.
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'spacing' |
      'sizing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Video.
  video?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
      webm?: string;
    }>;
  };

  // Video Play Icon.
  playIcon?: {
    decoration?: Element.Decoration.PickedAttributes<'icon'>;
  };

  // Video Overlay.
  overlay?: {
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  };

  thumbnail?: {
    innerContent?: FormatBreakpointStateAttr<{
      src?: string;
    }>
  };
}
