// @divi/types/src/module/element/index.ts
import { Advanced as AdvancedNamespace } from './advanced';
import { Components as ComponentsNamespace } from './components';
import { Decoration as DecorationNamespace } from './decoration';
import { InnerContent as InnerContentNamespace } from './inner-content';
import { Meta as MetaNamespace } from './meta';
import { ScriptData as ScriptDataNamespace } from './script-data';
import { Style as StyleNamespace } from './style';
import {
  type Type as ElementType,
  Types as TypesNamespace,
} from './types';


export namespace Element {
  export import InnerContent = InnerContentNamespace;
  export import Decoration = DecorationNamespace;
  export import Meta  = MetaNamespace;
  export import Advanced = AdvancedNamespace;

  /* eslint-disable @typescript-eslint/no-unused-vars */
  // TODO figure out a way to declare nested namespace without disabling no-unused-vars rule
  // imported namespace as marked as unused by eslint using this config; it shouldn't happen in the first place.
  export import ScriptData = ScriptDataNamespace;
  export import Style = StyleNamespace;
  export import Components = ComponentsNamespace;

  /* eslint-enable @typescript-eslint/no-unused-vars */

  export interface Attributes {
    innerContent?: InnerContent.Attributes;
    decoration?: Decoration.Attributes;
    advanced?: Advanced.Attributes;
    meta?: Meta.Attributes;
  }

  export type Part = keyof Attributes;

  export type AttributesPart = Attributes[Part]

  export type Type = ElementType;

  /* eslint-disable @typescript-eslint/no-unused-vars */
  export import Types = TypesNamespace;
  /* eslint-enable @typescript-eslint/no-unused-vars */
}
