// @divi/types/src/module/store/sticky/index.ts
import { type Breakpoint } from '../../../format-attr';
import { type Decoration } from '../../element/decoration';


export namespace Sticky {

  // module/options/sticky/script-data
  export type FormatSetting<TValue> = Partial<Record<Breakpoint, TValue>> | TValue;

  export type Setting = {
    id: string;
    selector: string;

    // TODO (D5, Refactor) maybe undo the camelCase conversion to underscore separated format.
    // camelCase format needs to be re-written because D5 still uses D4's frontend script.
    position: FormatSetting<Omit<Decoration.Sticky.AttributeValue['position'], 'topBottom'> | 'top_bottom'>;
    topOffset: FormatSetting<Decoration.Sticky.AttributeValue['offset']['top']>;
    bottomOffset: FormatSetting<Decoration.Sticky.AttributeValue['offset']['bottom']>;
    topLimit: FormatSetting<Decoration.Sticky.AttributeValue['limit']['top']>;
    bottomLimit: FormatSetting<Decoration.Sticky.AttributeValue['limit']['bottom']>;
    offsetSurrounding: FormatSetting<Decoration.Sticky.AttributeValue['offset']['surrounding']>;
    transition: Decoration.Sticky.AttributeValue['transition'] | FormatSetting<Decoration.Sticky.AttributeValue['transition']>;
    moduleAlignment?: Decoration.Sizing.AttributeValue['alignment'];
    styles?: {
      width?: Decoration.Sizing.AttributeValue['width'];
      'max-width'?: Decoration.Sizing.AttributeValue['maxWidth'];
      /* eslint-disable-next-line @typescript-eslint/naming-convention */
      module_alignment?: {
        desktop?: Decoration.Sizing.AttributeValue['alignment'] | '';
        tablet?: Decoration.Sizing.AttributeValue['alignment'] | '';
        phone?: Decoration.Sizing.AttributeValue['alignment'] | '';
      };
      positioning?: Decoration.Position.AttributeValue['mode'];
      /* eslint-disable-next-line @typescript-eslint/naming-convention */
      vertical_offset?: Decoration.Position.AttributeValue['offset']['vertical'];
      /* eslint-disable-next-line @typescript-eslint/naming-convention */
      horizontal_offset?: Decoration.Position.AttributeValue['offset']['horizontal'];
    }
    stickyStyles?: {
      width?: Decoration.Sizing.AttributeValue['width'];
      'max-width'?: Decoration.Sizing.AttributeValue['maxWidth'];
      /* eslint-disable-next-line @typescript-eslint/naming-convention */
      position_origin_r?: Decoration.Position.AttributeValue['origin']['relative'];
      /* eslint-disable-next-line @typescript-eslint/naming-convention */
      horizontal_offset?: Decoration.Position.AttributeValue['offset']['horizontal'];
      /* eslint-disable-next-line @typescript-eslint/naming-convention */
      vertical_offset?: Decoration.Position.AttributeValue['offset']['vertical'];
    }
  }
}
