// @divi/types/src/shortcode-module/advanced-field/box-shadow/index.ts
/* eslint-disable @typescript-eslint/naming-convention */
export interface ShortcodeModuleAdvancedFieldDefinitionBoxShadow {
  css?: {
    custom_style?: boolean;
    important?: boolean;
    hover?: string;
    main?: string;
    overlay?: string | boolean;
    show_if?: Record<string, string>;
    show_if_not?: Record<string, string>;
  }
  default?: ShortcodeModuleBoxShadowFieldDefault;
  default_on_fronts?: {
    color?: string;
    position?: string;
  }
  label?: string;
  name?: string;
  option_category?: string;
  show_if?: Record<string, string>;
  tab_slug?: string;
  toggle_slug?: string;
}

export interface ShortcodeModuleBoxShadowFieldPreset {
  value: string;
  icon?: string;
  content?: string;
  fields: Record<string, string>;
}

export type ShortcodeModuleBoxShadowFieldPresets = ShortcodeModuleBoxShadowFieldPreset[];

export interface ShortcodeModuleBoxShadowPreset {
  horizontal: string;
  vertical: string;
  blur: string;
  spread: string;
  position: string;
}

export type ShortcodeModuleBoxShadowPresetName = keyof ShortcodeModuleBoxShadowPresetOptions;

export type ShortcodeModuleBoxShadowPresetOptions = {
  none?: string;
  preset1?: string;
  preset2?: string;
  preset3?: string;
  preset4?: string;
  preset5?: string;
  preset6?: string;
  preset7?: string;
}

export type ShortcodeModuleBoxShadowFieldDefault = [
  string,
  ShortcodeModuleBoxShadowPresetOptions
];
