// @divi/types/src/conversion.ts
// @divi/module and @divi/module-library
export type ModuleAttributeConversionMap = Record<string, string>;

export type ValueExpansionFunction = (value: string) => any;

export type ValueExpansionFunctionMap = Record<string, ValueExpansionFunction>;

export interface ModuleConversionOutline {
  advanced?: {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    admin_label?: string;
    animation?: string;
    background?: string;
    borders?: Record<string, string>;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    disabled_on?: string;
    dividers?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    form_field?: Record<string, string>;
    height?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    image_icon?: Record<string, string>;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    link_options?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    margin_padding?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    max_width?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    module?: string;
    overflow?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    position_fields?: string;
    text?: string;
    transform?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    z_index?: string;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    box_shadow?: Record<string, string>;
    button?: Record<string, string>;
    fonts?: Record<string, string>;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    text_shadow?: Record<string, string>;
    filters?: Record<string, string>;
    sizing?: string;
    scroll?: Record<string, string> | string;
    sticky?: Record<string, string> | string;
    transition?: string;
  };
  css?: Record<string, string>;
  module?: Record<string, string>;
  valueExpansionFunctionMap?: ValueExpansionFunctionMap;
}

export type AdvancedOptionFunctionsMap = Record<string, (...args: any[]) => ModuleConversionMap>;

export interface ModuleConversionMap {
  attributeMap: ModuleAttributeConversionMap;
  optionEnableMap: Record<string, string>;
  valueExpansionFunctionMap: ValueExpansionFunctionMap;
}

// @divi/conversion
export type ModuleName = string;

export type ModuleLibraryConversionMap = Record<ModuleName, ModuleConversionMap>;

export type ConversionModuleLibraryAttributeMap = Record<ModuleName, ModuleConversionMap>;

export type ConversionModuleLibraryValueExpansionFunctionMap = Record<ModuleName, ValueExpansionFunctionMap>;
