// @divi/types/src/utils.ts
import { type PostStatus } from '@divi/types-external';

/**
 * This file contains generics that serves as utility type.
 *
 * IMPORTANT: DO NOT IMPORT ANY VARIABLE HERE. DOING SO MIGHT CAUSE DEPENDENCY CYCLE.
 */

/**
 * Get property type of given object type.
 */
export type PropertyTypeOf<T> = T[keyof T];

/**
 * Get keys of union types.
 */
export type KeysOfUnion<T> = T extends T ? keyof T: never;

/**
 * Transform certain properties as required.
 *
 * @since ??
 */
export type TransformAsRequired<T, TK extends keyof T> = Omit<T, TK> & {
    [P in TK]: T[P];
}

/**
 * Transform certain properties as optional.
 *
 * @since ??
 */
export type TransformAsOptional<T, TK extends keyof T> = Omit<T, TK> & {
    [P in TK]?: T[P];
}

export interface PostObject {
  ID:                    number;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_author:           string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_date:             Date;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_date_gmt:         Date;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_content:          string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_title:            string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_excerpt:          string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_status:           PostStatus;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  comment_status:        string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  ping_status:           string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_password:         string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_name:             string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  to_ping:               string;
  pinged:                string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_modified:         Date;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_modified_gmt:     Date;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_content_filtered: string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_parent:           number;
  guid:                  string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  menu_order:            number;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_type:             string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  post_mime_type:        string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  comment_count:         string;
  filter:                string;
}

/* eslint-disable @typescript-eslint/naming-convention */
export interface Taxonomy {
  term_id: number
  name: string
  slug: string
  term_group: number
  term_taxonomy_id: number
  taxonomy: string
  description: string
  parent: number
  filter: string
}
