// @divi/types/src/module/element/decoration/index.ts
import { Animation as AnimationNamespace } from './animation';
import { Background as BackgroundNamespace } from './background';
import { BodyFont as BodyFontNamespace } from './body-font';
import { Border as BorderNamespace } from './border';
import { BoxShadow as BoxShadowNamespace } from './box-shadow';
import { Button as ButtonNamespace } from './button';
import { Conditions as ConditionsNamespace } from './conditions';
import { DisabledOn as DisabledOnNamespace } from './disabled-on';
import { Filters as FiltersNamespace } from './filters';
import { Font as FontNamespace } from './font';
import { HeadingFont as HeadingFontNamespace } from './heading-font';
import { Icon as IconNamespace } from './icon';
import { InlineFont as InlineFontNamespace } from './inline-font';
import { Overflow as OverflowNamespace } from './overflow';
import { Position as PositionNamespace } from './position';
import { Scroll as ScrollNamespace } from './scroll';
import { Sizing as SizingNamespace } from './sizing';
import { Spacing as SpacingNamespace } from './spacing';
import { Sticky as StickyNamespace } from './sticky';
import { Transform as TransformNamespace } from './transform';
import { Transition as TransitionNamespace } from './transition';
import { ZIndex as ZIndexNamespace } from './z-index';


export namespace Decoration {
  export import Animation = AnimationNamespace;
  export import Background = BackgroundNamespace;
  export import BodyFont = BodyFontNamespace;
  export import Border = BorderNamespace;
  export import BoxShadow = BoxShadowNamespace;
  export import Button = ButtonNamespace;
  export import Conditions = ConditionsNamespace;
  export import DisabledOn = DisabledOnNamespace;
  export import Filters = FiltersNamespace;
  export import Font = FontNamespace;
  export import HeadingFont = HeadingFontNamespace;
  export import Icon = IconNamespace;
  export import InlineFont = InlineFontNamespace;
  export import Overflow = OverflowNamespace;
  export import Position = PositionNamespace;
  export import Scroll = ScrollNamespace;
  export import Sizing = SizingNamespace;
  export import Spacing = SpacingNamespace;
  export import Sticky = StickyNamespace;
  export import Transform = TransformNamespace;
  export import Transition = TransitionNamespace;
  export import ZIndex = ZIndexNamespace;

  export type Attributes = {
    animation?: Animation.Attributes;
    background?: Background.Attributes;
    bodyFont?: BodyFont.Attributes;
    border?: Border.Attributes;
    boxShadow?: BoxShadow.Attributes;
    button?: Button.Attributes;
    conditions?: Conditions.Attributes;
    disabledOn?: DisabledOn.Attributes;
    filters?: Filters.Attributes;
    font?: Font.Attributes;
    icon?: Icon.Attributes;
    inlineFont?: InlineFont.Attributes;
    headingFont?: HeadingFont.Attributes;
    overflow?: Overflow.Attributes;
    position?: Position.Attributes;
    scroll?: Scroll.Attributes;
    sizing?: Sizing.Attributes;
    spacing?: Spacing.Attributes;
    sticky?: Sticky.Attributes;
    transform?: Transform.Attributes;
    transition?: Transition.Attributes;
    zIndex?: ZIndex.Attributes;
  }

  export type PickedAttributes<TAttrNames extends keyof Attributes> = Pick<Attributes, TAttrNames>
}
