// @divi/types/src/module/library/fullwidth-header/index.ts
// Internal dependencies.
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type Background } from '../../element/decoration/background';
import { type InternalAttrs } from '../internal';


export interface FullwidthHeaderCssAttr extends Css.AttributeValue {
  headerContainer?: string;
  headerImage?: string;
  logo?: string;
  title?: string;
  content?: string;
  subtitle?: string;
  button1?: string;
  button2?: string;
  scrollButton?: string;
}

export type FullwidthHeaderCssGroupAttr = FormatBreakpointStateAttr<FullwidthHeaderCssAttr>;

export interface FullwidthHeaderAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: FullwidthHeaderCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
      headerFullscreen?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Logo
  logo?: {
    innerContent?: Element.Types.Image.InnerContent.Attributes;
  }

  // Title
  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Subhead
  subhead?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Content
  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    advanced?: {
      maxWidth?: FormatBreakpointStateAttr<string>;
      orientation?: FormatBreakpointStateAttr<'center' | 'bottom'>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'bodyFont' |
      'inlineFont'
    >;
  };

  // Image
  image?: {
    innerContent?: Element.Types.Image.InnerContent.Attributes;
    advanced?: {
      orientation?: FormatBreakpointStateAttr<'center' | 'bottom'>;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'border' |
      'boxShadow' |
      'filters'
    >;
  }

  // Overlay.
  overlay?: {
    decoration?: {
      background?: Background.Attributes;
    };
  }

  // Button.
  buttonOne?: Element.Types.Button.Attributes;
  buttonTwo?: Element.Types.Button.Attributes;

  scrollDown?: {
    decoration?: Element.Decoration.PickedAttributes<'icon'>;
  };
}
