// @divi/types/src/field-library/map/index.ts
import { type FormEvent } from 'react';

import { type Map as InnerContentMap } from '../../module/element/inner-content/map';


export namespace Map {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {object} [event] Event.
   * @property {object} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: InnerContentMap.AttributeValue;
  }

  /**
   * @typedef {object} Props
   * @property {string} name Name.
   * @property {string} [className] Class name.
   * @property {Function} onChange On change callback.
   * @property {object} [defaultValue] Default value.
   * @property {object} value Value.
   * @property {boolean} ["readonly"] Read only.
   * @property {boolean} [hideUpdateButton] Hide update button.
   * @property {boolean} [hasMarker] Has marker.
   * @property {object[]} [extraMarkers] Extra markers.
   * @property {string} [imagesUri] Images URI.
   * @property {string} [googleMapAPIKey] Google map API key.
   * @property {boolean} [hasAPIKey] Has API key.
   */
  export interface Props {
    name: string;
    className?: string;
    onChange: (params: OnChangeCallbackParams) => void;
    defaultValue?: InnerContentMap.AttributeValue;
    value: InnerContentMap.AttributeValue;
    readonly?: boolean;
    hideUpdateButton?: boolean;
    hasMarker?: boolean;
    extraMarkers?: InnerContentMap.AttributeValue[];
    imagesUri?: string;
    googleMapAPIKey?: string;
    hasAPIKey?: boolean;
  }
}
