// @divi/types/src/module/library/accordion-item/index.ts
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface AccordionItemCssAttr extends Css.AttributeValue {
  openToggle?: string;
  toggleTitle?: string;
  toggleIcon?: string;
  toggleContent?: string;
}

export type AccordionItemCssGroupAttr = FormatBreakpointStateAttr<AccordionItemCssAttr>;

export interface AccordionItemAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: AccordionItemCssGroupAttr;

  // Module.
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      open?: FormatBreakpointStateAttr<string>;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
    'background' |
    'border' |
    'boxShadow' |
    'filters' |
    'overflow' |
    'position' |
    'scroll' |
    'sizing' |
    'spacing' |
    'transform' |
    'transition' |
    'zIndex'
    >;
  };

  // Title.
  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Content.
  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: Element.Decoration.PickedAttributes<
      'bodyFont' |
      'inlineFont'
    >;
  };

  // Open Toggle.
  openToggle?: {
    decoration?: Element.Decoration.PickedAttributes<
    'background' |
    'font'
    >;
  };

  // Closed Toggle.
  closedToggle?: {
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'font'
    >;
  };

  // Closed Toggle Icon.
  closedToggleIcon?: {
    decoration?: Element.Decoration.PickedAttributes<'icon' | 'font'>;
  };
}
