// @divi/types/src/field-library/upload/index.ts
import {
  type MouseEvent,
  type ReactNode,
} from 'react';

import {
  type AttrState,
  type Breakpoint,
} from '../../format-attr';
import { type Background } from '../../module/element/decoration/background';


export namespace Upload {

  // Original `onChangeCallbackParams` is located at
  // '/src/components/modals/components/module/containers/types'.  component
  // need to handle more returned value.

  /**
   * @typedef {object} OnChangeCallbackParams  component onChange callback params.
   * @property {MouseEvent} [event] Event object.
   * @property {string|number|Record<string, string|number>} [inputValue] Input value.
   */
  export interface OnChangeCallbackParams {
    event?: MouseEvent;
    inputValue?: string | number | Record<string, string | number>;
  }

  /**
   * @typedef {Function} OnChangeCallback  component onChange callback.
   * @param {OnChangeCallbackParams} params  Component onChange callback params.
   */
  export type OnChangeCallback = (params: OnChangeCallbackParams) => void;

  /**
   * @typedef {object} Props  component props.
   * @property {string} name Field name.
   * @property {string} [className] Field class name.
   * @property {Function} onChange Field onChange callback.
   * @property {Function} onReset Field onReset callback.
   * @property {string|number} value Field value.
   * @property {string|number} [defaultValue] Field default value.
   * @property {string} [dataType] Field data type.
   * @property {boolean} [isBackgroundType] Whether field is background type.
   * @property {object} [backgroundAttrs] Background attributes.
   * @property {boolean} ["readonly"] Whether field is readonly.
   * @property {boolean} [hideMetadata] Whether field is hide metadata.
   * @property {boolean} [embed] Whether field is embed.
   * @property {boolean} [showResetButton] Whether field is show reset button.
   * @property {string} [chooseText] Field choose text.
   * @property {string} [uploadButtonText] Field upload button text.
   * @property {boolean} [attachmentId] Whether field is attachment id.
   * @property {boolean} [renderPreview] Whether field is render preview.
   * @property {Function} [renderBefore] Field render before.
   * @property {Function} [renderAfter] Field render after.
   */
  export interface Props {
    name: string;
    className?: string;
    onChange: OnChangeCallback;
    onReset: (event: MouseEvent) => void;
    value: string|number;
    defaultValue?: string|number;
    dataType?: string;
    isBackgroundType?: boolean,
    backgroundAttrs?: Background.AttributeValue,
    readonly?: boolean,
    hideMetadata?: boolean,
    embed?: boolean,
    showResetButton?: boolean,
    chooseText?: string,
    uploadButtonText?: string,
    attachmentId?: boolean,
    renderPreview?: boolean;
    renderBefore?: (state: string) => ReactNode;
    renderAfter?: (state: string) => ReactNode;
  }

  /**
   * @typedef {object} AttributeSize  component attribute size.
   * @property {number} height Attribute size height.
   * @property {string} orientation Attribute size orientation.
   * @property {string} url Attribute size url.
   * @property {string} width Attribute size width.
   */
  export interface AttributeSize {
    height: number;
    orientation: string;
    url: string;
    width: string;
  }

  /**
   * @typedef {object} AttributeSizes  component attribute sizes.
   */
  export type AttributeSizes = Record<string, AttributeSize>;

  /**
   * @typedef {object} ContainerProps  component container props.
   * @property {object} [attr] Background group attribute.
   * @property {'desktop' | 'tablet' | 'phone'} [appBreakpoint] App breakpoint.
   * @property {'value' | 'hover' | 'sticky'} [appState] App state.
   * @property {boolean} [isBackgroundType] Whether field is background type.
   */
  export type ContainerProps = Record<string, any> & {
    attr?: Background.Attributes;
    appBreakpoint?: Breakpoint,
    appState?: AttrState,
    isBackgroundType?: boolean,
  }
}
