// @divi/types/src/module/settings/group/library/index.ts
import { AdminLabel as AdminLabelNamespace } from './admin-label';
import { Animation as AnimationNamespace } from './animation';
import { AttributesRel as AttributesRelNamespace } from './attributes-rel';
import { Background as BackgroundNamespace } from './background';
import { Border as BorderNamespace } from './border';
import { BoxShadow as BoxShadowNamespace } from './box-shadow';
import { Button as ButtonNamespace } from './button';
import { ButtonIcon as ButtonIconNamespace } from './button-icon';
import { ButtonLink as ButtonLinkNamespace } from './button-link';
import { Composite as CompositeNamespace } from './composite';
import { Conditions as ConditionsNamespace } from './conditions';
import { Css as CssNamespace } from './css';
import { DisabledOn as DisabledOnNamespace } from './disabled-on';
import { Dividers as DividersNamespace } from './dividers';
import { EmailService as EmailServiceNamespace } from './email-service';
import { Filters as FiltersNamespace } from './filters';
import { Font as FontNamespace } from './font';
import { FontBody as FontBodyNamespace } from './font-body';
import { FontHeader as FontHeaderNamespace } from './font-header';
import { FormField as FormFieldNamespace } from './form-field';
import { Gutter as GutterNamespace } from './gutter';
import { Icon as IconNamespace } from './icon';
import { IdClasses as IdClassesNamespace } from './id-classes';
import { Link as LinkNamespace } from './link';
import { type GroupLibraryMapWithoutComposite } from './map';
import { Overflow as OverflowNamespace } from './overflow';
import { Position as PositionNamespace } from './position';
import { PositionSettings as PositionSettingsNamespace } from './position-settings';
import { Scroll as ScrollNamespace } from './scroll';
import { ScrollSettings as ScrollSettingsNamespace } from './scroll-settings';
import { Sizing as SizingNamespace } from './sizing';
import { Spacing as SpacingNamespace } from './spacing';
import { SpamProtection as SpamProtectionNamespace } from './spam-protection';
import { Sticky as StickyNamespace } from './sticky';
import { Text as TextNamespace } from './text';
import { TextShadow as TextShadowNamespace } from './text-shadow';
import { Transform as TransformNamespace } from './transform';
import { Transition as TransitionNamespace } from './transition';
import { VisibilitySettings as VisibilitySettingsNamespace } from './visibility-settings';
import { ZIndex as ZIndexNamespace } from './z-index';


export namespace Library {
  /* eslint-disable @typescript-eslint/no-unused-vars */
  export import AdminLabel = AdminLabelNamespace;
  export import Animation = AnimationNamespace;
  export import AttributesRel = AttributesRelNamespace;
  export import Background = BackgroundNamespace;
  export import Border = BorderNamespace;
  export import BoxShadow = BoxShadowNamespace;
  export import Button = ButtonNamespace;
  export import ButtonIcon = ButtonIconNamespace;
  export import ButtonLink = ButtonLinkNamespace;
  export import Composite = CompositeNamespace;
  export import Conditions = ConditionsNamespace;
  export import Css = CssNamespace;
  export import DisabledOn = DisabledOnNamespace;
  export import Dividers = DividersNamespace;
  export import EmailService = EmailServiceNamespace;
  export import Filters = FiltersNamespace;
  export import Font = FontNamespace;
  export import FontBody = FontBodyNamespace;
  export import FontHeader = FontHeaderNamespace;
  export import FormField = FormFieldNamespace;
  export import Gutter = GutterNamespace;
  export import Icon = IconNamespace;
  export import IdClasses = IdClassesNamespace;
  export import Link = LinkNamespace;
  export import Overflow = OverflowNamespace;
  export import Position = PositionNamespace;
  export import PositionSettings = PositionSettingsNamespace;
  export import Scroll = ScrollNamespace;
  export import ScrollSettings = ScrollSettingsNamespace;
  export import Sizing = SizingNamespace;
  export import Spacing = SpacingNamespace;
  export import SpamProtection = SpamProtectionNamespace;
  export import Sticky = StickyNamespace;
  export import Text = TextNamespace;
  export import TextShadow = TextShadowNamespace;
  export import Transform = TransformNamespace;
  export import Transition = TransitionNamespace;
  export import VisibilitySettings = VisibilitySettingsNamespace;
  export import ZIndex = ZIndexNamespace;
  /* eslint-enable @typescript-eslint/no-unused-vars */

  export interface Map extends GroupLibraryMapWithoutComposite{
    'divi/composite': {
      props?: Composite.GroupProps;
    };
  }

  export type ComponentMap = {
    [TName in keyof Map]?: {
      type?: 'group';
      name?: TName;
      props?: Partial<Map[TName]['props']>;
    };
  }

  export type ComponentNames = keyof ComponentMap;

  export type Component = ComponentMap[ComponentNames];

  export type ComponentMapItem<TName extends ComponentNames = ComponentNames> = {
    name?: TName;
    type?: ComponentMap[TName]['type'];
    props?: ComponentMap[TName]['props'];
  }
}
