// @divi/types-external/src/wp/media.ts
// Properties listed here are outside our control, hence not conforming to our naming conventions.
/* eslint-disable @typescript-eslint/naming-convention */

export interface WpMediaFrame {
  open(): void;
  detach(): void;
  on(event: string, cb: (params: unknown) => void): void;
  state(): WpMediaFrameState;
}

export interface WpMediaFrameState {
  props: WpAttachment;
  get(prop: string): WpAttachments;
  display(attachment: WpAttachment): WpAttachment
  attributes?:Record<string, unknown>;
}

export interface WpMediaFrameUpdateSelection {
  filters: Record<string, unknown>;
  gallery: Record<string, unknown>;
  length: number;
  models: Record<string, unknown>;
  multiple: boolean;
  observers: Record<number, unknown>;
  props: Record<string, unknown>;
}

export interface MediaFrame {
  new(args: Record<string, unknown>): WpMediaFrame;
}

export type WpAttachments = WpAttachment[] & {
  first(): WpAttachment,
  add(wpAttachment: WpAttachment): WpAttachment,
};

export interface WpAttachment {
  id: number;
  caption: string;
  description: string;
  type: string;
  get(prop: string): string|number|Record<string, string|number>;
  toJSON?: () => WpAttachment;
  attributes?: Record<string, string>;
}

export interface WpMediaParams {
  frame?: string;
  state?: string;
  states?: unknown;
  metadata?: Record<string, any>;
  title?: string;
  editing?: boolean;
  multiple?: string | boolean;
  selection?: string|string[];
  library?: { type: string };
}

export interface WpMedia {
  (prop: WpMediaParams): WpMediaFrame,
  frames?: Record<string, WpMediaFrame>,
  view?: {
    MediaFrame: {
      ETSelect: MediaFrame,
    },
    settings?: {
      attachmentCounts?: {
        audio: number,
        video: number,
      },
      captions?: boolean,
      contentWidth?: number,
      defaultProps?: {
        link?: string,
        align?: string,
        size?: string,
      },
      embedExts?: string[],
      embedMimes?: Record<string, string>,
      mediaTrash?: number,
      mimeTypes?: Record<string, string>,
      months?: Record<number, { year: string, month: string, text: string }>,
      nonce?: Record<string, string>,
      oEmbedProxyUrl?: string,
      post?: Record<string, any>,
      tabUrl?: string,
      tabs: any[],
    }
  },
  string?: {
    props: (props?: unknown, attachment?: unknown) => unknown,
    link: (props?: unknown, attachment?: unknown) => unknown,
    audio: (props?: unknown, attachment?: unknown) => unknown,
    video: (props?: unknown, attachment?: unknown) => unknown,
    _audioVideo: (type?: unknown, props?: unknown, attachment?: unknown) => unknown,
    image: (props?: unknown, attachment?: unknown) => unknown,
  },
  controller?: {
    Library: {
      extend?: (props?: unknown) => {
        new(arg?: unknown): unknown,
      },
      prototype?: {
        defaults: unknown,
      },
    },
  },
  gallery?: {
    shortcode?: (selection: unknown) => ({
      attrs?: {
        named?: unknown,
      },
    }),
  },
  playlist?: {
    shortcode?: (selection: unknown) => ({
      attrs?: {
        named?: unknown,
      },
    }),
  },
  editor?: {
    insert?: (mediaString: string) => any,
  }
  attachment?(id: unknown): WpAttachment;
}
