// @divi/types/src/field-library/color-picker/index.ts
import { type ReactNode } from 'react';
import { type ImmutableObject } from 'seamless-immutable';

import { type AppUi } from '../../app-ui';
import { type Clipboard } from '../../clipboard';
import { type RightClickOptions } from '../../right-click-options';


export namespace ColorPicker {

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {string} inputValue Input value.
   */
  export interface OnChangeCallbackParams {
    inputValue: string;
  }

  /**
   * @typedef {object} Props
   * @property {string} addTitle Add title.
   * @property {string} cookiePath Cookie path.
   * @property {string[]} colorPalette Color palette.
   * @property {string[]} [globalColorPalette] Global color palette.
   * @property {string} defaultValue Default value.
   * @property {string} imagesUri Images URI.
   * @property {string} name Name.
   * @property {Function} onChange On change callback.
   * @property {Function} onReset On reset callback.
   * @property {Function} updateDefaultColors Update default colors.
   * @property {string} value Value.
   * @property {boolean} [animate] Animate.
   * @property {string} [className] Class name.
   * @property {boolean} [hasPreview] Has preview.
   * @property {boolean} [hasColorManager] Has color manager.
   * @property {boolean} [isAlpha] Is alpha.
   * @property {boolean} ["readonly"] Read only.
   * @property {boolean} [showAddButton] Show add button.
   * @property {boolean} [showCustomInput] Show custom input.
   * @property {boolean} [showEditButton] Show edit button.
   * @property {boolean} [showRemoveButton] Show remove button.
   * @property {boolean} [showResetButton] Show reset button.
   * @property {boolean} [showPaletteOnPickerActive] Show palette on picker active.
   * @property {boolean} [showPickerPalettes] Show picker palettes.
   * @property {boolean} [pickerOnStart] Picker on start.
   * @property {Function} [onClose] On close color picker.
   * @property {Function} [onAddGlobalColor] Callback function for on add global color.
   * @property {Function} [onEditGlobalColor] Callback function for on edit global color.
   * @property {Function} [onChangeGlobalColorValue] Called when the picker value changes while editing a global color.
   * @property {Function} [onDeleteGlobalColor] Callback function for on delete global color.
   * @property {Function} [updateDefaultColors] Callback function for update default colors.
   *
   */
  export interface Props {
    addTitle: string;
    attrName: string;
    cookiePath: string;
    colorPalette: string[];
    defaultValue: string;
    globalColorPalette?: string[];
    id: string;
    label: string;
    name: string;
    onChange: (params: OnChangeCallbackParams) => void;
    onReset: (event: React.MouseEvent) => void;
    onAddGlobalColor: (color: string) => void;
    onChangeGlobalColorValue: (colorId: string, color: string) => void;
    onDeleteGlobalColor: (colorId: string) => void;
    onEditGlobalColor: (colorId: string, color: string) => void;
    setRightClickOptions: (params: {
      options: RightClickOptions.Option.Items;
      cursorPosition: RightClickOptions.Cursor.Position;
      optionType: 'colorPicker';
      owner: RightClickOptions.Owner.Item;
      windowLocation: RightClickOptions.Window.Location;
    }) => void;
    toggleModuleInteraction: (status: AppUi.Store.Actions.ToggleModuleInteraction.Params['status']) => void;
    updateDefaultColors: (id: number | string, color: string) => void;
    value: string;

    animate?: boolean;
    children?: ReactNode;
    className?: string;
    groupName?: string;
    hasPreview?: boolean;
    hasColorManager?: boolean;
    isAlpha?: boolean;
    lastClipboardItem?: ImmutableObject<Clipboard.Item.ModuleFieldAttribute>;
    readonly?: boolean;
    showAddButton?: boolean;
    showCustomInput?: boolean;
    showEditButton?: boolean;
    showRemoveButton?: boolean;
    showResetButton?: boolean;
    showPaletteOnPickerActive?: boolean;
    showPickerPalettes?: boolean;
    subName?: string;
    pickerOnStart?: boolean; // Used in gradient preview.
    onCloseColorPicker?: () => void;
    getGlobalColorValue: (id: string) => string;
    isActiveGlobalColor: (id: string) => boolean;
  }

  export type ContainerProps = Partial<Omit<Props,
    'colorPalette' |
    'globalColorPalette' |
    'imagesUri' |
    'cookiePath' |
    'updateDefaultColors' |
    'onAddGlobalColor' |
    'onChangeGlobalColorValue' |
    'onEditGlobalColor' |
    'onDeleteGlobalColor' |
    'setRightClickOptions'
  >>;
}
