// @divi/types/src/module/element/decoration/transition/index.ts
import { type Css } from '../../../../css';
import { type FormatPropertySelectors } from '../../../../format-attr';
import { type Components } from '../../../components';
import { type AdvancedStyle } from '../../advanced-styles';
import { type Animation } from '../animation';
import { type Background } from '../background';
import { type BodyFont } from '../body-font';
import { type Border } from '../border';
import { type BoxShadow } from '../box-shadow';
import { type DisabledOn } from '../disabled-on';
import { type Filters } from '../filters';
import { type Font } from '../font';
import { type HeadingFont } from '../heading-font';
import { type Overflow } from '../overflow';
import { type Position } from '../position';
import { type Scroll } from '../scroll';
import { type Sizing } from '../sizing';
import { type Spacing } from '../spacing';
import { type Sticky } from '../sticky';
import { type Transform } from '../transform';
import { type ZIndex } from '../z-index';


export namespace Transition {
  export interface AttributeValue {
    duration?: string;
    delay?: string;
    speedCurve?: string;
    properties?: string[];
  }

  export interface Attributes {
    desktop?: {
      value?: AttributeValue;
    }
  }

  export type ElementAttributes = {
    animation?: Animation.Attributes;
    background?: Background.Attributes;
    bodyFont?: BodyFont.Attributes;
    border?: Border.Attributes;
    boxShadow?: BoxShadow.Attributes;
    transform?: Transform.Attributes;
    disabledOn?: DisabledOn.Attributes;
    filter?: Filters.Attributes;
    font?: Font.Attributes;
    headingFont?: HeadingFont.Attributes;
    overflow?: Overflow.Attributes;
    position?: Position.Attributes;
    scroll?: Scroll.Attributes;
    sizing?: Sizing.Attributes;
    spacing?: Spacing.Attributes;
    sticky?: Sticky.Attributes;
    zIndex?: ZIndex.Attributes;
  };

  export namespace Style {

    // TODO replace `TransitionStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.Transition>;
      attrs?: ElementAttributes;
      important?: boolean;
      advancedStyles?: AdvancedStyle.Item[];
    }
  }


}
