// @divi/types/src/module/element/decoration/animation/index.ts
import { type Css } from '../../../../css';
import {
  type FormatBreakpointStateAttr,
  type FormatPropertySelectors,
} from '../../../../format-attr';
import { type FormatStyleImportant } from '../../../../style';
import { type Components } from '../../../components';


export namespace Animation {
  export interface AttributeValue {
    style?: 'none' | 'fade' | 'slide' | 'bounce' | 'zoom' | 'flip' | 'fold' | 'roll';
    direction?: 'center' | 'left' | 'right' | 'bottom' | 'top';
    duration?: string;
    delay?: string;
    intensity?: {
      slide?: string;
      zoom?: string;
      flip?: string;
      fold?: string;
      roll?: string;
    };
    startingOpacity?: string;
    speedCurve?: 'ease-in-out' | 'ease' | 'ease-in' | 'ease-out' | 'linear';
    repeat?: 'once' | 'loop';
  }

  export type Attributes = FormatBreakpointStateAttr<AttributeValue>;

  export namespace Declaration {

    // TODO Replace `AnimationDeclarationParams` with this.
    export interface Params {
      style: AttributeValue['style'];
      attrValue: AttributeValue;
    }
  }

  export namespace Style {

    // TODO Replace `AnimationStyleProps` with this.
    export interface Props extends Components.Style.Props<Attributes> {
      propertySelectors?: FormatPropertySelectors<Css.Properties.Animation>;
      important?: boolean | FormatStyleImportant<Css.Properties.Animation>;
    }
  }
}
