// @divi/types/src/module/element/decoration/conditions/index.ts
import { type Author } from './condition-author';
import { type LoggedInStatus } from './condition-logged-in-status';
import { type UserRole } from './condition-user-role';


export namespace Conditions {
  type ValueOf<TObjectType, TValueType extends keyof TObjectType = keyof TObjectType> = TObjectType[TValueType];

  export type AllSettings = {
    loggedInStatus: LoggedInStatus.Settings;
    author: Author.Settings;
    userRole: UserRole.Settings;
  }
  export type Names = keyof AllSettings;

  export type Settings = ValueOf<AllSettings>;

  export type Item = {
    id: string;
    conditionName: Names;
    conditionSettings: Settings;
    operator: 'OR' | 'AND';
  }

  // TODO maybe remove ConditionsStore and the rest of the types since these are exact copy of it.
  // See: packages/module/src/options/conditions/group/conditions-fields/types.ts
  export type AttributeValue = Item[];

  // TODO consider using FormatBreakpointStateAttr<> to add responsive and state support.
  export type Attributes = AttributeValue;
}
