// @divi/types/src/module/library/portfolio/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Icon } from '../../../icon';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface PortfolioCategory {
  id: number;
  label: string;
  permalink: string;
}

export interface PortfolioThumbnail {
  src: string;
  width: number;
  height: number;
}

export interface PortfolioPost {
  id: number;
  title: string;
  permalink: string;
  thumbnails: { grid: PortfolioThumbnail; fullwidth: PortfolioThumbnail; }
  categories: PortfolioCategory[];
  classNames: string[];
}

export interface PortfolioMetadata {
  maxNumPages?: number;
  nextPageButtonLabel?: string;
  prevPageButtonLabel?: string;
  wpPagenavi?: string | null;
}

export interface PortfolioAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'conditions' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Title
  title?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Meta
  meta?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  // Pagination
  pagination?: {
    decoration?: Element.Decoration.PickedAttributes<'font'>;
  };

  portfolio?: {
    innerContent?: FormatBreakpointStateAttr<{
      postsNumber?: string;
      includedCategories?: string[];
    }>;
    advanced?: {
      showTitle?: FormatBreakpointStateAttr<OnOff>;
      showCategories?: FormatBreakpointStateAttr<OnOff>;
      showPagination?: FormatBreakpointStateAttr<OnOff>;
      layout?: FormatBreakpointStateAttr<
      'fullwidth' |
      'grid'
      >;
    }
  }

  image?: {
    decoration?: Element.Decoration.PickedAttributes<
    'border' |
    'boxShadow' |
    'filters'
    >;
  };

  overlay?: {
    advanced?: {
      hoverIcon?: Icon.Font.Attributes;
      iconColor?: FormatBreakpointStateAttr<string>;
    };
    decoration?: Element.Decoration.PickedAttributes<'background'>;
  }
}
