// @divi/types/src/colors/store/index.ts
import { type ImmutableObject } from 'seamless-immutable';


export namespace Store {

  /**
   * Type alias for a store state item.
   * It can be either a string or an array of strings.
   *
   * @since ??
   */
  export type StateItem = string | string[];

  /**
   * Represents the state of the store.
   *
   * This StoreState is an immutable object.
   *
   * @since ??
   */
  export interface State {

    /**
     * The default palette.
     */
    defaultPalette: string[];
  };

  export type ImmutableState = ImmutableObject<State>;

  /**
   * StoreInstance interface represents an instance of a store.
   *
   * @since ??
   */
  export interface Instance {

    /**
     * A function that dispatches an action within the store.
     *
     * @example
     *
     * ```ts
     * // Dispatching an action within the store
     * store.dispatch();
     * ```
     */
    dispatch: () => void;


    /**
     * A function that retrieves the current state of the store.
     *
     * @example
     *
     * ```ts
     * // Retrieving the current state of the store
     * const currentState = store.getState();
     * ```
     */
    getState: () => ImmutableState;
  };

  export namespace Actions {
    export namespace UpdateColorPalette {

      /**
       * Represents an action to update a color palette.
       *
       * This interface defines the properties of an object that represents an action to update a color palette.
       * It is typically used as the type of the action in a reducer function.
       *
       * @since ??
       */
      export type ReturnValue = {

        /**
         * The type of the action. This should always be set to 'UPDATE_PALETTE'.
         */
        type: 'UPDATE_PALETTE';

        /**
         * The ID of the color to update. It should be a unique identifier for a color in the palette.
         */
        id: number;

        /**
         * The new color value. It should be a string representing a color in a valid CSS format.
         *
         * @example '#ff0000'
         * @example 'rgb(255, 0, 0)'
         * @example 'rgba(255, 0, 0, 1)'
         * @example 'hsl(0, 100%, 50%)'
         * @example 'hsla(0, 100%, 50%, 1)'
         * @example 'red'
         */
        color: string;
      }
    }

    export type FunctionMap = {
      updateColorPalette: (
        id: UpdateColorPalette.ReturnValue['id'],
        color: UpdateColorPalette.ReturnValue['color'],
      ) => UpdateColorPalette.ReturnValue;
    };

    export type ReturnValue = UpdateColorPalette.ReturnValue;
  };

  export namespace Selectors {
    export type FunctionMap = {
      getDefaultPalette: () => ImmutableState['defaultPalette'];
    }
  }
}
