// @divi/types/src/module/library/post-slider/index.ts
import { type OnOff } from '../../../attr';
import {
  type FormatBreakpointStateAttr,
} from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type AttributesRel } from '../../element/inner-content/attributes-rel';
import { type InternalAttrs } from '../internal';


export interface PostSliderAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.Attributes;

  module?: {
    advanced?: {
        auto?: FormatBreakpointStateAttr<OnOff>;
        autoSpeed?: FormatBreakpointStateAttr<string>;
        autoIgnoreHover?: FormatBreakpointStateAttr<OnOff>;
        htmlAttributes?: Element.Advanced.IdClasses.Attributes;
        link?: Element.Advanced.Link.Attributes;
        text?: Element.Advanced.Text.Attributes;
    },
    decoration?: Pick<Element.Decoration.Attributes,
      'animation' |
      'background' |
      'conditions' |
      'disabledOn' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex' |
      'boxShadow' |
      'border' |
      'filters'
    >;
    meta?: Element.Meta.Attributes;
  }

  title?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  content?: {
    advanced?: {
      showOnMobile?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'bodyFont' |
      'sizing'
    >;
  };

  meta?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
    }
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  };

  image?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
      placement: FormatBreakpointStateAttr<string>;
      showOnMobile?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'boxShadow' |
      'border' |
      'filters'
    >;
  };

  // TODO: Need to recheck it in edit and style component.
  button?: {
    innerContent?: FormatBreakpointStateAttr<{
      text?: string;
      rel?: AttributesRel.AttributeValue;
    }>;
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
      showOnMobile?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Element.Types.Button.Decoration.Attributes;
  };

  post?: {
    advanced?: {
      contentSource?: FormatBreakpointStateAttr<string>;
      orderby?: FormatBreakpointStateAttr<string>;
      currentLoop?: FormatBreakpointStateAttr<OnOff>;
      number?: FormatBreakpointStateAttr<string>;
      categories?: FormatBreakpointStateAttr<string[]>;
      excerptManual?: FormatBreakpointStateAttr<OnOff>;
      excerptLength?: FormatBreakpointStateAttr<string>;
      offset?: FormatBreakpointStateAttr<string>;
    };
  };

  arrows?: {
    advanced?: {
      color?: FormatBreakpointStateAttr<string>;
      enable?: FormatBreakpointStateAttr<OnOff>;
    };
  };

  pagination?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'background'
    >;
  };

  contentOverlay?: {
    advanced?: {
      use?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'background' |
      'border'
    >;
  };

  slideOverlay?: {
    advanced?: {
      use?: FormatBreakpointStateAttr<OnOff>;
    };
    decoration?: Pick<Element.Decoration.Attributes,
      'background'
    >;
  };
}
