// @divi/types/src/module/library/index.ts
import { type ShortcodeModule } from '../../shortcode-module';
import { type PropertyTypeOf } from '../../utils';
import { type AccordionAttrs } from './accordion';
import { type AccordionItemAttrs } from './accordion-item';
import { type AudioAttrs } from './audio';
import { type BlogAttrs } from './blog';
import { type BlurbAttrs } from './blurb';
import { type ButtonAttrs } from './button';
import { type CircleCounterAttrs } from './circle-counter';
import { type CodeAttrs } from './code';
import { type ColumnAttrs } from './column';
import { type CommentsAttrs } from './comments';
import { type ContactFieldAttrs } from './contact-field';
import { type ContactFormAttrs } from './contact-form';
import { type CountdownTimerAttrs } from './countdown-timer';
import { type BarCounterAttrs } from './counter';
import { type BarCountersAttrs } from './counters';
import { type CtaAttrs } from './cta';
import { type DividerAttrs } from './divider';
import { type FilterablePortfolioAttrs } from './filterable-portfolio';
import { type FullwidthCodeAttrs } from './fullwidth-code';
import { type FullwidthHeaderAttrs } from './fullwidth-header';
import { type FullwidthImageAttrs } from './fullwidth-image';
import { type FullwidthMapAttrs } from './fullwidth-map';
import { type FullwidthMenuAttrs } from './fullwidth-menu';
import { type FullwidthPortfolioAttrs } from './fullwidth-portfolio';
import { type FullwidthPostSliderAttrs } from './fullwidth-post-slider';
import { type FullwidthPostTitleAttrs } from './fullwidth-post-title';
import { type FullwidthSliderAttrs } from './fullwidth-slider';
import { type GalleryAttrs } from './gallery';
import { type HeadingAttrs } from './heading';
import { type IconAttrs } from './icon';
import { type ImageAttrs } from './image';
import { type LoginAttrs } from './login';
import { type MapAttrs } from './map';
import { type MapPinAttrs } from './map-pin';
import { type MenuAttrs } from './menu';
import { type NumberCounterAttrs } from './number-counter';
import { type PortfolioAttrs } from './portfolio';
import { type PostNavAttrs } from './post-nav';
import { type PostSliderAttrs } from './post-slider';
import { type PostTitleAttrs } from './post-title';
import { type PricingTableAttrs } from './pricing-table';
import { type PricingTablesAttrs } from './pricing-tables';
import { type RowAttrs } from './row';
import { type SearchAttrs } from './search';
import { type SectionAttrs } from './section';
import { type SidebarAttrs } from './sidebar';
import { type SignupAttrs } from './signup';
import { type SignupCustomFieldAttrs } from './signup-custom-field';
import { type SlideAttrs } from './slide';
import { type SliderAttrs } from './slider';
import { type SocialMediaFollowAttrs } from './social-media-follow';
import { type SocialMediaFollowItemAttrs } from './social-media-follow-item';
import { type TabAttrs } from './tab';
import { type TabsAttrs } from './tabs';
import { type TeamMemberAttrs } from './team-member';
import { type TestimonialAttrs } from './testimonial';
import { type TextAttrs } from './text';
import { type ToggleAttrs } from './toggle';
import { type VideoAttrs } from './video';
import { type VideoSliderAttrs } from './video-slider';
import { type VideoSliderItemAttrs } from './video-slider-item';


export type ModuleAttrsWithChildren = ModuleAttrs & {
  children?: ModuleAttrs;
}

export type ModuleAttrs =
  AccordionAttrs |
  AccordionItemAttrs |
  AudioAttrs |
  BarCounterAttrs |
  BarCountersAttrs |
  BlogAttrs |
  BlurbAttrs |
  ButtonAttrs |
  CircleCounterAttrs |
  CodeAttrs |
  ColumnAttrs |
  CommentsAttrs |
  ContactFieldAttrs |
  ContactFormAttrs |
  CountdownTimerAttrs |
  CtaAttrs |
  DividerAttrs |
  FilterablePortfolioAttrs |
  FullwidthCodeAttrs |
  FullwidthHeaderAttrs |
  FullwidthImageAttrs |
  FullwidthMapAttrs |
  FullwidthMenuAttrs |
  FullwidthPortfolioAttrs |
  FullwidthPostSliderAttrs |
  FullwidthPostTitleAttrs |
  FullwidthSliderAttrs |
  GalleryAttrs |
  HeadingAttrs |
  IconAttrs |
  ImageAttrs |
  LoginAttrs |
  MapAttrs |
  MapPinAttrs |
  MenuAttrs |
  NumberCounterAttrs |
  PortfolioAttrs |
  PostNavAttrs |
  PostSliderAttrs |
  PostTitleAttrs |
  PricingTableAttrs |
  PricingTablesAttrs |
  RowAttrs |
  SearchAttrs |
  SectionAttrs |
  ShortcodeModule.Attributes |
  SidebarAttrs |
  SignupAttrs |
  SignupCustomFieldAttrs |
  SlideAttrs |
  SliderAttrs |
  SocialMediaFollowAttrs |
  SocialMediaFollowItemAttrs |
  TabAttrs |
  TabsAttrs |
  TeamMemberAttrs |
  TestimonialAttrs |
  TextAttrs |
  ToggleAttrs |
  VideoAttrs |
  VideoSliderAttrs |
  VideoSliderItemAttrs;

/**
 * `ModuleAttrAll` needs to be extracted by getting individual attribute using `PropertyTypeOf` of
 * each module attributes then combine them using union types. Attribute can't be retrieved by using
 * `PropertyTypeOf` because it'll return `unknown` or `string` (this somehow happens if there are
 * n amount of child modules' module attributes combined into `ModuleAttrs`).
 *
 * Attribute of `ShortcodeModule.Attributes` is intentionally excluded because including it will cause
 * `ModuleAttrAll` type to be `unknown`. On top of it all of its attributes are `string` and
 * `string` isn't even expected for `ModuleAttr` (that's the reason why `ModuleAttr` omits `string`
 * from `ModuleAttrAll`).
 *
 * @since ??
 */
export type ModuleAttrAll =
  PropertyTypeOf<AccordionAttrs> |
  PropertyTypeOf<AccordionItemAttrs> |
  PropertyTypeOf<AudioAttrs> |
  PropertyTypeOf<BarCounterAttrs> |
  PropertyTypeOf<BarCountersAttrs> |
  PropertyTypeOf<BlogAttrs> |
  PropertyTypeOf<BlurbAttrs> |
  PropertyTypeOf<ButtonAttrs> |
  PropertyTypeOf<CircleCounterAttrs> |
  PropertyTypeOf<CodeAttrs> |
  PropertyTypeOf<ColumnAttrs> |
  PropertyTypeOf<CommentsAttrs> |
  PropertyTypeOf<ContactFieldAttrs> |
  PropertyTypeOf<CountdownTimerAttrs> |
  PropertyTypeOf<CtaAttrs> |
  PropertyTypeOf<DividerAttrs> |
  PropertyTypeOf<FilterablePortfolioAttrs> |
  PropertyTypeOf<FullwidthCodeAttrs> |
  PropertyTypeOf<FullwidthHeaderAttrs> |
  PropertyTypeOf<FullwidthImageAttrs> |
  PropertyTypeOf<FullwidthMapAttrs> |
  PropertyTypeOf<FullwidthMenuAttrs> |
  PropertyTypeOf<FullwidthPortfolioAttrs> |
  PropertyTypeOf<FullwidthPostSliderAttrs> |
  PropertyTypeOf<FullwidthPostTitleAttrs> |
  PropertyTypeOf<FullwidthSliderAttrs> |
  PropertyTypeOf<GalleryAttrs> |
  PropertyTypeOf<HeadingAttrs> |
  PropertyTypeOf<IconAttrs> |
  PropertyTypeOf<ImageAttrs> |
  PropertyTypeOf<LoginAttrs> |
  PropertyTypeOf<MapAttrs> |
  PropertyTypeOf<MapPinAttrs> |
  PropertyTypeOf<MenuAttrs> |
  PropertyTypeOf<NumberCounterAttrs> |
  PropertyTypeOf<PortfolioAttrs> |
  PropertyTypeOf<PostNavAttrs> |
  PropertyTypeOf<PostSliderAttrs> |
  PropertyTypeOf<PostTitleAttrs> |
  PropertyTypeOf<PricingTableAttrs> |
  PropertyTypeOf<PricingTablesAttrs> |
  PropertyTypeOf<RowAttrs> |
  PropertyTypeOf<SearchAttrs> |
  PropertyTypeOf<SectionAttrs> |
  PropertyTypeOf<SidebarAttrs> |
  PropertyTypeOf<SlideAttrs> |
  PropertyTypeOf<SliderAttrs> |
  PropertyTypeOf<SocialMediaFollowAttrs> |
  PropertyTypeOf<SocialMediaFollowItemAttrs> |
  PropertyTypeOf<TabAttrs> |
  PropertyTypeOf<TabsAttrs> |
  PropertyTypeOf<TeamMemberAttrs> |
  PropertyTypeOf<TestimonialAttrs> |
  PropertyTypeOf<TextAttrs> |
  PropertyTypeOf<ToggleAttrs> |
  PropertyTypeOf<VideoAttrs> |
  PropertyTypeOf<VideoSliderAttrs> |
  PropertyTypeOf<VideoSliderItemAttrs>;

export type ModuleAttr = Omit<ModuleAttrAll, 'string'>;

export type ModuleAttrsByName<TModuleName> =
  TModuleName extends 'divi/accordion' ? AccordionAttrs :
  TModuleName extends 'divi/accordion-item' ? AccordionItemAttrs :
  TModuleName extends 'divi/audio' ? AudioAttrs :
  TModuleName extends 'divi/blog' ? BlogAttrs :
  TModuleName extends 'divi/blurb' ? BlurbAttrs :
  TModuleName extends 'divi/button' ? ButtonAttrs :
  TModuleName extends 'divi/circle-counter' ? CircleCounterAttrs :
  TModuleName extends 'divi/code' ? CodeAttrs :
  TModuleName extends 'divi/column' ? ColumnAttrs :
  TModuleName extends 'divi/column-inner' ? ColumnAttrs :
  TModuleName extends 'divi/contact-field' ? ContactFieldAttrs :
  TModuleName extends 'divi/contact-form' ? ContactFormAttrs :
  TModuleName extends 'divi/countdown-timer' ? CountdownTimerAttrs :
  TModuleName extends 'divi/counter' ? BarCounterAttrs :
  TModuleName extends 'divi/counters' ? BarCountersAttrs :
  TModuleName extends 'divi/cta' ? CtaAttrs :
  TModuleName extends 'divi/divider' ? DividerAttrs :
  TModuleName extends 'divi/filterable-portfolio' ? FilterablePortfolioAttrs :
  TModuleName extends 'divi/fullwidth-code' ? FullwidthCodeAttrs :
  TModuleName extends 'divi/fullwidth-header' ? FullwidthHeaderAttrs :
  TModuleName extends 'divi/fullwidth-image' ? FullwidthImageAttrs :
  TModuleName extends 'divi/fullwidth-map' ? FullwidthMapAttrs :
  TModuleName extends 'divi/fullwidth-menu' ? FullwidthMenuAttrs :
  TModuleName extends 'divi/fullwidth-portfolio' ? FullwidthPortfolioAttrs :
  TModuleName extends 'divi/fullwidth-post-slider' ? FullwidthPostSliderAttrs :
  TModuleName extends 'divi/fullwidth-post-title' ? FullwidthPostTitleAttrs :
  TModuleName extends 'divi/fullwidth-slider' ? FullwidthSliderAttrs :
  TModuleName extends 'divi/gallery' ? GalleryAttrs :
  TModuleName extends 'divi/heading' ? HeadingAttrs :
  TModuleName extends 'divi/icon' ? IconAttrs :
  TModuleName extends 'divi/image' ? ImageAttrs :
  TModuleName extends 'divi/login' ? LoginAttrs :
  TModuleName extends 'divi/map' ? MapAttrs :
  TModuleName extends 'divi/map-pin' ? MapPinAttrs :
  TModuleName extends 'divi/menu' ? MenuAttrs :
  TModuleName extends 'divi/number-counter' ? NumberCounterAttrs :
  TModuleName extends 'divi/portfolio' ? PortfolioAttrs :
  TModuleName extends 'divi/post-nav' ? PostNavAttrs :
  TModuleName extends 'divi/post-slider' ? PostSliderAttrs :
  TModuleName extends 'divi/post-title' ? PostTitleAttrs :
  TModuleName extends 'divi/pricing-table' ? PricingTableAttrs :
  TModuleName extends 'divi/pricing-tables' ? PricingTablesAttrs :
  TModuleName extends 'divi/row' ? RowAttrs :
  TModuleName extends 'divi/row-inner' ? RowAttrs :
  TModuleName extends 'divi/search' ? SearchAttrs :
  TModuleName extends 'divi/section' ? SectionAttrs :
  TModuleName extends 'divi/shortcode-module' ? ShortcodeModule.Attributes :
  TModuleName extends 'divi/sidebar' ? SidebarAttrs :
  TModuleName extends 'divi/signup' ? SignupAttrs :
  TModuleName extends 'divi/slide-custom-field' ? SignupCustomFieldAttrs :
  TModuleName extends 'divi/slide' ? SlideAttrs :
  TModuleName extends 'divi/slider' ? SliderAttrs :
  TModuleName extends 'divi/social-media-follow' ? SocialMediaFollowAttrs :
  TModuleName extends 'divi/social-media-follow-item' ? SocialMediaFollowItemAttrs :
  TModuleName extends 'divi/tab' ? TabAttrs :
  TModuleName extends 'divi/tabs' ? TabsAttrs :
  TModuleName extends 'divi/team-member' ? TeamMemberAttrs :
  TModuleName extends 'divi/testimonial' ? TestimonialAttrs :
  TModuleName extends 'divi/text' ? TextAttrs :
  TModuleName extends 'divi/toggle' ? ToggleAttrs :
  TModuleName extends 'divi/video' ? VideoAttrs :
  TModuleName extends 'divi/video-slider' ? VideoSliderAttrs :
  TModuleName extends 'divi/video-slider-item' ? VideoSliderItemAttrs :
  ModuleAttrs;

export * from './accordion';
export * from './accordion-item';
export * from './audio';
export * from './blog';
export * from './blurb';
export * from './button';
export * from './circle-counter';
export * from './code';
export * from './column';
export * from './comments';
export * from './contact-field';
export * from './contact-form';
export * from './countdown-timer';
export * from './counter';
export * from './counters';
export * from './cta';
export * from './divider';
export * from './filterable-portfolio';
export * from './fullwidth-code';
export * from './fullwidth-header';
export * from './fullwidth-image';
export * from './fullwidth-map';
export * from './fullwidth-menu';
export * from './fullwidth-portfolio';
export * from './fullwidth-post-slider';
export * from './fullwidth-post-title';
export * from './fullwidth-slider';
export * from './gallery';
export * from './global-layout';
export * from './heading';
export * from './icon';
export * from './image';
export * from './internal';
export * from './login';
export * from './map';
export * from './map-pin';
export * from './menu';
export * from './number-counter';
export * from './portfolio';
export * from './post-nav';
export * from './post-slider';
export * from './post-title';
export * from './pricing-table';
export * from './pricing-tables';
export * from './row';
export * from './search';
export * from './section';
export * from './sidebar';
export * from './signup';
export * from './signup-custom-field';
export * from './slide';
export * from './slider';
export * from './social-media-follow';
export * from './social-media-follow-item';
export * from './tab';
export * from './tabs';
export * from './team-member';
export * from './testimonial';
export * from './text';
export * from './toggle';
export * from './video';
export * from './video-slider';
export * from './video-slider-item';
