// @divi/types/src/field-library/draggable-list/index.ts
import {
  type FormEvent,
  type MouseEvent,
} from 'react';


export namespace DraggableList {

  /**
   * @typedef {object} Value
   * @property {string} id
   * @property {string} title
   */
  export interface Value {
    id: string;
    title: string;
  }

  /**
   * @typedef {object} OnChangeCallbackParams
   * @property {FormEvent} [event]
   * @property {Value[]} [inputValue]
   */
  export interface OnChangeCallbackParams {
    event?: FormEvent;
    inputValue?: Value[];
  }

  /**
   * @typedef {object} Props
   * @property {string} name Field name.
   * @property {string} [className] Class name.
   * @property {object[]} [defaultValue] Default value.
   * @property {object[]} value Value.
   * @property {string} [addTitle] Add title.
   * @property {boolean} [disableDuplicate] Disable duplicate.
   * @property {boolean} [disableDelete] Disable delete.
   * @property {boolean} [hideAddButton] Hide add button.
   * @property {boolean} ["readonly"] Read only.
   * @property {Function} onChange On change callback.
   * @property {Function} [onClickSettings] On click settings callback.
   * @property {Function} [onClickDuplicate] On click duplicate callback.
   * @property {Function} [onClickDelete] On click delete callback.
   * @property {Function} [onClickMore] On click more callback.
   * @property {Function} [onAddNew] On add new callback.
   * @property {Function} [onContextMenu] On context menu callback.
   * @property {Function} [onDragStart] On drag start callback.
   * @property {Function} [onDragEnd] On drag end callback.
   */
  export interface Props {
    name: string;
    className?: string;
    defaultValue?: Value[];
    value: Value[];
    addTitle?: string;
    disableDuplicate?: boolean;
    disableDelete?: boolean;
    hideAddButton?: boolean;
    readonly?: boolean;
    onChange: (params: OnChangeCallbackParams) => void;
    onClickSettings?: (id: string) => void;
    onClickDuplicate?: (id: string) => void;
    onClickDelete?: (id: string) => void;
    onClickMore?: (id: string, event: MouseEvent) => void;
    onAddNew?: () => void;
    onContextMenu?: (id: string, event: MouseEvent) => void;
    onDragStart?: () => void;
    onDragEnd?: () => void;
  }
}
