// @divi/types/src/css/index.ts
export namespace Css {
  export namespace Properties {

    // TODO replace `AnimationStyleCssProperties` with this.
    export type Animation =
      'animation-delay' |
      'animation-duration' |
      'animation-timing-function' |
      'opacity' |
      'transform';

    // TODO replace `BackgroundStyleCssProperties` with this.
    export type Background =
      'background-blend-mode' |
      'background-color' |
      'background-image' |
      'background-position' |
      'background-repeat' |
      'background-size';

    // TODO replace `TextShadowStyleCssProperties` with this
    export type TextShadow = 'text-shadow';


    // TODO replace `FontStyleCssProperties` with this.
    export type Font =
      'color' |
      'font-family' |
      'font-weight' |
      'font-size' |
      'font-style' |
      'font-variant' |
      'letter-spacing' |
      'line-height' |
      'text-align' |
      'text-decoration-line' |
      'text-decoration-color' |
      'text-decoration-style' |
      'text-transform';

    // TODO replace `ListFontStyleCssProperties` with this.
    // TODO feat(D5, Refactoring) - the use of FontStyleCssProperties is to be removed.
    export type ListFont = Font |
      'list-style-type' |
      'list-style-position' |
      'padding-left';

    // TODO replace `BlockquoteFontStyleCssProperties` with this.
    // TODO feat(D5, Refactoring) - the use of FontStyleCssProperties is to be removed.
    export type BlockquoteFont = Font |
      'border-left-width' |
      'border-left-style' |
      'border-left-color';

    // TODO replace `BorderStyleShorthandCssProperties` with this.
    export type BorderShorthand =
      'border' |
      'border-radius' |
      'border-style';

    // TODO replace `BorderStyleCssProperties` with this.
    export type Border =
      BorderShorthand |
      'border-top-left-radius' |
      'border-top-right-radius' |
      'border-bottom-left-radius' |
      'border-bottom-right-radius' |
      'border-top-style' |
      'border-right-style' |
      'border-bottom-style' |
      'border-left-style' |
      'border-color' |
      'border-top-color' |
      'border-right-color' |
      'border-bottom-color' |
      'border-left-color' |
      'border-width' |
      'border-top-width' |
      'border-right-width' |
      'border-bottom-width' |
      'border-left-width';

    // TODO replace `BoxShadowCssProperties` with this.
    export type BoxShadow = 'box-shadow';

    // TODO replace `FiltersStyleCssProperties` with this.
    export type Filters = 'filter' | 'mix-blend-mode';

    // TODO replace `TransformStyleCssProperties` with this.
    export type Transform =
      'transform' |
      'transform-origin' |
      'transition';

    // TODO replace `TransitionStyleCssProperties` with this.
    export type Transition = 'z-index';

    // TODO replace `DisabledOnStyleCssProperties` with this.
    export type DisabledOn = 'display' | 'opacity';

    // TODO replace `IconStyleCssProperties` with this.
    export type Icon =
      'font-family' |
      'font-size' |
      'font-weight' |
      'line-height' |
      'content' |
      'color' |
      'margin-top' |
      'margin-left';

    // TODO replace `OverflowStyleCssProperties` with this.
    export type Overflow = 'overflow-x' | 'overflow-y';

    // TODO replace `PositionStyleCssProperties` with this.
    export type Position =
      'position' |
      'top' |
      'right' |
      'bottom' |
      'left' |
      'transform';

    // TODO replace `SizingStyleCssProperties` with this.
    export type Sizing =
      'width' |
      'max-width' |
      'margin-left' |
      'margin-right' |
      'min-height' |
      'height' |
      'max-height';

    // TODO replace `SpacingStyleShorthandCssProperties` with this.
    export type SpacingShorthand =
      'margin' |
      'padding';

    // TODO replace `SpacingStyleCssProperties` with this.
    export type Spacing =
      SpacingShorthand |
      'margin-top' |
      'margin-right' |
      'margin-bottom' |
      'margin-left' |
      'padding-top' |
      'padding-right' |
      'padding-bottom' |
      'padding-left';

    // TODO replace `ButtonStyleCssProperties` with this.
    export type Button = 'text-align' |
      'left' |
      'margin-left' |
      'right' |
      'margin-right' |
      'color' |
      'content' |
      'display' |
      'font-family' |
      'font-size' |
      'font-weight' |
      'line-height'

       |
      'opacity';

  }
}
