// @divi/types/src/module/settings/field/index.ts
import { type ChangeEvent, type ReactNode } from 'react';
import { type ImmutableObject } from 'seamless-immutable';

import { type Clipboard } from '../../../clipboard';
import { type FieldLibrary } from '../../../field-library';
import {
  type AttrState,
  type Breakpoint,
} from '../../../format-attr';
import { type RightClickOptions } from '../../../right-click-options';
import {
  type ModuleAttr,
  type ModuleAttrs,
} from '../../library';


export namespace Field {
  export interface OnHasValueParams {
    value: string | number | null | undefined | Record<string, unknown> | Record<number, unknown>;
    defaultValue?: unknown;
  }

  export interface OnChangeParams {
    event?: ChangeEvent<HTMLInputElement>;
    inputValue?: string | string[];
    attrName: string;
    attrSubName?: string;
    moduleId: string;
    responsive?: boolean;
    responsiveMode: Breakpoint;
    stateMode: AttrState;
  }

  // TODO to be replaced with `Field.OnChangeParams`.
  export type OnChangeUnknownParams = Record<string, any>;

  export interface OnResetParams {
    attrName: string;
    attrSubName?: string;
    moduleId: string;
    responsiveMode: Breakpoint;
    stateMode: AttrState;
  }

  export interface GetValueParams {
    attrValue?: ModuleAttr;
    responsiveMode?: Breakpoint;
    stateMode?: AttrState;
  }

  export interface CallbackParams<TAttrs = ModuleAttrs, TAttr = ModuleAttr> {
    attrs: TAttrs;
    attr: TAttr;
    defaultAttr?: TAttr;
    attrName: string;
    responsiveMode: Breakpoint;
    stateMode: AttrState;
    subName?: string;
    hover?: boolean;
    responsive?: boolean;
    sticky?: boolean;
    getValue?: (params: GetValueParams) => any;
  }

  export type InputValue = string | string[] | Record<string, string> | number;

  /**
   * Module field container props.
   *
   * This type is used to define the props of a module field container component.
   *
   * @since ??
   *
   * @typedef {object} Field.ContainerProps
   * @property {string} [id] ID.
   * @property {string} label Label.
   * @property {string} description Description.
   * @property {string} attrName Attribute name.
   * @property {string} [subName] Sub name.
   * @property {string|((params: Field.CallbackParams) => string)} [placeholder] Placeholder.
   * @property {string} [className] Class name.
   * @property {number} [column] Column number.
   * @property {string|((onChangeParams: Field.OnChangeParams) => void)} [onChange] Change callback.
   * @property {string|((onHasValueParams: Field.OnHasValueParams) => boolean)} [onHasValue] Has value callback.
   * @property {string|((onResetParams: Field.OnResetParams) => void)} [onReset] Reset callback.
   * @property {(keyof Field.ContainerProps)[]} [onChangeParams] On change params.
   * @property {ReactNode} [children] Children component.
   * @property {boolean|((params: Field.CallbackParams) => boolean)} [visible] Visible.
   * @property {boolean} [multipleChoices] Multiple choices.
   * @property {(params: Field.GetValueParams) => any} [getValue] Get value callback.
   * @property {string} [groupName] Group name.
   * @property {boolean} [enable] Enable.
   * @property {string} [disabledDescription] Disabled description.
   * @property {ModuleAttr} [defaultAttr] Default attribute.
   * @property {FieldLibrary.Component} [component] Field component.
   * @property {object} [features] Features.
   * @property {boolean} [features.responsive] Responsive.
   * @property {boolean} [features.hover] Hover.
   * @property {boolean} [features.sticky] Sticky.
   * @property {boolean|{ type: 'text' | 'url' | 'image'; }} [features.dynamicContent] Dynamic content.
   */
  export interface ContainerProps {
    id?: string;
    label: string;
    description: string;
    attrName: string;
    subName?: string;
    placeholder?: string | ((params: CallbackParams) => string);
    className?: string;
    column?: number;

    // TODO to be replaced with Field.OnChangeParams.
    onChange?: string | ((onChangeParams: Record<string, any>) => void);
    onHasValue?: string | ((onHasValueParams: OnHasValueParams) => boolean);
    onReset?: string | ((onResetParams: OnResetParams) => void);
    enableReset?: boolean;
    onChangeParams?: (keyof Field.ContainerProps)[];
    children?: ReactNode;
    visible?: boolean | ((params: CallbackParams) => boolean);
    multipleChoices?: boolean;
    getValue?: (params: GetValueParams) => any;
    groupName?: string;
    enable?: boolean;
    disabledDescription?: string;
    defaultAttr?: ModuleAttr;
    component?: FieldLibrary.Component;
    features?: {
      responsive?: boolean;
      hover?: boolean;
      sticky?: boolean;
      dynamicContent?: false | {
        type: 'text' | 'url' | 'image';
      }
    }
  }

  export interface Props extends ContainerProps {
    attrs: ImmutableObject<ModuleAttrs>;
    attr: ModuleAttr;
    appState: AttrState;
    appBreakpoint: Breakpoint;
    defaultValue?: ModuleAttr;
    lastClipboardItem: ImmutableObject<Clipboard.Item.ModuleFieldAttribute>;
    hasSync?: boolean;
    isSynced?: boolean;
    onUnsyncGlobalModuleAttribute?: () => void;
    setRightClickOptions: (params: {
      options: RightClickOptions.Option.Items;
      cursorPosition: RightClickOptions.Cursor.Position;
      optionType: RightClickOptions.Owner.Item['optionType'];
      owner: RightClickOptions.Owner.Item;
      windowLocation: RightClickOptions.Window.Location;
    }) => void;
  }

  // Previously called `ModuleModalFieldBaseProps`
  export interface BaseProps {
    container?: boolean;
    groupName?: string;
    id?: string;
    priority?: number;
    render?: boolean;
  }
}

// TODO to be replaced with `Field.OnChangeParams`.
export type OnChangeParams = Record<string, any>;
