// @divi/types/src/module/library/contact-form/index.ts
import { type OnOff } from '../../../attr';
import {
  type FormatBreakpointStateAttr,
} from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import {
  type FieldElementAttr,
} from '../../options';
import {
  type InternalAttrs,
} from '../internal';


export interface ContactFormAttrs extends InternalAttrs {

  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: Css.AttributeValue;

  module?: {
    advanced?: {
        htmlAttributes?: Element.Advanced.IdClasses.Attributes;
        link?: Element.Advanced.Link.Attributes;
        text?: Element.Advanced.Text.Attributes;
        successMessage?: FormatBreakpointStateAttr<string>;
        spamProtection?: Element.Advanced.SpamProtection.Attributes;
    },
    decoration?: Pick<Element.Decoration.Attributes,
      'animation' |
      'background' |
      'conditions' |
      'disabledOn' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex' |
      'border' |
      'boxShadow' |
      'filters'
    >;
    meta?: Element.Meta.Attributes;
  },

  // Contact Form title.
  title?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
    innerContent?: FormatBreakpointStateAttr<string>;
  },

  // Contact form normal captcha.
  captcha?: {
    decoration?: Pick<Element.Decoration.Attributes,
      'font'
    >;
  },

  // Field: Normal, Focused, Placeholder.
  field?: FieldElementAttr;

  // Button: Form submit button.
  button?: Element.Types.Button.Attributes;

  // Redirect URL.
  redirect?: {
    advanced?: {
      useRedirect?: FormatBreakpointStateAttr<OnOff>;
    };
    innerContent?: FormatBreakpointStateAttr<string>;
  },

  // Contact Form :: Email :: receiver email address and email message content.
  email?: {
    advanced?: {
      receiver?: FormatBreakpointStateAttr<string>;
    };
    innerContent?: FormatBreakpointStateAttr<string>;
  }
}
