// @divi/types/src/module/library/blog/index.ts
import { type OnOff } from '../../../attr';
import { type FormatBreakpointStateAttr } from '../../../format-attr';
import { type Css } from '../../css';
import { type Element } from '../../element';
import { type InternalAttrs } from '../internal';


export interface BlogCssAttr extends Css.AttributeValue {
  title?: string;
  content?: string;
  postMeta?: string;
  pagenavi?: string;
  featuredImage?: string;
  readMore?: string;
}

export type BlogCssGroupAttr = FormatBreakpointStateAttr<BlogCssAttr>;


export interface BlogAttrs extends InternalAttrs {

  css?: BlogCssGroupAttr;

  // Module
  module?: {
    meta?: Element.Meta.Attributes;
    advanced?: {
      htmlAttributes?: Element.Advanced.IdClasses.Attributes;
      link?: Element.Advanced.Link.Attributes;
      text?: Element.Advanced.Text.Attributes;
    };
    decoration?: Element.Decoration.PickedAttributes<
      'background' |
      'filters' |
      'animation' |
      'sizing' |
      'boxShadow' |
      'spacing' |
      'transform' |
      'disabledOn' |
      'overflow' |
      'position' |
      'scroll' |
      'sticky' |
      'transition' |
      'zIndex'
    >;
  };

  // Image
  image?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
    },
    decoration?: Element.Decoration.PickedAttributes<
      'filters' |
      'border' |
      'boxShadow'
    >;
  };

  // Title
  title?: {
    decoration?: Element.Decoration.PickedAttributes< 'font' >;
  };

  // Meta
  meta?: {
    advanced?: {
      showAuthor?: FormatBreakpointStateAttr<OnOff>;
      showDate?: FormatBreakpointStateAttr<OnOff>;
      showCategories?: FormatBreakpointStateAttr<OnOff>;
      showComments?: FormatBreakpointStateAttr<OnOff>;
    },
    decoration?: Element.Decoration.PickedAttributes< 'font' >;
  };

  // Read more
  readMore?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
    },
    decoration?: Element.Decoration.PickedAttributes< 'font' >;
  };

  // Pagination
  pagination?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
    },
    decoration?: Element.Decoration.PickedAttributes< 'font' >;
  };

  // Content
  content?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
    },
    decoration?: Element.Decoration.PickedAttributes< 'bodyFont' >;
  };

  // Post
  post?: {
    advanced?: {
      type?: FormatBreakpointStateAttr<string>;
      number?: FormatBreakpointStateAttr<string>;
      categories?: FormatBreakpointStateAttr<string[]>;
      dateFormat?: FormatBreakpointStateAttr<string>;
      excerptContent?: FormatBreakpointStateAttr<OnOff>;
      excerptManual?: FormatBreakpointStateAttr<OnOff>;
      excerptLength?: FormatBreakpointStateAttr<string>;
      offset?: FormatBreakpointStateAttr<string>;
      showExcerpt?: FormatBreakpointStateAttr<OnOff>;
      currentLoop?: FormatBreakpointStateAttr<OnOff>;
    },
    decoration?: Element.Decoration.PickedAttributes< 'border' >;
  };

  overlay?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
    },
    decoration?: Element.Decoration.PickedAttributes< 'background' >;
  };
  overlayIcon?: {
    decoration?: Element.Decoration.PickedAttributes< 'icon' >;
  };

  fullwidth?: {
    advanced?: {
      enable?: FormatBreakpointStateAttr<OnOff>;
    },
    decoration?: Element.Decoration.PickedAttributes< 'border' >;
  };

  masonry?: {
    decoration?: Element.Decoration.PickedAttributes< 'background' >;
  };
}

export interface BlogPostCategory {
  id: number;
  name: string;
  slug: string;
  link: string;
}
export interface BlogPost {
  id: number;
  classNames: string;
  title: string;
  permalink: string;
  thumbnail: {
    src: string;
    alt: string;
    width: string;
    height: string;
  } | null;
  content: string;
  date: string;
  comment: string;
  author: {
    name: string;
    link: string;
  };
  categories: BlogPostCategory[]
}

export interface BlogMetadata {
  maxNumPages?: number;
  wpPagenavi?: string;
}
