// @divi/types/src/shortcode-module/advanced-field/index.ts
/* eslint-disable @typescript-eslint/naming-convention */
import { type ShortcodeModuleAdvancedFieldDefinitionBorder } from './border';
import { type ShortcodeModuleAdvancedFieldDefinitionBoxShadow } from './box-shadow';
import { type ShortcodeModuleAdvancedFieldDefinitionButton } from './button';
import { type ShortcodeModuleAdvancedFieldDefinitionFilters } from './filters';
import { type ShortcodeModuleAdvancedFieldDefinitionFont } from './font';
import { type ShortcodeModuleAdvancedFieldDefinitionFormField } from './form-field';
import { type ShortcodeModuleAdvancedFieldDefinitionPosition } from './position';
import { type ShortcodeModuleAdvancedFieldDefinitionScroll } from './scroll';
import { type ShortcodeModuleAdvancedFieldDefinitionMaxWidth } from './sizing';
import { type ShortcodeModuleAdvancedFieldDefinitionSpacing } from './spacing';
import { type ShortcodeModuleAdvancedFieldDefinitionSticky } from './sticky';
import { type ShortcodeModuleAdvancedFieldDefinitionText } from './text';
import { type ShortcodeModuleAdvancedFieldDefinitionTransform } from './transform';
import { type ShortcodeModuleAdvancedFieldDefinitionZIndex } from './z-index';


export * from './border';
export * from './box-shadow';
export * from './button';
export * from './filters';
export * from './font';
export * from './form-field';
export * from './position';
export * from './scroll';
export * from './sizing';
export * from './spacing';
export * from './sticky';
export * from './text';
export * from './text-shadow';
export * from './transform';
export * from './z-index';

export interface ShortcodeModuleAdvancedFieldDefinition {
  animation?: [];
  background?: boolean | {
    css?: {
      hover?: string;
      important?: string | boolean;
      main?: string;
      mask?: string;
      mask_selector?: string;
      pattern?: string;
      pattern_selector?: string;
    };
    has_background_color_toggle?: boolean;
    options?: Record<
      string,
      boolean | [] | string | Record<string, string | string[] | boolean | Record<string, string>>
    >;
    settings?: {
      color?: string;
      tab_slug?: string;
      toggle_slug?: string;
    };
    use_background_color?: boolean | string;
    use_background_color_gradient?: boolean | string;
    use_background_color_reset?: boolean | string;
    use_background_image?: boolean | string;
    use_background_image_parallax?: boolean;
    use_background_mask?: boolean;
    use_background_pattern?: boolean;
    use_background_video?: boolean;
  };
  border?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_css?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_dropdown_menus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_dropdown_menus_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_form?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_form_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_form_field?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_form_field_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_form_notice?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_form_notice_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_fields?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_fields_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_fullwidth?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_image?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_portrait?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_price?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_radio_button?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_radio_button_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_sale_badge?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_selected_radio_button?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_selected_radio_button_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_table?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_table_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_table_row?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_table_row_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_table_cell?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_table_cell_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_tooltip?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  border_tooltip_focus?: ShortcodeModuleAdvancedFieldDefinitionBorder;
  borders?: Record<string, [] | boolean | ShortcodeModuleAdvancedFieldDefinitionBorder>;
  box_shadow?: Record<string, [] | boolean | ShortcodeModuleAdvancedFieldDefinitionBoxShadow>;
  button?: boolean | Record<string, ShortcodeModuleAdvancedFieldDefinitionButton>;
  child_filters?: ShortcodeModuleAdvancedFieldDefinitionFilters['child_filters_target'];
  dividers?: [];
  disabled_on?: [];
  filters?: [] | ShortcodeModuleAdvancedFieldDefinitionFilters;
  fonts?: boolean | Record<string, ShortcodeModuleAdvancedFieldDefinitionFont>;
  form_field?: Record<string, ShortcodeModuleAdvancedFieldDefinitionFormField>;
  height?: [] | boolean | {
    css?: {
      main?: string;
    };
    options?: {
      height?: {
        default?: string;
        default_tablet?: string;
        default_phone?: string;
      }
    };
    use_height?: boolean;
    use_max_height?: boolean;
    use_min_height?: boolean;
  };
  icon_settings?: {
    css?: {
      main?: string;
    }
  };
  image?: {
    css?: {
      main?: string | string[];
    }
  };
  image_icon?: Record<string, {
    margin_padding?: {
      css?: Record<string, string>;
    };
    option_category?: string;
    tab_slug?: string;
    toggle_slug?: string;
    label?: string;
    css?: Record<string, string>;
    disable_toggle?: boolean;
  }>;
  link_options?: boolean | [];
  margin_padding?: boolean | ShortcodeModuleAdvancedFieldDefinitionSpacing;
  max_width?: boolean | [] | ShortcodeModuleAdvancedFieldDefinitionMaxWidth;
  module?: [];
  overflow?: boolean | [] | {
    default?: string;
  };
  position_fields?: [] | boolean | ShortcodeModuleAdvancedFieldDefinitionPosition;
  scroll_effects?: boolean | ShortcodeModuleAdvancedFieldDefinitionScroll;
  sticky?: boolean | ShortcodeModuleAdvancedFieldDefinitionSticky;
  text?: boolean | [] | ShortcodeModuleAdvancedFieldDefinitionText;
  text_shadow?: [] | {
    default?: boolean | [];
  };
  transform?: [] | ShortcodeModuleAdvancedFieldDefinitionTransform;
  transition?: boolean | [];
  z_index?: [] | boolean | ShortcodeModuleAdvancedFieldDefinitionZIndex;
}
