// @divi/types/src/module/settings/group/library/animation/index.ts
import { type FieldLibrary } from '../../../../../field-library';
import { type Element } from '../../../../element';
import { type GroupFieldProps } from '../../field';
import { type GroupProps as GroupPropsSource } from '../../props';


export namespace Animation {
  export interface GroupProps extends GroupPropsSource {
    defaultGroupAttr?: Element.Decoration.Animation.Attributes;
    fields?: {
      style?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Animation.Attributes;
        component?: FieldLibrary.ComponentMap['divi/button-options'];
      };
      direction?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Animation.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      duration?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Animation.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      delay?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Animation.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      startingOpacity?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Animation.Attributes;
        component?: FieldLibrary.ComponentMap['divi/range'];
      };
      speedCurve?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Animation.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      repeat?: GroupFieldProps & {
        defaultAttr?: Element.Decoration.Animation.Attributes;
        component?: FieldLibrary.ComponentMap['divi/select'];
      };
      [key: string]: GroupFieldProps & {
        component?: FieldLibrary.ComponentMap['divi/range'] | FieldLibrary.ComponentMap['divi/button-options'] | FieldLibrary.ComponentMap['divi/select'];
      };
    }
  }
}
